---
title: "Contents"
description: "Tutorials on some of the common steps during scRNAseq data analysis using three popular toolkits: [ {{< fa brands r-project >}} Seurat](https://satijalab.org/seurat/), [{{< fa brands r-project >}} Bioconductor](https://bioconductor.org/packages/release/bioc/html/scran.html) and [{{< fa brands python >}} Scanpy](https://scanpy.readthedocs.io/en/stable/)."
date: ""
author: ""
toc: false
number-sections: false
sidebar: false
code-tools: false
format: html
---



::: {.callout-tip}
We perform the same steps with all three toolkits, but there are some small differences as all methods are not implemented everywhere. It's up to you which toolkit you want to work with.  
You can run the labs either in a Singularity container on Uppmax (recommended) or using Docker locally on your system. Instructions on running the labs are provided [here](/other/containers.qmd).  
To download file, *Right click > Save Link As...*.
:::

| Topic | {{< fa brands r-project >}} Seurat | {{< fa brands r-project >}} Bioconductor | {{< fa brands python >}} Scanpy |
|:--------|:---------------------------:|:---------------------------------:|:------------------------:|
| {{< fa clipboard-list >}} Quality Control | [{{< fa file-lines >}}](labs/seurat/seurat_01_qc.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_01_qc.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_01_qc.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_01_qc.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_01_qc.html)  [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_01_qc.ipynb) |
| {{< fa brands hubspot >}} Dimensionality reduction | [{{< fa file-lines >}}](labs/seurat/seurat_02_dimred.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_02_dimred.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_02_dimred.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_02_dimred.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_02_dimred.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_02_dimred.ipynb) |
| {{< fa diagram-project >}} Data integration | [{{< fa file-lines >}}](labs/seurat/seurat_03_integration.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_03_integration.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_03_integration.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_03_integration.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_03_integration.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_03_integration.ipynb) |
| {{< fa circle-nodes >}} Clustering | [{{< fa file-lines >}}](labs/seurat/seurat_04_clustering.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_04_clustering.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_04_clustering.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_04_clustering.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_04_clustering.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_04_clustering.ipynb) |
| {{< fa chart-simple >}} Differential expression | [{{< fa file-lines >}}](labs/seurat/seurat_05_dge.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_05_dge.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_05_dge.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_05_dge.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_05_dge.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_05_dge.ipynb) |
| {{< fa bacterium >}} Celltype prediction | [{{< fa file-lines >}}](labs/seurat/seurat_06_celltyping.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_06_celltyping.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_06_celltyping.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_06_celltyping.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_06_celltyping.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_06_celltyping.ipynb) |
| {{< fa route >}} Trajectory inference | [{{< fa file-lines >}}](labs/seurat/seurat_08_trajectory.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_08_trajectory.qmd) | | [{{< fa file-lines >}}](labs/scanpy/scanpy_08_trajectory.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_08_trajectory.ipynb) |
| {{< fa microscope >}} Spatial transcriptomics | [{{< fa file-lines >}}](labs/seurat/seurat_07_spatial.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_07_spatial.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_07_spatial.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_07_spatial.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_07_spatial.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_07_spatial.ipynb) |

Many additional information and explanations can be found in the [Single Cell Glossary](https://nbisweden.github.io/single-cell-pbl/glossary_of_terms_single_cell.html)

## Useful resources

- Single cell RNA-seq course at from Hemberg lab [{{< fa square-arrow-up-right >}}](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
- Single cell RNA-seq course in Python [{{< fa square-arrow-up-right >}}](https://chanzuckerberg.github.io/scRNA-python-workshop/intro/about)
- Single cell RNA-seq course at Broad [{{< fa square-arrow-up-right >}}](https://broadinstitute.github.io/2019_scWorkshop/)
- Repository listing many scRNA-seq tools [{{< fa square-arrow-up-right >}}](https://github.com/seandavi/awesome-single-cell)
- SingleCellExperiment objects for many datasets [{{< fa square-arrow-up-right >}}](https://hemberg-lab.github.io/scRNA.seq.datasets/)
- Conquer datasets - many different datasets based on a salmon pipeline [{{< fa square-arrow-up-right >}}](http://imlspenticton.uzh.ch:3838/conquer/)
- The Human Cell Atlas project [{{< fa square-arrow-up-right >}}](https://www.humancellatlas.org/)
- The github repository for this course [{{< fa brands github >}}](https://github.com/NBISweden/workshop-scRNAseq)
- Bitbucket repository for QC report scripts [{{< fa brands bitbucket >}}](https://bitbucket.org/asbj/qc-summary_scrnaseq)
- Bitbucket repository for the NBIS scRNAseq pipeline [{{< fa brands bitbucket >}}](https://bitbucket.org/scilifelab-lts/lts-workflows-sm-scrnaseq)

***
