---
layout: default
title: Single cell RNA-seq analysis course
---


# Single cell RNA-seq analysis course
Scilifelab Solna, Rooms Air & Fire,  2020-01-27 - 2020-01-29 (and 2020-01-30 optional day)

### Important dates

Application open: September 13, 2019

Application deadline: December 16, 2019

Confirmation to accepted students: December 20, 2019

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/410/list.svg" width="40" height="40"> Course content
***

The first 3 days of the course will cover the basic steps in single cell RNAseq (scRNAseq) processing and data analysis. Topics covered will include:

* An overview of the current scRNAseq technologies
* Basic overview of pipelines for processing raw reads into expression values
* Quality control of scRNAseq data
* Dimensionality reduction and clustering techniques
* Data normalization
* Differential gene expression for scRNAseq data
* Trajectory analysis
* Comparison of different analysis pipelines such as Seurat, Scran and Scanpy

On the fourth day we offer a "Bring your own data" day where you can analyse your data using the tools that you have discovered during the course, with TAs present to help you run the analyses and discuss your results. If you do not have your own data you can still select a public dataset of interest and analyse during that day.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/3874/contact-form.svg" width="40" height="40"> Application
***

###### How to apply

Fill in this [application](https://forms.gle/6VFVTFep6iYk7wLr9) by December 16th and please indicate whether you are planning to participate on the optional day of January 30th. 

###### Course fee

A course fee* of 2000 SEK will be invoiced to accepted participants. This includes lunches, coffee and course dinner.

*Please note that NBIS cannot invoice individuals

###### Entry requirements


Required for being able to follow the course and complete the computer exercises

* Basic knowledge in R/Linux.
* Be able to bring your own laptop with R installed for the practical computational exercises. Instructions on installation will be sent by email to accepted participants.
* Some programming/scripting experience is required (preferably in R).


* **Desirable**: Previous experience with RNA-seq analysis and/or participation in NGS/RNA-seq course is an advantage.

Due to limited space the course can accommodate maximum of 25 participants. If we receive more applications, participants will be selected based on several criteria. Selection criteria include correct entry requirements, motivation to attend the course as well as gender and geographical balance.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/158264/schedule.svg" width="40" height="40"> Schedule
***

Course schedule can be found here: [Schedule]() (TBA)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/6672/exercise.svg" width="40" height="40"> Exercises
***

All exercises for the afternoon sessions can be found at [Exercises](). (TBA)

For working on Uppmax: to use the allocations we have for the course, please look [here](login.md).

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/19652/maths-class-materials-cross-of-a-pencil-and-a-ruler.svg" width="40" height="40"> Precourse material
***

Please read carefully the [Precourse material](precourse) before the course start. 

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/4199/placeholder-on-a-map.svg" width="40" height="40"> Address and travel suggestions
***

[Travel Info](travel.md)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/38706/group-of-people.svg" width="40" height="40"> Course leaders
***

If you have any questions regarding the course, please email to Åsa Björklund or Paulo Czarnewski

* [Åsa Björklund](http://nbis.se/about/staff/asa-bjorklund/) (asa.bjorklund@scilifelab.se)
* [Paulo Czarnewski](https://nbis.se/about/staff/paulo-czarnewski/) (paulo.czarnewski@scilifelab.se)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/19262/link.svg" width="40" height="40"> Useful links
***

* [Single cell RNA-seq course at from Hemberg lab](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
* [Single cell RNA-seq course at Broad](https://broadinstitute.github.io/2019_scWorkshop/)
* [Repository listing many scRNA-seq tools](https://github.com/seandavi/awesome-single-cell)
* [SingleCellExperiment objects for many datasets](https://hemberg-lab.github.io/scRNA.seq.datasets/)
* [Conquer datasets - many different datasets based on a salmon pipeline](http://imlspenticton.uzh.ch:3838/conquer/)
* [The Human Cell Atlas project](https://www.humancellatlas.org/) 
* [The github repository for this course. Also contains the R markdown scripts that was used to generate all the exercises](https://github.com/NBISweden/workshop-scRNAseq)
* [Bitbucket repository for QC report scripts](https://bitbucket.org/asbj/qc-summary_scrnaseq)
* [Bitbucket repository for the NBIS scRNAseq pipeline](https://bitbucket.org/scilifelab-lts/lts-workflows-sm-scrnaseq)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/83468/navigation-history-interface-symbol-of-a-clock-with-an-arrow.svg" width="40" height="40"> Past courses
***

* [Stockholm February 4-6, 2019](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2019-02-04/)
* [Stockholm May 21-23, 2018](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2018-05-21/)
* [Uppsala October 2-4, 2017](https://scilifelab.github.io/courses/scrnaseq/1710/)

