<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Assignment of cell identities based on gene expression patterns using reference data.">

<title> Celltype prediction</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo." class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-bacterium" aria-label="bacterium"></i> Celltype prediction</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit</p>
                  <div>
        <div class="description">
          Assignment of cell identities based on gene expression patterns using reference data.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">03-Jan-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#meta-ct_read" id="toc-meta-ct_read" class="nav-link active" data-scroll-target="#meta-ct_read"><span class="header-section-number">1</span> Read data</a></li>
  <li><a href="#meta-ct_ref" id="toc-meta-ct_ref" class="nav-link" data-scroll-target="#meta-ct_ref"><span class="header-section-number">2</span> Reference data</a></li>
  <li><a href="#scmap" id="toc-scmap" class="nav-link" data-scroll-target="#scmap"><span class="header-section-number">3</span> scMap</a>
  <ul>
  <li><a href="#scmap-cluster" id="toc-scmap-cluster" class="nav-link" data-scroll-target="#scmap-cluster"><span class="header-section-number">3.1</span> scMap cluster</a></li>
  </ul></li>
  <li><a href="#scmap-cell" id="toc-scmap-cell" class="nav-link" data-scroll-target="#scmap-cell"><span class="header-section-number">4</span> scMap cell</a></li>
  <li><a href="#meta-ct_scpred" id="toc-meta-ct_scpred" class="nav-link" data-scroll-target="#meta-ct_scpred"><span class="header-section-number">5</span> scPred</a></li>
  <li><a href="#meta-ct_compare" id="toc-meta-ct_compare" class="nav-link" data-scroll-target="#meta-ct_compare"><span class="header-section-number">6</span> Compare results</a></li>
  <li><a href="#meta-ct_gsea" id="toc-meta-ct_gsea" class="nav-link" data-scroll-target="#meta-ct_gsea"><span class="header-section-number">7</span> GSEA with celltype markers</a>
  <ul>
  <li><a href="#meta-ct_gsea_deg" id="toc-meta-ct_gsea_deg" class="nav-link" data-scroll-target="#meta-ct_gsea_deg"><span class="header-section-number">7.1</span> DEG overlap</a></li>
  <li><a href="#meta-ct_gsea_annot" id="toc-meta-ct_gsea_annot" class="nav-link" data-scroll-target="#meta-ct_gsea_annot"><span class="header-section-number">7.2</span> With annotated gene sets</a></li>
  </ul></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">8</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run R commands unless otherwise specified.</p>
</div>
</div>
<p>Celltype prediction can either be performed on indiviudal cells where each cell gets a predicted celltype label, or on the level of clusters. All methods are based on similarity to other datasets, single cell or sorted bulk RNAseq, or uses known marker genes for each celltype.<br>
We will select one sample from the Covid data, <code>ctrl_13</code> and predict celltype by cell on that sample.<br>
Some methods will predict a celltype to each cell based on what it is most similar to even if the celltype of that cell is not included in the reference. Other methods include an uncertainty so that cells with low similarity scores will be unclassified.<br>
There are multiple different methods to predict celltypes, here we will just cover a few of those.</p>
<p>Here we will use a reference PBMC dataset from the <code>scPred</code> package which is provided as a Seurat object with counts. And we will test classification based on the <code>scPred</code> and <code>scMap</code> methods. Finally we will use gene set enrichment predict celltype based on the DEGs of each cluster.</p>
<section id="meta-ct_read" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="meta-ct_read"><span class="header-section-number">1</span> Read data</h2>
<p>First, lets load required libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(scater)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(scran)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(dplyr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(cowplot)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pheatmap)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(rafalib)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(scPred)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(scmap)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s read in the saved Covid-19 data object from the clustering step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load the data and select 'ctrl_13` sample</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>alldata <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/results/covid_qc_dr_int_cl.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s read in the saved Covid-19 data object from the clustering step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ctrl.sce <span class="ot">&lt;-</span> alldata[, alldata<span class="sc">@</span>colData<span class="sc">$</span>sample <span class="sc">==</span> <span class="st">'ctrl.13'</span>]</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># remove all old dimensionality reductions as they will mess up the analysis further down</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDims</span>(ctrl.sce) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="meta-ct_ref" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="meta-ct_ref"><span class="header-section-number">2</span> Reference data</h2>
<p>Load the reference dataset with annotated labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> scPred<span class="sc">::</span>pbmc_1</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>reference</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
32838 features across 3500 samples within 1 assay 
Active assay: RNA (32838 features, 0 variable features)</code></pre>
</div>
</div>
<p>Convert to a SCE object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">=</span> Seurat<span class="sc">::</span><span class="fu">as.SingleCellExperiment</span>(reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Rerun analysis pipeline. Run normalization, feature selection and dimensionality reduction</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Normalize</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">&lt;-</span> <span class="fu">computeSumFactors</span>(ref.sce)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(ref.sce)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Variable genes</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>var.out <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(ref.sce, <span class="at">method=</span><span class="st">"loess"</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>hvg.ref <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(var.out, <span class="at">n=</span><span class="dv">1000</span>)  </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Dim reduction</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(ref.sce, <span class="at">exprs_values =</span> <span class="st">"logcounts"</span>,  <span class="at">scale =</span> T,</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>              <span class="at">ncomponents =</span> <span class="dv">30</span>, <span class="at">subset_row =</span> hvg.ref)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(ref.sce, <span class="at">dimred =</span> <span class="st">"PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(ref.sce,<span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"cell_type"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-7-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Run all steps of the analysis for the ctrl sample as well. Use the clustering from the integration lab with resolution 0.5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Normalize</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>ctrl.sce <span class="ot">&lt;-</span> <span class="fu">computeSumFactors</span>(ctrl.sce)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>ctrl.sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(ctrl.sce)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Variable genes</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>var.out <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(ctrl.sce, <span class="at">method=</span><span class="st">"loess"</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>hvg.ctrl <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(var.out, <span class="at">n=</span><span class="dv">1000</span>)  </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Dim reduction</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>ctrl.sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(ctrl.sce, <span class="at">exprs_values =</span> <span class="st">"logcounts"</span>,  <span class="at">scale =</span> T, <span class="at">ncomponents =</span> <span class="dv">30</span>, <span class="at">subset_row =</span> hvg.ctrl)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>ctrl.sce <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(ctrl.sce, <span class="at">dimred =</span> <span class="st">"PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(ctrl.sce,<span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"louvain_SNNk15"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-9-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="scmap" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="scmap"><span class="header-section-number">3</span> scMap</h2>
<p>The scMap package is one method for projecting cells from a scRNA-seq experiment on to the cell-types or individual cells identified in a different experiment. It can be run on different levels, either projecting by cluster or by single cell, here we will try out both.</p>
<p>For scmap cell type labels must be stored in the <code>cell_type1</code> column of the <code>colData</code> slots, and gene ids that are consistent across both datasets must be stored in the <code>feature_symbol</code> column of the <code>rowData</code> slots.</p>
<section id="scmap-cluster" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="scmap-cluster"><span class="header-section-number">3.1</span> scMap cluster</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add in slot cell_type1</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>ref.sce<span class="sc">@</span>colData<span class="sc">$</span>cell_type1 <span class="ot">=</span> ref.sce<span class="sc">@</span>colData<span class="sc">$</span>cell_type</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create a rowData slot with feature_symbol</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>rd <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">feature_symbol=</span><span class="fu">rownames</span>(ref.sce))</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(rd) <span class="ot">=</span> <span class="fu">rownames</span>(ref.sce)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(ref.sce) <span class="ot">=</span> rd</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co"># same for the ctrl dataset</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># create a rowData slot with feature_symbol</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>rd <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">feature_symbol=</span><span class="fu">rownames</span>(ctrl.sce))</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(rd) <span class="ot">=</span> <span class="fu">rownames</span>(ctrl.sce)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(ctrl.sce) <span class="ot">=</span> rd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we can select variable features in both datasets.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select features</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(ctrl.sce) <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">counts</span>(ctrl.sce))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">logcounts</span>(ctrl.sce) <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">logcounts</span>(ctrl.sce))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>ctrl.sce <span class="ot">&lt;-</span> <span class="fu">selectFeatures</span>(ctrl.sce, <span class="at">suppress_plot =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(ref.sce) <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">counts</span>(ref.sce))</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="fu">logcounts</span>(ref.sce) <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">logcounts</span>(ref.sce))</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">&lt;-</span> <span class="fu">selectFeatures</span>(ref.sce, <span class="at">suppress_plot =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we need to index the reference dataset by cluster, default is the clusters in <code>cell_type1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">&lt;-</span> <span class="fu">indexCluster</span>(ref.sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we project the Covid-19 dataset onto that index.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>project_cluster <span class="ot">&lt;-</span> <span class="fu">scmapCluster</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">projection =</span> ctrl.sce,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">index_list =</span> <span class="fu">list</span>(</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">ref =</span> <span class="fu">metadata</span>(ref.sce)<span class="sc">$</span>scmap_cluster_index</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># projected labels</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(project_cluster<span class="sc">$</span>scmap_cluster_labs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     B cell  CD4 T cell  CD8 T cell         cDC       cMono      ncMono 
         66         108         133          38         217         144 
    NK cell         pDC Plasma cell  unassigned 
        256           2           3         208 </code></pre>
</div>
</div>
<p>Then add the predictions to metadata and plot UMAP.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add in predictions</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>ctrl.sce<span class="sc">@</span>colData<span class="sc">$</span>scmap_cluster <span class="ot">&lt;-</span> project_cluster<span class="sc">$</span>scmap_cluster_labs</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(ctrl.sce,<span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"scmap_cluster"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-14-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="scmap-cell" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="scmap-cell"><span class="header-section-number">4</span> scMap cell</h2>
<p>We can instead index the refernce data based on each single cell and project our data onto the closest neighbor in that dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ref.sce <span class="ot">&lt;-</span> <span class="fu">indexCell</span>(ref.sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Again we need to index the reference dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>project_cell <span class="ot">&lt;-</span> <span class="fu">scmapCell</span>(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">projection =</span> ctrl.sce,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">index_list =</span> <span class="fu">list</span>(</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">ref =</span> <span class="fu">metadata</span>(ref.sce)<span class="sc">$</span>scmap_cell_index</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We now get a table with index for the 5 nearest neigbors in the reference dataset for each cell in our dataset. We will select the celltype of the closest neighbor and assign it to the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>cell_type_pred <span class="ot">&lt;-</span> <span class="fu">colData</span>(ref.sce)<span class="sc">$</span>cell_type1[project_cell<span class="sc">$</span>ref[[<span class="dv">1</span>]][<span class="dv">1</span>,]]</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(cell_type_pred)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>cell_type_pred
     B cell  CD4 T cell  CD8 T cell         cDC       cMono      ncMono 
        101         192         266          22         213         189 
    NK cell         pDC Plasma cell 
        188           2           2 </code></pre>
</div>
</div>
<p>Then add the predictions to metadata and plot umap.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add in predictions</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>ctrl.sce<span class="sc">@</span>colData<span class="sc">$</span>scmap_cell <span class="ot">&lt;-</span> cell_type_pred</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(ctrl.sce,<span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"scmap_cell"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-18-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Plot both:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>( <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(ctrl.sce,<span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"scmap_cluster"</span>),</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(ctrl.sce,<span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"scmap_cell"</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-19-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_scpred" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="meta-ct_scpred"><span class="header-section-number">5</span> scPred</h2>
<p>scPred will train a classifier based on all principal components. First, <code>getFeatureSpace</code> will create a scPred object stored in the <code>@misc</code> slot where it extracts the PCs that best separates the different celltypes. Then <code>trainModel</code> will do the actual training for each celltype.</p>
<p>scPred works with Seurat objects, so we will convert both objects to seurat objects. You may see a lot of warnings about renaming things, but as long as you do not see an Error, you should be fine.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(Seurat))</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.Seurat</span>(ref.sce)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.Seurat</span>(ctrl.sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The loadings matrix is lost when converted to Seurat object, and scPred needs that information. So we need to rerun PCA with Seurat and the same hvgs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VariableFeatures</span>(reference) <span class="ot">=</span> hvg.ref</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> reference <span class="sc">%&gt;%</span> <span class="fu">ScaleData</span>(<span class="at">verbose=</span>F) <span class="sc">%&gt;%</span> <span class="fu">RunPCA</span>(<span class="at">verbose=</span>F)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="fu">VariableFeatures</span>(ctrl) <span class="ot">=</span> hvg.ctrl</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> ctrl <span class="sc">%&gt;%</span> <span class="fu">ScaleData</span>(<span class="at">verbose=</span>F) <span class="sc">%&gt;%</span> <span class="fu">RunPCA</span>(<span class="at">verbose=</span>F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">getFeatureSpace</span>(reference, <span class="st">"cell_type"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Extracting feature space for each cell type...
DONE!</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">trainModel</span>(reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Training models for each cell type...
DONE!</code></pre>
</div>
</div>
<p>scPred will train a classifier based on all principal components. First, <code>getFeatureSpace</code> will create a scPred object stored in the <code>@misc</code> slot where it extracts the PCs that best separates the different celltypes. Then <code>trainModel</code> will do the actual training for each celltype.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get_scpred</span>(reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'scPred' object
✔  Prediction variable = cell_type 
✔  Discriminant features per cell type
✔  Training model(s)
Summary

|Cell type   |    n| Features|Method    |   ROC|  Sens|  Spec|
|:-----------|----:|--------:|:---------|-----:|-----:|-----:|
|B cell      |  280|       50|svmRadial | 1.000| 1.000| 1.000|
|CD4 T cell  | 1620|       50|svmRadial | 0.994| 0.972| 0.963|
|CD8 T cell  |  945|       50|svmRadial | 0.973| 0.859| 0.971|
|cDC         |   26|       50|svmRadial | 0.994| 0.727| 0.999|
|cMono       |  212|       50|svmRadial | 1.000| 0.957| 0.997|
|ncMono      |   79|       50|svmRadial | 1.000| 0.962| 0.999|
|NK cell     |  312|       50|svmRadial | 0.998| 0.926| 0.995|
|pDC         |   20|       50|svmRadial | 1.000| 0.950| 1.000|
|Plasma cell |    6|       50|svmRadial | 1.000| 1.000| 1.000|</code></pre>
</div>
</div>
<p>You can optimize parameters for each dataset by chaning parameters and testing different types of models, see more at: https://powellgenomicslab.github.io/scPred/articles/introduction.html. But for now, we will continue with this model. Now, lets predict celltypes on our data, where scPred will align the two datasets with Harmony and then perform classification.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">scPredict</span>(ctrl, reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Matching reference with new dataset...
     ─ 1000 features present in reference loadings
     ─ 937 features shared between reference and new dataset
     ─ 93.7% of features in the reference are present in new dataset
●  Aligning new data to reference...
●  Classifying cells...
DONE!</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl, <span class="at">group.by =</span> <span class="st">"scpred_prediction"</span>, <span class="at">label =</span> T, <span class="at">repel =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-25-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Now plot how many cells of each celltypes can be found in each cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>louvain_SNNk15, <span class="at">fill =</span> scpred_prediction)) <span class="sc">+</span> <span class="fu">geom_bar</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-26-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Add the predictions into the SCE object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ctrl.sce<span class="sc">@</span>colData<span class="sc">$</span>scpred_prediction <span class="ot">=</span> ctrl<span class="sc">$</span>scpred_prediction</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="meta-ct_compare" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-ct_compare"><span class="header-section-number">6</span> Compare results</h2>
<p>Now we will compare the output of the two methods using the convenient function in scPred <code>crossTab</code> that prints the overlap between two metadata slots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">crossTab</span>(ctrl, <span class="st">"scmap_cell"</span>, <span class="st">"scpred_prediction"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["B cell"],"name":[1],"type":["int"],"align":["right"]},{"label":["CD4 T cell"],"name":[2],"type":["int"],"align":["right"]},{"label":["CD8 T cell"],"name":[3],"type":["int"],"align":["right"]},{"label":["cDC"],"name":[4],"type":["int"],"align":["right"]},{"label":["cMono"],"name":[5],"type":["int"],"align":["right"]},{"label":["ncMono"],"name":[6],"type":["int"],"align":["right"]},{"label":["NK cell"],"name":[7],"type":["int"],"align":["right"]},{"label":["pDC"],"name":[8],"type":["int"],"align":["right"]},{"label":["Plasma cell"],"name":[9],"type":["int"],"align":["right"]}],"data":[{"1":"99","2":"3","3":"1","4":"0","5":"2","6":"1","7":"1","8":"0","9":"0","_rn_":"B cell"},{"1":"1","2":"114","3":"28","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","_rn_":"CD4 T cell"},{"1":"0","2":"46","3":"194","4":"0","5":"2","6":"0","7":"120","8":"0","9":"1","_rn_":"CD8 T cell"},{"1":"0","2":"0","3":"0","4":"7","5":"2","6":"2","7":"0","8":"0","9":"0","_rn_":"cDC"},{"1":"0","2":"0","3":"0","4":"14","5":"167","6":"75","7":"0","8":"0","9":"0","_rn_":"cMono"},{"1":"0","2":"0","3":"0","4":"0","5":"12","6":"88","7":"0","8":"0","9":"0","_rn_":"ncMono"},{"1":"0","2":"0","3":"22","4":"0","5":"0","6":"0","7":"61","8":"0","9":"0","_rn_":"NK cell"},{"1":"0","2":"0","3":"0","4":"0","5":"1","6":"0","7":"0","8":"0","9":"0","_rn_":"pDC"},{"1":"0","2":"1","3":"1","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","_rn_":"Plasma cell"},{"1":"1","2":"28","3":"20","4":"1","5":"27","6":"23","7":"6","8":"2","9":"1","_rn_":"unassigned"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
<section id="meta-ct_gsea" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="meta-ct_gsea"><span class="header-section-number">7</span> GSEA with celltype markers</h2>
<p>Another option, where celltype can be classified on cluster level is to use gene set enrichment among the DEGs with known markers for different celltypes. Similar to how we did functional enrichment for the DEGs in the Differential expression exercise. There are some resources for celltype gene sets that can be used. Such as <a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/">CellMarker</a>, <a href="https://panglaodb.se/">PanglaoDB</a> or celltype gene sets at <a href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a>. We can also look at overlap between DEGs in a reference dataset and the dataset you are analysing.</p>
<section id="meta-ct_gsea_deg" class="level3" data-number="7.1">
<h3 data-number="7.1" class="anchored" data-anchor-id="meta-ct_gsea_deg"><span class="header-section-number">7.1</span> DEG overlap</h3>
<p>First, lets extract top DEGs for our Covid-19 dataset and the reference dataset. When we run differential expression for our dataset, we want to report as many genes as possible, hence we set the cutoffs quite lenient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run differential expression in our dataset, using clustering at resolution 0.3</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>DGE_list <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>(</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> alldata,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">groups =</span> <span class="fu">as.character</span>(alldata<span class="sc">@</span>colData<span class="sc">$</span>louvain_SNNk15),</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pval.type =</span> <span class="st">"all"</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.prop =</span> <span class="dv">0</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute differential gene expression in reference dataset (that has cell annotation)</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>ref_DGE <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>(</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> ref.sce,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">groups =</span> <span class="fu">as.character</span>(ref.sce<span class="sc">@</span>colData<span class="sc">$</span>cell_type),</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pval.type =</span> <span class="st">"all"</span>,</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">direction =</span> <span class="st">"up"</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the top cell marker genes in reference dataset</span></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co"># select top 50 with hihgest foldchange among top 100 signifcant genes.</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>ref_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(ref_DGE, <span class="cf">function</span>(x){</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>  x<span class="sc">$</span>logFC <span class="ot">&lt;-</span> <span class="fu">rowSums</span>( <span class="fu">as.matrix</span>( x[,<span class="fu">grep</span>(<span class="st">"logFC"</span>,<span class="fu">colnames</span>(x))] ))</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  x <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="fu">filter</span>(p.value <span class="sc">&lt;</span> <span class="fl">0.01</span>) <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="sc">-</span><span class="dv">100</span>, p.value) <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="dv">50</span>, logFC) <span class="sc">%&gt;%</span> <span class="fu">rownames</span>()</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="fu">unlist</span>(<span class="fu">lapply</span>(ref_list, length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     B cell  CD4 T cell  CD8 T cell         cDC       cMono      ncMono 
         50          50          19          17          50          50 
    NK cell         pDC Plasma cell 
         50          50          24 </code></pre>
</div>
</div>
<p>Now we can run GSEA for the DEGs from our dataset and check for enrichment of top DEGs in the reference dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(fgsea))</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co"># run fgsea for each of the clusters in the list</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(DGE_list, <span class="cf">function</span>(x){</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  x<span class="sc">$</span>logFC <span class="ot">&lt;-</span> <span class="fu">rowSums</span>( <span class="fu">as.matrix</span>( x[,<span class="fu">grep</span>(<span class="st">"logFC"</span>,<span class="fu">colnames</span>(x))] ))</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  gene_rank <span class="ot">&lt;-</span> <span class="fu">setNames</span>(x<span class="sc">$</span>logFC, <span class="fu">rownames</span>(x))</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>  fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>( <span class="at">pathways=</span>ref_list, <span class="at">stats=</span>gene_rank,<span class="at">nperm=</span><span class="dv">10000</span>)</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fgseaRes)</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">names</span>(DGE_list)</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a><span class="co"># You can filter and resort the table based on ES, NES or pvalue</span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>pval <span class="sc">&lt;</span> <span class="fl">0.1</span> , ]} )</span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>size <span class="sc">&gt;</span> <span class="dv">2</span> , ]} )</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[<span class="fu">order</span>(x<span class="sc">$</span>NES,<span class="at">decreasing =</span> T), ]} )</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`1`
      pathway         pval         padj         ES       NES nMoreExtreme size
1:     B cell 0.0001927897 0.0004709083  0.9654400  1.996271            0   47
2: CD4 T cell 0.0001913143 0.0004709083  0.8611463  1.797727            0   50
3:        cDC 0.0004054328 0.0007297790  0.9567266  1.693487            1   17
4: CD8 T cell 0.0011836654 0.0015218555 -0.9039497 -1.625338            5   17
5:      cMono 0.0008307373 0.0012461059 -0.7922962 -1.683510            3   47
6:     ncMono 0.0002092926 0.0004709083 -0.8909656 -1.902068            0   49
7:    NK cell 0.0002092926 0.0004709083 -0.9173376 -1.958368            0   49
                                               leadingEdge
1:          CD37,MS4A1,TNFRSF13C,CXCR4,LINC00926,BANK1,...
2:                   RPS29,RPS6,RPL13,RPL32,RPL3,RPL21,...
3: HLA-DRA,HLA-DQB1,HLA-DPB1,HLA-DRB1,HLA-DPA1,HLA-DMA,...
4:                        CCL5,IL32,GZMH,CD3D,CD2,LYAR,...
5:                S100A6,S100A9,LYZ,S100A8,TYROBP,FCN1,...
6:               S100A4,FCER1G,S100A11,AIF1,TIMP1,PSAP,...
7:                     NKG7,HCST,ITGB2,GNLY,MYO1F,CST7,...

$`2`
       pathway         pval         padj         ES       NES nMoreExtreme size
1:       cMono 0.0001488538 0.0005616225  0.9506666  1.883687            0   47
2:      ncMono 0.0004424779 0.0006637168  0.8730863  1.739447            2   49
3: Plasma cell 0.0247007894 0.0317581578 -0.7585071 -1.513369           96   24
4:  CD8 T cell 0.0002416043 0.0005616225 -0.9090058 -1.702997            0   17
5:     NK cell 0.0003103662 0.0005616225 -0.7923197 -1.802458            0   49
6:      B cell 0.0003045067 0.0005616225 -0.8792018 -1.988035            0   47
7:  CD4 T cell 0.0003120125 0.0005616225 -0.9273714 -2.115810            0   50
                                leadingEdge
1:  S100A8,S100A9,LYZ,S100A12,VCAN,FCN1,...
2: S100A11,AIF1,S100A4,SAT1,FCER1G,MAFB,...
3:    PEBP1,ISG20,FKBP11,MIF,CYCS,SPCS2,...
4:         CCL5,IL32,GZMH,CD3D,CD2,LYAR,...
5:       GNLY,NKG7,GZMA,CTSW,B2M,FGFBP2,...
6: RPS5,RPL23A,CXCR4,RPL13A,RPL18A,CD52,...
7:  RPL3,RPS4X,RPL5,EEF1A1,RPS27A,RPL14,...

$`3`
      pathway         pval         padj         ES       NES nMoreExtreme size
1: CD4 T cell 0.0001886081 0.0006382979  0.9809186  2.085454            0   50
2:     B cell 0.0362153963 0.0465626524  0.6874085  1.446835          190   47
3:    NK cell 0.0029787234 0.0044680851 -0.7551115 -1.632298           13   49
4:        cDC 0.0010066438 0.0018119589 -0.9165423 -1.669600            4   17
5:        pDC 0.0008460237 0.0018119589 -0.8104411 -1.741714            3   47
6:      cMono 0.0002115059 0.0006382979 -0.9118854 -1.959727            0   47
7:     ncMono 0.0002127660 0.0006382979 -0.9404478 -2.032933            0   49
                                                leadingEdge
1:                    IL7R,LDHB,PIK3IP1,RPL3,RPS12,RPS6,...
2:                RPS5,RPL13A,RPL23A,RPL18A,RPS23,RPL12,...
3:                     NKG7,GNLY,CST7,GZMA,MYO1F,FGFBP2,...
4: HLA-DRA,HLA-DRB1,HLA-DPA1,HLA-DPB1,HLA-DQB1,HLA-DRB5,...
5:                      PLEK,NPC2,PTPRE,PLAC8,CTSB,IRF8,...
6:                S100A9,S100A8,LYZ,TYROBP,FCN1,S100A12,...
7:                    FCER1G,PSAP,AIF1,IFITM3,LST1,FTH1,...

$`4`
      pathway         pval         padj         ES       NES nMoreExtreme size
1:     ncMono 0.0012733447 0.0022920204 -0.7748661 -1.424156           11   49
2:     B cell 0.0009587728 0.0021572387 -0.7796204 -1.426295            8   47
3:    NK cell 0.0008488964 0.0021572387 -0.7813979 -1.436161            7   49
4:        cDC 0.0040390441 0.0060585661 -0.8695803 -1.450825           35   17
5:      cMono 0.0001065303 0.0004793864 -0.8311165 -1.520506            0   47
6: CD4 T cell 0.0001059322 0.0004793864 -0.9078329 -1.671131            0   50
                                               leadingEdge
1:               S100A4,S100A11,CEBPB,AIF1,IFITM2,SAT1,...
2:                   RPS23,RPS11,CD52,RPL13A,FAU,RPL12,...
3:                   ITGB2,NKG7,MYO1F,GNLY,IFITM1,HCST,...
4: HLA-DRA,HLA-DRB1,HLA-DPB1,HLA-DPA1,HLA-DQB1,HLA-DMA,...
5:                  JUND,S100A6,TYROBP,NFKBIA,LYZ,FCN1,...
6:                RPL34,RPS29,RPL30,RPS14,EEF1A1,RPL36,...

$`5`
       pathway         pval        padj         ES       NES nMoreExtreme size
1:      ncMono 0.0001183992 0.001065593  0.9741230  1.858844            0   49
2:         cDC 0.0012654668 0.002847300  0.9222958  1.570016            8   17
3:       cMono 0.0058542413 0.007526882  0.7928335  1.505493           48   47
4: Plasma cell 0.0105050505 0.011818182 -0.7416214 -1.630275           25   24
5:     NK cell 0.0019280206 0.003470437 -0.6633798 -1.693144            2   49
6:      B cell 0.0030637255 0.004595588 -0.6718164 -1.698590            4   47
7:  CD8 T cell 0.0003460208 0.001557093 -0.9177352 -1.878823            0   17
8:  CD4 T cell 0.0006455778 0.001936733 -0.8405692 -2.153844            0   50
                                               leadingEdge
1:              LST1,AIF1,COTL1,FCER1G,FCGR3A,SERPINA1,...
2: HLA-DPA1,HLA-DRA,HLA-DPB1,HLA-DRB1,HLA-DRB5,HLA-DMA,...
3:                   TYROBP,LYZ,S100A6,FCN1,TKT,S100A9,...
4:                  ISG20,PEBP1,FKBP11,CYCS,JCHAIN,MIF,...
5:                   CST7,IFITM1,GZMM,FGFBP2,CTSW,PRF1,...
6:           CXCR4,RPL13A,RPL23A,MS4A1,TNFRSF13C,BANK1,...
7:                        CCL5,IL32,CD3D,GZMH,CD2,CD8A,...
8:                  RPS3,RPL31,IL7R,LDHB,MGAT4A,RPS27A,...

$`6`
       pathway         pval         padj         ES       NES nMoreExtreme size
1:     NK cell 0.0001600000 0.0004797441  0.9195441  1.907252            0   49
2:  CD4 T cell 0.0001598210 0.0004797441  0.9046969  1.881553            0   50
3:  CD8 T cell 0.0001743071 0.0004797441  0.9658685  1.697378            0   17
4: Plasma cell 0.0518931711 0.0667197914  0.7634732  1.420681          306   24
5:         cDC 0.0304806565 0.0457209848 -0.8134075 -1.511360          129   17
6:      ncMono 0.0002665245 0.0004797441 -0.8961485 -2.002228            0   49
7:       cMono 0.0002642008 0.0004797441 -0.9201712 -2.042908            0   47
                                            leadingEdge
1:                    NKG7,GNLY,CST7,GZMA,CTSW,GZMM,...
2:                  IL7R,RPS29,RPS3,RPL3,RPS6,RPL13,...
3:                    CCL5,IL32,GZMH,CD3D,LYAR,CD8A,...
4:             RPL36AL,FKBP11,PPIB,PEBP1,ISG20,CYCS,...
5: HLA-DRA,HLA-DRB5,HLA-DMA,BASP1,HLA-DQB1,HLA-DRB1,...
6:                 FCER1G,AIF1,LST1,FTH1,COTL1,PSAP,...
7:            S100A9,S100A8,LYZ,TYROBP,FCN1,S100A12,...

$`7`
      pathway         pval         padj         ES       NES nMoreExtreme size
1:      cMono 0.0001076890 0.0004846005  0.9372421  1.664286            0   47
2:     ncMono 0.0001073077 0.0004846005  0.9250568  1.648509            0   49
3:        cDC 0.0078417179 0.0100822088  0.8859262  1.443820           64   17
4:        pDC 0.0844281714 0.0949816929  0.7118244  1.264006          783   47
5: CD4 T cell 0.0073855244 0.0100822088 -0.6240606 -1.610574            4   50
6:     B cell 0.0013966480 0.0026354319 -0.7369408 -1.874564            0   47
7: CD8 T cell 0.0005837712 0.0017513135 -0.9412358 -1.937687            0   17
8:    NK cell 0.0014641288 0.0026354319 -0.8162374 -2.091285            0   49
                                               leadingEdge
1:                LYZ,S100A9,S100A8,FCN1,TYROBP,S100A6,...
2:                 AIF1,PSAP,S100A11,FCER1G,COTL1,LST1,...
3: HLA-DRA,HLA-DRB1,HLA-DPA1,HLA-DQB1,HLA-DPB1,HLA-DMA,...
4:                       NPC2,CTSB,PLEK,PTPRE,CD4,IRF8,...
5:                   IL7R,PIK3IP1,RPS3,RPL3,RPS29,TCF7,...
6:               CXCR4,RPS5,CD79B,MS4A1,TNFRSF13C,CD52,...
7:                        CCL5,IL32,GZMH,CD3D,CD2,CD8A,...
8:                     NKG7,GNLY,CST7,GZMA,CTSW,FGFBP2,...

$`8`
      pathway         pval         padj         ES       NES nMoreExtreme size
1:    NK cell 0.0002020610 0.0006061831  0.9782014  2.105539            0   49
2: CD8 T cell 0.0129659643 0.0166705256  0.8704934  1.575693           63   17
3: CD4 T cell 0.0061252717 0.0091879075 -0.7321603 -1.585427           30   50
4:        cDC 0.0017765495 0.0031977892 -0.8986806 -1.636249            8   17
5:     ncMono 0.0003958045 0.0008905601 -0.8203990 -1.769846            1   49
6:     B cell 0.0001978239 0.0006061831 -0.8415950 -1.804030            0   47
7:      cMono 0.0001978239 0.0006061831 -0.8948426 -1.918170            0   47
                                               leadingEdge
1:                     GNLY,NKG7,PRF1,FGFBP2,CST7,CTSW,...
2:                   CCL5,GZMH,IL32,LYAR,LINC01871,CD2,...
3:                 TMEM123,IL7R,RPS13,RPL22,TPT1,RPS28,...
4: HLA-DRA,HLA-DRB1,HLA-DQB1,HLA-DPA1,HLA-DRB5,HLA-DMA,...
5:                  COTL1,FTH1,AIF1,LST1,SAT1,SERPINA1,...
6:                   CD37,RPS11,RPS23,CD52,MS4A1,CD79B,...
7:                  S100A9,S100A8,LYZ,FCN1,TKT,S100A12,...</code></pre>
</div>
</div>
<p>Selecing top significant overlap per cluster, we can now rename the clusters according to the predicted labels. OBS! Be aware that if you have some clusters that have non-significant p-values for all the gene sets, the cluster label will not be very reliable. Also, the gene sets you are using may not cover all the celltypes you have in your dataset and hence predictions may just be the most similar celltype. Also, some of the clusters have very similar p-values to multiple celltypes, for instance the ncMono and cMono celltypes are equally good for some clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(res,<span class="cf">function</span>(x){<span class="fu">as.data.frame</span>(x)[<span class="dv">1</span>,<span class="dv">1</span>]}))</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>alldata<span class="sc">@</span>colData<span class="sc">$</span>ref_gsea <span class="ot">&lt;-</span> new.cluster.ids[<span class="fu">as.character</span>(alldata<span class="sc">@</span>colData<span class="sc">$</span>louvain_SNNk15)]</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>( <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>   <span class="fu">plotReducedDim</span>(alldata, <span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"louvain_SNNk15"</span>),</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>   <span class="fu">plotReducedDim</span>(alldata, <span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"ref_gsea"</span>)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-32-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>Compare to results with the other celltype prediction methods in the ctrl_13 sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>ctrl.sce<span class="sc">@</span>colData<span class="sc">$</span>ref_gsea <span class="ot">=</span> alldata<span class="sc">@</span>colData<span class="sc">$</span>ref_gsea[alldata<span class="sc">@</span>colData<span class="sc">$</span>sample <span class="sc">==</span> <span class="st">"ctrl.13"</span>]</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>( <span class="at">ncol =</span> <span class="dv">3</span>,</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plotReducedDim</span>(ctrl.sce, <span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"ref_gsea"</span>),</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plotReducedDim</span>(ctrl.sce, <span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"scmap_cell"</span>),          </span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plotReducedDim</span>(ctrl.sce, <span class="at">dimred =</span> <span class="st">"UMAP"</span>,<span class="at">colour_by =</span> <span class="st">"scpred_prediction"</span>)</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-33-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="1344"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_gsea_annot" class="level3" data-number="7.2">
<h3 data-number="7.2" class="anchored" data-anchor-id="meta-ct_gsea_annot"><span class="header-section-number">7.2</span> With annotated gene sets</h3>
<p>We have dowloaded the celltype gene lists from http://bio-bigdata.hrbmu.edu.cn/CellMarker/CellMarker_download.html and converted the excel file to a csv for you. Read in the gene lists and do some filtering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the human marker table</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"data/cellmarker/Human_cell_markers.txt"</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers [ markers<span class="sc">$</span>speciesType <span class="sc">==</span> <span class="st">"Human"</span>, ]</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers [ markers<span class="sc">$</span>cancerType <span class="sc">==</span> <span class="st">"Normal"</span>, ]</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Filter by tissue (to reduce computational time and have tissue-specific classification)</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a><span class="co"># sort(unique(markers$tissueType))</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="co"># grep("blood",unique(markers$tissueType),value = T)</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a><span class="co"># markers &lt;- markers [ markers$tissueType %in% c("Blood","Venous blood",</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                "Serum","Plasma",</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                "Spleen","Bone marrow","Lymph node"), ]</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a><span class="co"># remove strange characters etc.</span></span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>( <span class="fu">unique</span>(markers<span class="sc">$</span>cellName) , <span class="cf">function</span>(x){</span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">paste</span>(markers<span class="sc">$</span>geneSymbol[markers<span class="sc">$</span>cellName <span class="sc">==</span> x],<span class="at">sep=</span><span class="st">","</span>)</span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"[[]|[]]| |-"</span>,<span class="st">","</span>,x)</span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>( x , <span class="at">split =</span> <span class="st">","</span>))</span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">unique</span>(x [ <span class="sc">!</span> x <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">""</span>,<span class="st">"NA"</span>,<span class="st">"family"</span>) ])</span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">casefold</span>(x,<span class="at">upper =</span> T)</span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb44-22"><a href="#cb44-22" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(celltype_list) <span class="ot">&lt;-</span> <span class="fu">unique</span>(markers<span class="sc">$</span>cellName)</span>
<span id="cb44-23"><a href="#cb44-23" aria-hidden="true" tabindex="-1"></a><span class="co"># celltype_list &lt;- lapply(celltype_list , function(x) {x[1:min(length(x),50)]} )</span></span>
<span id="cb44-24"><a href="#cb44-24" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> celltype_list[ <span class="fu">unlist</span>(<span class="fu">lapply</span>(celltype_list,length)) <span class="sc">&lt;</span> <span class="dv">100</span> ]</span>
<span id="cb44-25"><a href="#cb44-25" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> celltype_list[ <span class="fu">unlist</span>(<span class="fu">lapply</span>(celltype_list,length)) <span class="sc">&gt;</span> <span class="dv">5</span> ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run fgsea for each of the clusters in the list</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(DGE_list, <span class="cf">function</span>(x) {</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  x<span class="sc">$</span>logFC <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">as.matrix</span>(x[, <span class="fu">grep</span>(<span class="st">"logFC"</span>, <span class="fu">colnames</span>(x))]))</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  gene_rank <span class="ot">&lt;-</span> <span class="fu">setNames</span>(x<span class="sc">$</span>logFC, <span class="fu">rownames</span>(x))</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>(<span class="at">pathways =</span> celltype_list, <span class="at">stats =</span> gene_rank, <span class="at">nperm =</span> <span class="dv">10000</span>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fgseaRes)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">names</span>(DGE_list)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="co"># You can filter and resort the table based on ES, NES or pvalue</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>pval <span class="sc">&lt;</span> <span class="fl">0.01</span> , ]} )</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>size <span class="sc">&gt;</span> <span class="dv">5</span> , ]} )</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[<span class="fu">order</span>(x<span class="sc">$</span>NES,<span class="at">decreasing =</span> T), ]} )</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a><span class="co"># show top 3 for each cluster.</span></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(res,head,<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`1`
                       pathway        pval       padj         ES       NES
1:           Follicular B cell 0.008166440 0.07598316  0.8576833  1.579356
2:              Pyramidal cell 0.004578025 0.05737792 -0.9698602 -1.492998
3: CD4+CD25+ regulatory T cell 0.001990446 0.03401853 -0.9787623 -1.506702
   nMoreExtreme size                         leadingEdge
1:           41   22 MS4A1,CD69,CD22,FCER2,CD40,PAX5,...
2:           22    6                           CD3E,NRGN
3:            9    6            CD3E,CD3D,CD3G,PTPRC,CD4

$`2`
                  pathway         pval       padj        ES      NES
1:             Neutrophil 0.0001360359 0.01373265 0.9184505 1.936824
2: CD1C+_B dendritic cell 0.0001460920 0.01373265 0.9323468 1.864847
3:           Stromal cell 0.0039203860 0.08544969 0.8450690 1.623956
   nMoreExtreme size                                  leadingEdge
1:            0   80 S100A8,S100A9,S100A12,MNDA,NAMPT,S100A11,...
2:            0   53      S100A8,S100A9,LYZ,S100A12,VCAN,FCN1,...
3:           25   38          VIM,TIMP2,BST1,TIMP1,ANPEP,CD44,...

$`3`
             pathway         pval        padj        ES      NES nMoreExtreme
1: Naive CD8+ T cell 0.0001857010 0.005096508 0.8631756 1.994474            0
2: Naive CD4+ T cell 0.0001926411 0.005096508 0.9346574 1.861827            0
3:       CD4+ T cell 0.0001932367 0.005096508 0.9241719 1.761954            0
   size                           leadingEdge
1:   91 LDHB,PIK3IP1,TCF7,NPM1,NOSIP,RPS8,...
2:   34   IL7R,RPS5,EEF1B2,TCF7,NOSIP,MAL,...
3:   25       IL7R,LTB,CD3E,CD3D,CD3G,CD2,...

$`4`
                             pathway        pval      padj         ES       NES
1:                     Megakaryocyte 0.005434783 0.0984022  0.7665186  1.710242
2: Effector CD8+ memory T (Tem) cell 0.009724809 0.1142665 -0.6875657 -1.309668
3:                        Neutrophil 0.004751575 0.0984022 -0.6971934 -1.329514
   nMoreExtreme size                                leadingEdge
1:            4   26        PF4,PPBP,GP9,ITGA2B,CD9,RASGRP2,...
2:           93   79    AHNAK,ZEB2,EFHD2,LGALS1,ARL4C,MYO1F,...
3:           45   80 ITGB2,PTPRC,S100A11,CD44,IFITM2,S100A9,...

$`5`
            pathway         pval       padj        ES      NES nMoreExtreme
1: Mesenchymal cell 0.0001149293 0.02160671 0.8525202 1.655527            0
2:     Stromal cell 0.0014890185 0.06998387 0.8594653 1.595253           11
3:       Neutrophil 0.0008843688 0.05542044 0.7777406 1.541984            7
   size                               leadingEdge
1:   59       COTL1,S100A4,CTSC,VIM,ZEB2,HES4,...
2:   38     PECAM1,VIM,TIMP1,CD44,ICAM3,PTPRC,...
3:   80 LST1,FCGR3A,S100A11,CD68,PECAM1,ITGB2,...

$`6`
                 pathway         pval      padj        ES      NES nMoreExtreme
1: CD4+ cytotoxic T cell 0.0001527417 0.0106395 0.8725797 1.948876            0
2:   Natural killer cell 0.0001530925 0.0106395 0.7808456 1.739837            0
3:           CD8+ T cell 0.0003395586 0.0106395 0.9518468 1.712349            1
   size                       leadingEdge
1:   86 CCL5,NKG7,GZMH,GNLY,CST7,GZMA,...
2:   84 NKG7,CD3D,GNLY,CD3E,GZMA,CD3G,...
3:   19 NKG7,CD3D,CD3E,CD3G,CD8A,GZMK,...

$`7`
                  pathway         pval       padj        ES      NES
1:             Neutrophil 0.0001028807 0.01000639 0.8778882 1.604758
2: CD1C+_B dendritic cell 0.0001064509 0.01000639 0.8898762 1.593123
3:           Stromal cell 0.0004386927 0.02749141 0.8623636 1.511856
   nMoreExtreme size                                 leadingEdge
1:            0   80 S100A9,S100A8,S100A11,LST1,CD14,S100A12,...
2:            0   53        LYZ,S100A9,S100A8,FCN1,VCAN,CD14,...
3:            3   38         VIM,CD44,TIMP2,TIMP1,ICAM1,BST1,...

$`8`
                             pathway         pval        padj        ES
1:             CD4+ cytotoxic T cell 0.0002023882 0.007654723 0.9422147
2: Effector CD8+ memory T (Tem) cell 0.0002027164 0.007654723 0.8946984
3:               Natural killer cell 0.0002035831 0.007654723 0.8465522
        NES nMoreExtreme size                           leadingEdge
1: 2.165615            0   86   GNLY,NKG7,GZMB,PRF1,FGFBP2,CCL5,...
2: 2.037385            0   79 GNLY,GZMB,FGFBP2,KLRD1,GZMH,SPON2,...
3: 1.939876            0   84   GNLY,NKG7,GZMB,GZMA,CD247,KLRD1,...</code></pre>
</div>
</div>
<p>#CT_GSEA8:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(x)[<span class="dv">1</span>, <span class="dv">1</span>]</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>alldata<span class="sc">@</span>colData<span class="sc">$</span>cellmarker_gsea <span class="ot">&lt;-</span> new.cluster.ids[<span class="fu">as.character</span>(alldata<span class="sc">@</span>colData<span class="sc">$</span>louvain_SNNk15)]</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotReducedDim</span>(alldata, <span class="at">dimred =</span> <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"cellmarker_gsea"</span>),</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotReducedDim</span>(alldata, <span class="at">dimred =</span> <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"ref_gsea"</span>)</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_06_celltyping_files/figure-html/unnamed-chunk-36-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="bioc_06_celltyping_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you think that the methods overlap well? Where do you see the most inconsistencies?</p>
</div>
</div>
<p>In this case we do not have any ground truth, and we cannot say which method performs best. You should keep in mind, that any celltype classification method is just a prediction, and you still need to use your common sense and knowledge of the biological system to judge if the results make sense.</p>
<p>Finally, lets save the data with predictions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(ctrl.sce,<span class="st">"data/results/ctrl13_qc_dr_int_cl_celltype.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="meta-session" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">8</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] fgsea_1.28.0                caret_6.0-94               
 [3] lattice_0.21-8              SeuratObject_4.1.3         
 [5] Seurat_4.3.0                scmap_1.24.0               
 [7] scPred_1.9.2                rafalib_1.0.0              
 [9] pheatmap_1.0.12             cowplot_1.1.1              
[11] dplyr_1.1.2                 scran_1.30.0               
[13] scater_1.30.1               ggplot2_3.4.2              
[15] scuttle_1.12.0              SingleCellExperiment_1.24.0
[17] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[19] GenomicRanges_1.54.1        GenomeInfoDb_1.38.2        
[21] IRanges_2.36.0              S4Vectors_0.40.2           
[23] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[25] matrixStats_1.0.0          

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.0-1     bitops_1.0-7             
  [3] lubridate_1.9.2           httr_1.4.6               
  [5] RColorBrewer_1.1-3        tools_4.3.0              
  [7] sctransform_0.3.5         utf8_1.2.3               
  [9] R6_2.5.1                  lazyeval_0.2.2           
 [11] uwot_0.1.14               withr_2.5.0              
 [13] sp_1.6-1                  gridExtra_2.3            
 [15] progressr_0.13.0          cli_3.6.1                
 [17] spatstat.explore_3.2-1    labeling_0.4.2           
 [19] spatstat.data_3.0-1       randomForest_4.7-1.1     
 [21] proxy_0.4-27              ggridges_0.5.4           
 [23] pbapply_1.7-0             harmony_1.2.0            
 [25] parallelly_1.36.0         limma_3.58.1             
 [27] rstudioapi_0.14           FNN_1.1.3.2              
 [29] generics_0.1.3            ica_1.0-3                
 [31] spatstat.random_3.1-5     Matrix_1.5-4             
 [33] ggbeeswarm_0.7.2          fansi_1.0.4              
 [35] abind_1.4-5               lifecycle_1.0.3          
 [37] yaml_2.3.7                edgeR_4.0.3              
 [39] recipes_1.0.6             SparseArray_1.2.3        
 [41] Rtsne_0.16                grid_4.3.0               
 [43] promises_1.2.0.1          dqrng_0.3.0              
 [45] crayon_1.5.2              miniUI_0.1.1.1           
 [47] beachmat_2.18.0           pillar_1.9.0             
 [49] knitr_1.43                metapod_1.10.1           
 [51] future.apply_1.11.0       codetools_0.2-19         
 [53] fastmatch_1.1-3           leiden_0.4.3             
 [55] googleVis_0.7.1           glue_1.6.2               
 [57] data.table_1.14.8         vctrs_0.6.2              
 [59] png_0.1-8                 gtable_0.3.3             
 [61] kernlab_0.9-32            gower_1.0.1              
 [63] xfun_0.39                 S4Arrays_1.2.0           
 [65] mime_0.12                 prodlim_2023.03.31       
 [67] survival_3.5-5            timeDate_4022.108        
 [69] iterators_1.0.14          hardhat_1.3.0            
 [71] lava_1.7.2.1              statmod_1.5.0            
 [73] bluster_1.12.0            ellipsis_0.3.2           
 [75] fitdistrplus_1.1-11       ROCR_1.0-11              
 [77] ipred_0.9-14              nlme_3.1-162             
 [79] RcppAnnoy_0.0.20          irlba_2.3.5.1            
 [81] vipor_0.4.5               KernSmooth_2.23-20       
 [83] rpart_4.1.19              colorspace_2.1-0         
 [85] nnet_7.3-18               tidyselect_1.2.0         
 [87] compiler_4.3.0            BiocNeighbors_1.20.1     
 [89] DelayedArray_0.28.0       plotly_4.10.2            
 [91] scales_1.2.1              lmtest_0.9-40            
 [93] stringr_1.5.0             digest_0.6.31            
 [95] goftest_1.2-3             spatstat.utils_3.0-3     
 [97] rmarkdown_2.22            RhpcBLASctl_0.23-42      
 [99] XVector_0.42.0            htmltools_0.5.5          
[101] pkgconfig_2.0.3           sparseMatrixStats_1.14.0 
[103] fastmap_1.1.1             rlang_1.1.1              
[105] htmlwidgets_1.6.2         shiny_1.7.4              
[107] DelayedMatrixStats_1.24.0 farver_2.1.1             
[109] zoo_1.8-12                jsonlite_1.8.5           
[111] BiocParallel_1.36.0       ModelMetrics_1.2.2.2     
[113] BiocSingular_1.18.0       RCurl_1.98-1.12          
[115] magrittr_2.0.3            GenomeInfoDbData_1.2.11  
[117] patchwork_1.1.2           munsell_0.5.0            
[119] Rcpp_1.0.10               viridis_0.6.3            
[121] reticulate_1.30           stringi_1.7.12           
[123] pROC_1.18.2               zlibbioc_1.48.0          
[125] MASS_7.3-58.4             plyr_1.8.8               
[127] parallel_4.3.0            listenv_0.9.0            
[129] ggrepel_0.9.3             deldir_1.0-9             
[131] splines_4.3.0             tensor_1.5               
[133] locfit_1.5-9.8            igraph_1.4.3             
[135] spatstat.geom_3.2-1       reshape2_1.4.4           
[137] ScaledMatrix_1.10.0       evaluate_0.21            
[139] foreach_1.5.2             httpuv_1.6.11            
[141] RANN_2.6.1                tidyr_1.3.0              
[143] purrr_1.0.1               polyclip_1.10-4          
[145] future_1.32.0             scattermore_1.2          
[147] rsvd_1.0.5                xtable_1.8-4             
[149] e1071_1.7-13              later_1.3.1              
[151] viridisLite_0.4.2         class_7.3-21             
[153] tibble_3.2.1              beeswarm_0.4.0           
[155] cluster_2.1.4             timechange_0.2.0         
[157] globals_0.16.2           </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2024 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.353</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"loop":true,"closeEffect":"zoom","selector":".lightbox","openEffect":"zoom","descPosition":"bottom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>