<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Combining single-cell gene expression data with spatial information to reveal the spatial distribution of gene activity within tissues.">

<title> Spatial Transcriptomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-microscope" aria-label="microscope"></i> Spatial Transcriptomics</h1>
            <p class="subtitle lead"><i class="fa-brands fa-python" aria-label="python"></i> Scanpy Toolkit</p>
                  <div>
        <div class="description">
          Combining single-cell gene expression data with spatial information to reveal the spatial distribution of gene activity within tissues.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">16-Jan-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#meta-st_prep" id="toc-meta-st_prep" class="nav-link active" data-scroll-target="#meta-st_prep"><span class="header-section-number">1</span> Preparation</a></li>
  <li><a href="#meta-st_qc" id="toc-meta-st_qc" class="nav-link" data-scroll-target="#meta-st_qc"><span class="header-section-number">2</span> Quality control</a>
  <ul>
  <li><a href="#meta-st_qc_filter" id="toc-meta-st_qc_filter" class="nav-link" data-scroll-target="#meta-st_qc_filter"><span class="header-section-number">2.1</span> Filter spots</a></li>
  <li><a href="#meta-st_qc_top" id="toc-meta-st_qc_top" class="nav-link" data-scroll-target="#meta-st_qc_top"><span class="header-section-number">2.2</span> Top expressed genes</a></li>
  <li><a href="#meta-st_qc_filterg" id="toc-meta-st_qc_filterg" class="nav-link" data-scroll-target="#meta-st_qc_filterg"><span class="header-section-number">2.3</span> Filter genes</a></li>
  </ul></li>
  <li><a href="#meta-st_analysis" id="toc-meta-st_analysis" class="nav-link" data-scroll-target="#meta-st_analysis"><span class="header-section-number">3</span> Analysis</a>
  <ul>
  <li><a href="#meta-st_analysis_dimred" id="toc-meta-st_analysis_dimred" class="nav-link" data-scroll-target="#meta-st_analysis_dimred"><span class="header-section-number">3.1</span> Dimensionality reduction and clustering</a></li>
  <li><a href="#meta-st_analysis_int" id="toc-meta-st_analysis_int" class="nav-link" data-scroll-target="#meta-st_analysis_int"><span class="header-section-number">3.2</span> Integration</a></li>
  <li><a href="#meta-st_analysis_svg" id="toc-meta-st_analysis_svg" class="nav-link" data-scroll-target="#meta-st_analysis_svg"><span class="header-section-number">3.3</span> Spatially Variable Features</a></li>
  </ul></li>
  <li><a href="#meta-st_ss" id="toc-meta-st_ss" class="nav-link" data-scroll-target="#meta-st_ss"><span class="header-section-number">4</span> Single cell data</a></li>
  <li><a href="#meta-st_sub" id="toc-meta-st_sub" class="nav-link" data-scroll-target="#meta-st_sub"><span class="header-section-number">5</span> Subset ST for cortex</a></li>
  <li><a href="#meta-st_deconv" id="toc-meta-st_deconv" class="nav-link" data-scroll-target="#meta-st_deconv"><span class="header-section-number">6</span> Deconvolution</a>
  <ul>
  <li><a href="#meta-st_deconv_genes" id="toc-meta-st_deconv_genes" class="nav-link" data-scroll-target="#meta-st_deconv_genes"><span class="header-section-number">6.1</span> Select genes for deconvolution</a></li>
  </ul></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">7</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run Python commands unless it starts with <code>%%bash</code>, in which case, those chunks run shell commands.</p>
</div>
</div>
<p>Adapted from tutorials by Giovanni Palla (https://scanpy-tutorials.readthedocs.io/en/latest/spatial/integration-scanorama.html) and Carlos Talavera-López (https://docs.scvi-tools.org/en/latest/tutorials/notebooks/stereoscope_heart_LV_tutorial.html)</p>
<p>Spatial transcriptomic data with the Visium platform is in many ways similar to scRNAseq data. It contains UMI counts for 5-20 cells instead of single cells, but is still quite sparse in the same way as scRNAseq data is, but with the additional information about spatial location in the tissue.<br>
Here we will first run quality control in a similar manner to scRNAseq data, then QC filtering, dimensionality reduction, integration and clustering. Then we will use scRNAseq data from mouse cortex to run LabelTransfer to predict celltypes in the Visium spots.<br>
We will use two <strong>Visium</strong> spatial transcriptomics dataset of the mouse brain (Sagittal), which are publicly available from the <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/">10x genomics website</a>. Note, that these dataset have already been filtered for spots that does not overlap with the tissue.</p>
<section id="meta-st_prep" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="meta-st_prep"><span class="header-section-number">1</span> Preparation</h2>
<p>Load packages</p>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> an</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanorama</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(action<span class="op">=</span><span class="st">"ignore"</span>, category<span class="op">=</span><span class="pp">Warning</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.logging.print_versions() # gives errror!!</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>sc.set_figure_params(facecolor<span class="op">=</span><span class="st">"white"</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">8</span>))</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load ST data</p>
<p>The function <code>datasets.visium_sge()</code> downloads the dataset from 10x genomics and returns an AnnData object that contains counts, images and spatial coordinates. We will calculate standards QC metrics with <code>pp.calculate_qc_metrics()</code> and visualize them.</p>
<p>When using your own Visium data, use Scanpy’s read_visium() function to import it.</p>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(<span class="st">"./data/spatial/visium"</span>):</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    os.makedirs(<span class="st">"./data/spatial/visium"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>adata_anterior <span class="op">=</span> sc.datasets.visium_sge(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    sample_id<span class="op">=</span><span class="st">"V1_Mouse_Brain_Sagittal_Anterior"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>adata_posterior <span class="op">=</span> sc.datasets.visium_sge(</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    sample_id<span class="op">=</span><span class="st">"V1_Mouse_Brain_Sagittal_Posterior"</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>reading /work/labs/scanpy/data/V1_Mouse_Brain_Sagittal_Anterior/filtered_feature_bc_matrix.h5
 (0:00:00)
reading /work/labs/scanpy/data/V1_Mouse_Brain_Sagittal_Posterior/filtered_feature_bc_matrix.h5
 (0:00:00)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>adata_anterior.var_names_make_unique()</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>adata_posterior.var_names_make_unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To make sure that both images are included in the merged object, use uns_merge=“unique”.</p>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge into one dataset</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>library_names <span class="op">=</span> [<span class="st">"V1_Mouse_Brain_Sagittal_Anterior"</span>, <span class="st">"V1_Mouse_Brain_Sagittal_Posterior"</span>]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata_anterior.concatenate(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    adata_posterior,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    batch_key<span class="op">=</span><span class="st">"library_id"</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    uns_merge<span class="op">=</span><span class="st">"unique"</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    batch_categories<span class="op">=</span>library_names</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>AnnData object with n_obs × n_vars = 6050 × 32285
    obs: 'in_tissue', 'array_row', 'array_col', 'library_id'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'spatial'
    obsm: 'spatial'</code></pre>
</div>
</div>
<p>As you can see, we now have the slot spatial in obsm, which contains the spatial information from the Visium platform.</p>
</section>
<section id="meta-st_qc" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="meta-st_qc"><span class="header-section-number">2</span> Quality control</h2>
<p>Similar to scRNA-seq we use statistics on number of counts, number of features and percent mitochondria for quality control.</p>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add info on mitochondrial and hemoglobin genes to the objects.</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'mt'</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">'mt-'</span>) </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'hb'</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.contains((<span class="st">"^Hb.*-"</span>))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">'mt'</span>,<span class="st">'hb'</span>], percent_top<span class="op">=</span><span class="va">None</span>, log1p<span class="op">=</span><span class="va">False</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'n_genes_by_counts'</span>, <span class="st">'total_counts'</span>, <span class="st">'pct_counts_mt'</span>, <span class="st">'pct_counts_hb'</span>], jitter<span class="op">=</span><span class="fl">0.4</span>, groupby <span class="op">=</span> <span class="st">'library_id'</span>, rotation<span class="op">=</span> <span class="dv">45</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-7-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="scanpy_08_spatial_files/figure-html/cell-7-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>We can also plot the same data onto the tissue section.</p>
<p>In scanpy, this is a bit tricky when you have multiple sections, as you would have to subset and plot them separately.</p>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># need to plot the two sections separately and specify the library_id</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> library <span class="kw">in</span> library_names:</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(adata[adata.obs.library_id <span class="op">==</span> library,:], library_id<span class="op">=</span>library, color <span class="op">=</span> [<span class="st">"total_counts"</span>, <span class="st">"n_genes_by_counts"</span>,<span class="st">'pct_counts_mt'</span>, <span class="st">'pct_counts_hb'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-8-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="scanpy_08_spatial_files/figure-html/cell-8-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-8-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="scanpy_08_spatial_files/figure-html/cell-8-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>As you can see, the spots with low number of counts/features and high mitochondrial content are mainly towards the edges of the tissue. It is quite likely that these regions are damaged tissue. You may also see regions within a tissue with low quality if you have tears or folds in your section.<br>
But remember, for some tissue types, the amount of genes expressed and proportion mitochondria may also be a biological features, so bear in mind what tissue you are working on and what these features mean.</p>
<section id="meta-st_qc_filter" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="meta-st_qc_filter"><span class="header-section-number">2.1</span> Filter spots</h3>
<p>Select all spots with less than <strong>25%</strong> mitocondrial reads, less than <strong>20%</strong> hb-reads and <strong>500</strong> detected genes. You must judge for yourself based on your knowledge of the tissue what are appropriate filtering criteria for your dataset.</p>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>keep <span class="op">=</span> (adata.obs[<span class="st">'pct_counts_hb'</span>] <span class="op">&lt;</span> <span class="dv">20</span>) <span class="op">&amp;</span> (adata.obs[<span class="st">'pct_counts_mt'</span>] <span class="op">&lt;</span> <span class="dv">25</span>) <span class="op">&amp;</span> (adata.obs[<span class="st">'n_genes_by_counts'</span>] <span class="op">&gt;</span> <span class="dv">1000</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">sum</span>(keep))</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[keep,:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>5749</code></pre>
</div>
</div>
<p>And replot onto tissue sections.</p>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> library <span class="kw">in</span> library_names:</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(adata[adata.obs.library_id <span class="op">==</span> library,:], library_id<span class="op">=</span>library, color <span class="op">=</span> [<span class="st">"total_counts"</span>, <span class="st">"n_genes_by_counts"</span>,<span class="st">'pct_counts_mt'</span>, <span class="st">'pct_counts_hb'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-10-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="scanpy_08_spatial_files/figure-html/cell-10-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-10-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="scanpy_08_spatial_files/figure-html/cell-10-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-st_qc_top" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="meta-st_qc_top"><span class="header-section-number">2.2</span> Top expressed genes</h3>
<p>As for scRNA-seq data, we will look at what the top expressed genes are.</p>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>sc.pl.highest_expr_genes(adata, n_top<span class="op">=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>normalizing counts per cell
    finished (0:00:00)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-11-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="scanpy_08_spatial_files/figure-html/cell-11-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>As you can see, the mitochondrial genes are among the top expressed genes. Also the lncRNA gene Bc1 (brain cytoplasmic RNA 1). Also one hemoglobin gene.</p>
</section>
<section id="meta-st_qc_filterg" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="meta-st_qc_filterg"><span class="header-section-number">2.3</span> Filter genes</h3>
<p>We will remove the <em>Bc1</em> gene, hemoglobin genes (blood contamination) and the mitochondrial genes.</p>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">'mt-'</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>hb_genes <span class="op">=</span> adata.var_names.<span class="bu">str</span>.contains(<span class="st">'^Hb.*-'</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>remove <span class="op">=</span> np.add(mito_genes, hb_genes)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>remove[adata.var_names <span class="op">==</span> <span class="st">"Bc1"</span>] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>keep <span class="op">=</span> np.invert(remove)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">sum</span>(remove))</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[:,keep]</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(adata.n_obs, adata.n_vars)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>22
5749 32263</code></pre>
</div>
</div>
</section>
</section>
<section id="meta-st_analysis" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="meta-st_analysis"><span class="header-section-number">3</span> Analysis</h2>
<p>We will proceed with the data in a very similar manner to scRNA-seq data.</p>
<p>As we have two sections, we will select variable genes with batch_key=“library_id” and then take the union of variable genes for further analysis. The idea is to avoid including batch specific genes in the analysis.</p>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save the counts to a separate object for later, we need the normalized counts in raw for DEG dete</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>counts_adata <span class="op">=</span> adata.copy()</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co"># take 1500 variable genes per batch and then use the union of them.</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata, flavor<span class="op">=</span><span class="st">"seurat"</span>, n_top_genes<span class="op">=</span><span class="dv">1500</span>, inplace<span class="op">=</span><span class="va">True</span>, batch_key<span class="op">=</span><span class="st">"library_id"</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co"># subset for variable genes</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> adata</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[:,adata.var.highly_variable_nbatches <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co"># scale data</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>sc.pp.scale(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>normalizing counts per cell
    finished (0:00:00)
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:02)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
... as `zero_center=True`, sparse input is densified and may lead to large memory consumption</code></pre>
</div>
</div>
<p>Now we can plot gene expression of individual genes, the gene <em>Hpca</em> is a strong hippocampal marker and <em>Ttr</em> is a marker of the choroid plexus.</p>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> library <span class="kw">in</span> library_names:</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(adata[adata.obs.library_id <span class="op">==</span> library,:], library_id<span class="op">=</span>library, color <span class="op">=</span> [<span class="st">"Ttr"</span>, <span class="st">"Hpca"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-14-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="scanpy_08_spatial_files/figure-html/cell-14-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-14-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="scanpy_08_spatial_files/figure-html/cell-14-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<section id="meta-st_analysis_dimred" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="meta-st_analysis_dimred"><span class="header-section-number">3.1</span> Dimensionality reduction and clustering</h3>
<p>We can then now run dimensionality reduction and clustering using the same workflow as we use for scRNA-seq analysis.</p>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, key_added<span class="op">=</span><span class="st">"clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing neighbors
WARNING: You’re trying to run this on 2405 dimensions of `.X`, if you really want this, set `use_rep='X'`.
         Falling back to preprocessing with `sc.pp.pca` and default params.
computing PCA
    with n_comps=50
    finished (0:00:00)
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:08)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:09)
running Leiden clustering
    finished: found 23 clusters and added
    'clusters', the cluster labels (adata.obs, categorical) (0:00:01)</code></pre>
</div>
</div>
<p>We can then plot clusters onto umap or onto the tissue section.</p>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    adata, color<span class="op">=</span>[<span class="st">"clusters"</span>, <span class="st">"library_id"</span>], palette<span class="op">=</span>sc.pl.palettes.default_20</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Length of palette colors is smaller than the number of categories (palette length: 20, categories length: 23. Some categories will have the same color.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-16-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="scanpy_08_spatial_files/figure-html/cell-16-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>As we are plotting the two sections separately, we need to make sure that they get the same colors by fetching cluster colors from a dict.</p>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>clusters_colors <span class="op">=</span> <span class="bu">dict</span>(</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">zip</span>([<span class="bu">str</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(adata.obs.clusters.cat.categories))], adata.uns[<span class="st">"clusters_colors"</span>])</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">10</span>))</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, library <span class="kw">in</span> <span class="bu">enumerate</span>(</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"V1_Mouse_Brain_Sagittal_Anterior"</span>, <span class="st">"V1_Mouse_Brain_Sagittal_Posterior"</span>]</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    ad <span class="op">=</span> adata[adata.obs.library_id <span class="op">==</span> library, :].copy()</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>        ad,</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>        img_key<span class="op">=</span><span class="st">"hires"</span>,</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>        library_id<span class="op">=</span>library,</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>        color<span class="op">=</span><span class="st">"clusters"</span>,</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>        size<span class="op">=</span><span class="fl">1.5</span>,</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>        palette<span class="op">=</span>[</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>            v</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> k, v <span class="kw">in</span> clusters_colors.items()</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> k <span class="kw">in</span> ad.obs.clusters.unique().tolist()</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>        ],</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>        legend_loc<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>        show<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>        ax<span class="op">=</span>axs[i],</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-17-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="scanpy_08_spatial_files/figure-html/cell-17-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-st_analysis_int" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="meta-st_analysis_int"><span class="header-section-number">3.2</span> Integration</h3>
<p>Quite often there are strong batch effects between different ST sections, so it may be a good idea to integrate the data across sections.</p>
<p>We will do a similar integration as in the Data Integration lab, here we will use Scanorama for integration.</p>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>adatas <span class="op">=</span> {}</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> batch <span class="kw">in</span> library_names:</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    adatas[batch] <span class="op">=</span> adata[adata.obs[<span class="st">'library_id'</span>] <span class="op">==</span> batch,]</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>adatas </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<pre><code>{'V1_Mouse_Brain_Sagittal_Anterior': View of AnnData object with n_obs × n_vars = 2597 × 2405
     obs: 'in_tissue', 'array_row', 'array_col', 'library_id', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_hb', 'pct_counts_hb', 'clusters'
     var: 'gene_ids', 'feature_types', 'genome', 'mt', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'mean', 'std'
     uns: 'spatial', 'library_id_colors', 'log1p', 'hvg', 'neighbors', 'umap', 'leiden', 'clusters_colors'
     obsm: 'spatial', 'X_pca', 'X_umap'
     obsp: 'distances', 'connectivities',
 'V1_Mouse_Brain_Sagittal_Posterior': View of AnnData object with n_obs × n_vars = 3152 × 2405
     obs: 'in_tissue', 'array_row', 'array_col', 'library_id', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_hb', 'pct_counts_hb', 'clusters'
     var: 'gene_ids', 'feature_types', 'genome', 'mt', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'mean', 'std'
     uns: 'spatial', 'library_id_colors', 'log1p', 'hvg', 'neighbors', 'umap', 'leiden', 'clusters_colors'
     obsm: 'spatial', 'X_pca', 'X_umap'
     obsp: 'distances', 'connectivities'}</code></pre>
</div>
</div>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanorama</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co">#convert to list of AnnData objects</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>adatas <span class="op">=</span> <span class="bu">list</span>(adatas.values())</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># run scanorama.integrate</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>scanorama.integrate_scanpy(adatas, dimred <span class="op">=</span> <span class="dv">50</span>)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all the integrated matrices.</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>scanorama_int <span class="op">=</span> [ad.obsm[<span class="st">'X_scanorama'</span>] <span class="cf">for</span> ad <span class="kw">in</span> adatas]</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="co"># make into one matrix.</span></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>all_s <span class="op">=</span> np.concatenate(scanorama_int)</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(all_s.shape)</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="co"># add to the AnnData object</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">"Scanorama"</span>] <span class="op">=</span> all_s</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Found 2405 genes among all datasets
[[0.         0.47824413]
 [0.         0.        ]]
Processing datasets (0, 1)
(5749, 50)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="18">
<pre><code>AnnData object with n_obs × n_vars = 5749 × 2405
    obs: 'in_tissue', 'array_row', 'array_col', 'library_id', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_hb', 'pct_counts_hb', 'clusters'
    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'mean', 'std'
    uns: 'spatial', 'library_id_colors', 'log1p', 'hvg', 'neighbors', 'umap', 'leiden', 'clusters_colors'
    obsm: 'spatial', 'X_pca', 'X_umap', 'Scanorama'
    obsp: 'distances', 'connectivities'</code></pre>
</div>
</div>
<p>Then we run dimensionality reduction and clustering as before.</p>
<div class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, use_rep<span class="op">=</span><span class="st">"Scanorama"</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, key_added<span class="op">=</span><span class="st">"clusters"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    adata, color<span class="op">=</span>[<span class="st">"clusters"</span>, <span class="st">"library_id"</span>], palette<span class="op">=</span>sc.pl.palettes.default_20</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:00)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:09)
running Leiden clustering
    finished: found 19 clusters and added
    'clusters', the cluster labels (adata.obs, categorical) (0:00:01)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-20-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="scanpy_08_spatial_files/figure-html/cell-20-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>As we have new clusters, we again need to make a new dict for cluster colors</p>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>clusters_colors <span class="op">=</span> <span class="bu">dict</span>(</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">zip</span>([<span class="bu">str</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(adata.obs.clusters.cat.categories))], adata.uns[<span class="st">"clusters_colors"</span>])</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">10</span>))</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, library <span class="kw">in</span> <span class="bu">enumerate</span>(</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"V1_Mouse_Brain_Sagittal_Anterior"</span>, <span class="st">"V1_Mouse_Brain_Sagittal_Posterior"</span>]</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    ad <span class="op">=</span> adata[adata.obs.library_id <span class="op">==</span> library, :].copy()</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>        ad,</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>        img_key<span class="op">=</span><span class="st">"hires"</span>,</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>        library_id<span class="op">=</span>library,</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>        color<span class="op">=</span><span class="st">"clusters"</span>,</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>        size<span class="op">=</span><span class="fl">1.5</span>,</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>        palette<span class="op">=</span>[</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>            v</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> k, v <span class="kw">in</span> clusters_colors.items()</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> k <span class="kw">in</span> ad.obs.clusters.unique().tolist()</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>        ],</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>        legend_loc<span class="op">=</span><span class="st">"on data"</span>,</span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>        show<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>        ax<span class="op">=</span>axs[i],</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-21-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="scanpy_08_spatial_files/figure-html/cell-21-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you see any differences between the integrated and non-integrated clustering? Judge for yourself, which of the clusterings do you think looks best? As a reference, you can compare to brain regions in the <a href="https://mouse.brain-map.org/experiment/thumbnails/100042147?image_type=atlas">Allen brain atlas</a>.</p>
</div>
</div>
</section>
<section id="meta-st_analysis_svg" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="meta-st_analysis_svg"><span class="header-section-number">3.3</span> Spatially Variable Features</h3>
<p>There are two main workflows to identify molecular features that correlate with spatial location within a tissue. The first is to perform differential expression based on spatially distinct clusters, the other is to find features that have spatial patterning without taking clusters or spatial annotation into account. First, we will do differential expression between clusters just as we did for the scRNAseq data before.</p>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run t-test </span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, <span class="st">"clusters"</span>, method<span class="op">=</span><span class="st">"wilcoxon"</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot as heatmap for cluster5 genes</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_heatmap(adata, groups<span class="op">=</span><span class="st">"5"</span>, n_genes<span class="op">=</span><span class="dv">10</span>, groupby<span class="op">=</span><span class="st">"clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ranking genes
    finished: added to `.uns['rank_genes_groups']`
    'names', sorted np.recarray to be indexed by group ids
    'scores', sorted np.recarray to be indexed by group ids
    'logfoldchanges', sorted np.recarray to be indexed by group ids
    'pvals', sorted np.recarray to be indexed by group ids
    'pvals_adj', sorted np.recarray to be indexed by group ids (0:00:15)
WARNING: dendrogram data not found (using key=dendrogram_clusters). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.
    using 'X_pca' with n_pcs = 50
Storing dendrogram info using `.uns['dendrogram_clusters']`
WARNING: Groups are not reordered because the `groupby` categories and the `var_group_labels` are different.
categories: 0, 1, 2, etc.
var_group_labels: 5</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-22-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="scanpy_08_spatial_files/figure-html/cell-22-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot onto spatial location</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>top_genes <span class="op">=</span> sc.get.rank_genes_groups_df(adata, group<span class="op">=</span><span class="st">'5'</span>,log2fc_min<span class="op">=</span><span class="dv">0</span>)[<span class="st">'names'</span>][:<span class="dv">3</span>]</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> library <span class="kw">in</span> [<span class="st">"V1_Mouse_Brain_Sagittal_Anterior"</span>, <span class="st">"V1_Mouse_Brain_Sagittal_Posterior"</span>]:</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(adata[adata.obs.library_id <span class="op">==</span> library,:], library_id<span class="op">=</span>library, color <span class="op">=</span> top_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-23-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="scanpy_08_spatial_files/figure-html/cell-23-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-23-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="scanpy_08_spatial_files/figure-html/cell-23-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Spatial transcriptomics allows researchers to investigate how gene expression trends varies in space, thus identifying spatial patterns of gene expression. For this purpose there are multiple methods, such as SpatailDE, SPARK, Trendsceek, HMRF and Splotch.</p>
<p>We use SpatialDE Svensson et al., a Gaussian process-based statistical framework that aims to identify spatially variable genes.</p>
<div class="{callout-warning}">
<p>Takes a long time to run, so skip this step for now and download the precomputed file.</p>
</div>
<div class="cell" data-results="hide" data-execution_count="23">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># slow step</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> NaiveDE</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> SpatialDE</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>counts <span class="op">=</span> sc.get.obs_df(adata, keys<span class="op">=</span><span class="bu">list</span>(adata.var_names), use_raw<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>total_counts <span class="op">=</span> sc.get.obs_df(adata, keys<span class="op">=</span>[<span class="st">"total_counts"</span>])</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>norm_expr <span class="op">=</span> NaiveDE.stabilize(counts.T).T</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>resid_expr <span class="op">=</span> NaiveDE.regress_out(</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>    total_counts, norm_expr.T, <span class="st">"np.log(total_counts)"</span>).T</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>results <span class="op">=</span> SpatialDE.run(adata.obsm[<span class="st">"spatial"</span>], resid_expr)</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'data/spatial/visium/scanpy_spatialde.pkl'</span>, <span class="st">'wb'</span>) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>    pickle.dump(results, <span class="bu">file</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> urllib.request</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>path_data <span class="op">=</span> <span class="st">"https://export.uppmax.uu.se/naiss2023-23-3/workshops/workshop-scrnaseq"</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>path_file <span class="op">=</span> <span class="st">"data/spatial/visium/scanpy_spatialde.pkl"</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(path_file):</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>    file_url <span class="op">=</span> os.path.join(</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>        path_data, <span class="st">"spatial/visium/results/scanpy_spatialde.pkl"</span>)</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>    urllib.request.urlretrieve(file_url, path_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'data/spatial/visium/scanpy_spatialde.pkl'</span>, <span class="st">'rb'</span>) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    results <span class="op">=</span> pickle.load(<span class="bu">file</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We concatenate the results with the DataFrame of annotations of variables: `adata.var`.</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>results.index <span class="op">=</span> results[<span class="st">"g"</span>]</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>adata.var <span class="op">=</span> pd.concat(</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>    [adata.var, results.loc[adata.var.index.values, :]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(<span class="st">'./data/spatial/visium/adata_processed_sc.h5ad'</span>)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Then we can inspect significant genes that varies in space and visualize them with `sc.pl.spatial` function.</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>results.sort_values(<span class="st">"qval"</span>).head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">FSV</th>
<th data-quarto-table-cell-role="th">M</th>
<th data-quarto-table-cell-role="th">g</th>
<th data-quarto-table-cell-role="th">l</th>
<th data-quarto-table-cell-role="th">max_delta</th>
<th data-quarto-table-cell-role="th">max_ll</th>
<th data-quarto-table-cell-role="th">max_mu_hat</th>
<th data-quarto-table-cell-role="th">max_s2_t_hat</th>
<th data-quarto-table-cell-role="th">model</th>
<th data-quarto-table-cell-role="th">n</th>
<th data-quarto-table-cell-role="th">s2_FSV</th>
<th data-quarto-table-cell-role="th">s2_logdelta</th>
<th data-quarto-table-cell-role="th">time</th>
<th data-quarto-table-cell-role="th">BIC</th>
<th data-quarto-table-cell-role="th">max_ll_null</th>
<th data-quarto-table-cell-role="th">LLR</th>
<th data-quarto-table-cell-role="th">pval</th>
<th data-quarto-table-cell-role="th">qval</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">g</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Efnb3</td>
<td>1.775241e-01</td>
<td>4</td>
<td>Efnb3</td>
<td>544.733658</td>
<td>4.490546e+00</td>
<td>-4701.690768</td>
<td>0.920436</td>
<td>6.950994e-02</td>
<td>SE</td>
<td>5749</td>
<td>8.838199e-06</td>
<td>4.029374e-04</td>
<td>0.002817</td>
<td>9438.008662</td>
<td>-5374.163817</td>
<td>672.473049</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">S100a16</td>
<td>5.205988e-02</td>
<td>4</td>
<td>S100a16</td>
<td>544.733658</td>
<td>1.764863e+01</td>
<td>-4013.533873</td>
<td>1.703474</td>
<td>3.034361e-02</td>
<td>SE</td>
<td>5749</td>
<td>7.639943e-06</td>
<td>2.521451e-03</td>
<td>0.002874</td>
<td>8061.694871</td>
<td>-4099.970498</td>
<td>86.436625</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">S100a5</td>
<td>3.082312e-01</td>
<td>4</td>
<td>S100a5</td>
<td>544.733658</td>
<td>2.175292e+00</td>
<td>-989.712703</td>
<td>-0.692740</td>
<td>3.893413e-02</td>
<td>SE</td>
<td>5749</td>
<td>1.244686e-05</td>
<td>3.006748e-04</td>
<td>0.006836</td>
<td>2014.052530</td>
<td>-2128.916543</td>
<td>1139.203841</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">S100a6</td>
<td>1.198049e-01</td>
<td>4</td>
<td>S100a6</td>
<td>544.733658</td>
<td>7.120948e+00</td>
<td>-4911.277757</td>
<td>0.024394</td>
<td>4.364682e-02</td>
<td>SE</td>
<td>5749</td>
<td>1.525869e-05</td>
<td>1.234171e-03</td>
<td>0.005581</td>
<td>9857.182640</td>
<td>-5087.893957</td>
<td>176.616199</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Cers2</td>
<td>6.170160e-02</td>
<td>4</td>
<td>Cers2</td>
<td>544.733658</td>
<td>1.473933e+01</td>
<td>-4909.254062</td>
<td>1.688963</td>
<td>3.873848e-02</td>
<td>SE</td>
<td>5749</td>
<td>6.963833e-06</td>
<td>1.699495e-03</td>
<td>0.002615</td>
<td>9853.135249</td>
<td>-5025.230150</td>
<td>115.976088</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Car14</td>
<td>6.721518e-02</td>
<td>4</td>
<td>Car14</td>
<td>544.733658</td>
<td>1.345078e+01</td>
<td>-4078.211327</td>
<td>1.646634</td>
<td>3.422252e-02</td>
<td>SE</td>
<td>5749</td>
<td>5.827188e-06</td>
<td>1.224504e-03</td>
<td>0.002738</td>
<td>8191.049780</td>
<td>-4232.446324</td>
<td>154.234997</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Hmgcs2</td>
<td>1.043706e-01</td>
<td>4</td>
<td>Hmgcs2</td>
<td>544.733658</td>
<td>8.317319e+00</td>
<td>64.273783</td>
<td>-0.697246</td>
<td>9.799863e-03</td>
<td>SE</td>
<td>5749</td>
<td>1.273224e-05</td>
<td>1.280151e-03</td>
<td>0.002757</td>
<td>-93.920441</td>
<td>-106.666922</td>
<td>170.940704</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Atp1a1</td>
<td>1.951715e-01</td>
<td>4</td>
<td>Atp1a1</td>
<td>544.733658</td>
<td>3.996873e+00</td>
<td>-2655.884269</td>
<td>-1.945325</td>
<td>6.109633e-02</td>
<td>SE</td>
<td>5749</td>
<td>1.387274e-05</td>
<td>5.578607e-04</td>
<td>0.002461</td>
<td>5346.395662</td>
<td>-3228.572142</td>
<td>572.687873</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Vangl1</td>
<td>1.997762e-09</td>
<td>4</td>
<td>Vangl1</td>
<td>544.733658</td>
<td>4.851652e+08</td>
<td>523.655680</td>
<td>-0.633082</td>
<td>9.297993e-10</td>
<td>SE</td>
<td>5749</td>
<td>5.712059e-09</td>
<td>1.036289e+09</td>
<td>0.011918</td>
<td>-1012.684235</td>
<td>435.773794</td>
<td>87.881886</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Tspan2</td>
<td>2.008700e-01</td>
<td>4</td>
<td>Tspan2</td>
<td>544.733658</td>
<td>3.855987e+00</td>
<td>-4904.385928</td>
<td>2.862779</td>
<td>1.358988e-01</td>
<td>SE</td>
<td>5749</td>
<td>1.765625e-05</td>
<td>6.842354e-04</td>
<td>0.002182</td>
<td>9843.398981</td>
<td>-5358.727526</td>
<td>454.341598</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
</section>
<section id="meta-st_ss" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="meta-st_ss"><span class="header-section-number">4</span> Single cell data</h2>
<p>We can use a scRNA-seq dataset as a reference to predict the proportion of different celltypes in the Visium spots. Keep in mind that it is important to have a reference that contains all the celltypes you expect to find in your spots. Ideally it should be a scRNA-seq reference from the exact same tissue. We will use a reference scRNA-seq dataset of ~14,000 adult mouse cortical cell taxonomy from the Allen Institute, generated with the SMART-Seq2 protocol.</p>
<p>Conveniently, you can also download the pre-processed dataset in h5ad format from here. Here with bash code:</p>
<div class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> urllib.request</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>path_data <span class="op">=</span> <span class="st">"https://export.uppmax.uu.se/naiss2023-23-3/workshops/workshop-scrnaseq"</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>path_file <span class="op">=</span> <span class="st">"data/spatial/visium/allen_cortex.h5ad"</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(path_file):</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>    file_url <span class="op">=</span> os.path.join(</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>        path_data, <span class="st">"spatial/visium/allen_cortex.h5ad"</span>)</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>    urllib.request.urlretrieve(file_url, path_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>adata_cortex <span class="op">=</span> sc.read_h5ad(<span class="st">"data/spatial/visium/allen_cortex.h5ad"</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>adata_cortex</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>AnnData object with n_obs × n_vars = 14249 × 34617
    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'sample_id', 'sample_type', 'organism', 'donor', 'sex', 'age_days', 'eye_condition', 'genotype', 'driver_lines', 'reporter_lines', 'brain_hemisphere', 'brain_region', 'brain_subregion', 'injection_label_direction', 'injection_primary', 'injection_secondary', 'injection_tract', 'injection_material', 'injection_exclusion_criterion', 'facs_date', 'facs_container', 'facs_sort_criteria', 'rna_amplification_set', 'library_prep_set', 'library_prep_avg_size_bp', 'seq_name', 'seq_tube', 'seq_batch', 'total_reads', 'percent_exon_reads', 'percent_intron_reads', 'percent_intergenic_reads', 'percent_rrna_reads', 'percent_mt_exon_reads', 'percent_reads_unique', 'percent_synth_reads', 'percent_ecoli_reads', 'percent_aligned_reads_total', 'complexity_cg', 'genes_detected_cpm_criterion', 'genes_detected_fpkm_criterion', 'tdt_cpm', 'gfp_cpm', 'class', 'subclass', 'cluster', 'confusion_score', 'cluster_correlation', 'core_intermediate_call'
    var: 'features'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>adata_cortex.obs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="28">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">orig.ident</th>
<th data-quarto-table-cell-role="th">nCount_RNA</th>
<th data-quarto-table-cell-role="th">nFeature_RNA</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">sample_type</th>
<th data-quarto-table-cell-role="th">organism</th>
<th data-quarto-table-cell-role="th">donor</th>
<th data-quarto-table-cell-role="th">sex</th>
<th data-quarto-table-cell-role="th">age_days</th>
<th data-quarto-table-cell-role="th">eye_condition</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">genes_detected_cpm_criterion</th>
<th data-quarto-table-cell-role="th">genes_detected_fpkm_criterion</th>
<th data-quarto-table-cell-role="th">tdt_cpm</th>
<th data-quarto-table-cell-role="th">gfp_cpm</th>
<th data-quarto-table-cell-role="th">class</th>
<th data-quarto-table-cell-role="th">subclass</th>
<th data-quarto-table-cell-role="th">cluster</th>
<th data-quarto-table-cell-role="th">confusion_score</th>
<th data-quarto-table-cell-role="th">cluster_correlation</th>
<th data-quarto-table-cell-role="th">core_intermediate_call</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">F1S4_160108_001_A01</td>
<td>0</td>
<td>1730700.0</td>
<td>9029</td>
<td>527128530</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>225675</td>
<td>M</td>
<td>53</td>
<td>Normal</td>
<td>...</td>
<td>10445</td>
<td>9222</td>
<td>248.86</td>
<td>248.86</td>
<td>GABAergic</td>
<td>Vip</td>
<td>Vip Arhgap36 Hmcn1</td>
<td>0.4385</td>
<td>0.837229</td>
<td>Intermediate</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">F1S4_160108_001_B01</td>
<td>0</td>
<td>1909605.0</td>
<td>10207</td>
<td>527128536</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>225675</td>
<td>M</td>
<td>53</td>
<td>Normal</td>
<td>...</td>
<td>11600</td>
<td>10370</td>
<td>289.61</td>
<td>289.61</td>
<td>GABAergic</td>
<td>Lamp5</td>
<td>Lamp5 Lsp1</td>
<td>0.1025</td>
<td>0.878743</td>
<td>Core</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">F1S4_160108_001_C01</td>
<td>0</td>
<td>1984948.0</td>
<td>10578</td>
<td>527128542</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>225675</td>
<td>M</td>
<td>53</td>
<td>Normal</td>
<td>...</td>
<td>11848</td>
<td>10734</td>
<td>281.06</td>
<td>281.06</td>
<td>GABAergic</td>
<td>Lamp5</td>
<td>Lamp5 Lsp1</td>
<td>0.0195</td>
<td>0.887084</td>
<td>Core</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">F1S4_160108_001_D01</td>
<td>0</td>
<td>2291552.0</td>
<td>8482</td>
<td>527128548</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>225675</td>
<td>M</td>
<td>53</td>
<td>Normal</td>
<td>...</td>
<td>9494</td>
<td>8561</td>
<td>390.02</td>
<td>390.02</td>
<td>GABAergic</td>
<td>Vip</td>
<td>Vip Crispld2 Htr2c</td>
<td>0.2734</td>
<td>0.843552</td>
<td>Core</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">F1S4_160108_001_E01</td>
<td>0</td>
<td>1757463.0</td>
<td>8697</td>
<td>527128554</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>225675</td>
<td>M</td>
<td>53</td>
<td>Normal</td>
<td>...</td>
<td>10012</td>
<td>8791</td>
<td>253.92</td>
<td>253.92</td>
<td>GABAergic</td>
<td>Lamp5</td>
<td>Lamp5 Plch2 Dock5</td>
<td>0.7532</td>
<td>0.854994</td>
<td>Core</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">FYS4_171004_104_C01</td>
<td>2</td>
<td>949356.0</td>
<td>9141</td>
<td>645142562</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>350650</td>
<td>M</td>
<td>51</td>
<td>Normal</td>
<td>...</td>
<td>9629</td>
<td>9229</td>
<td>432.15</td>
<td>432.15</td>
<td>Glutamatergic</td>
<td>L5 PT</td>
<td>L5 PT VISp C1ql2 Cdh13</td>
<td>0.0477</td>
<td>0.885255</td>
<td>Core</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FYS4_171004_104_D01</td>
<td>2</td>
<td>998736.0</td>
<td>6927</td>
<td>645142573</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>350650</td>
<td>M</td>
<td>51</td>
<td>Normal</td>
<td>...</td>
<td>7701</td>
<td>7023</td>
<td>217.83</td>
<td>217.83</td>
<td>GABAergic</td>
<td>Sst</td>
<td>Sst Hpse Sema3c</td>
<td>0.1064</td>
<td>0.854499</td>
<td>Core</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">FYS4_171004_104_F01</td>
<td>2</td>
<td>1002766.0</td>
<td>6936</td>
<td>645142613</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>350650</td>
<td>M</td>
<td>51</td>
<td>Normal</td>
<td>...</td>
<td>7888</td>
<td>7054</td>
<td>91.88</td>
<td>91.88</td>
<td>Glutamatergic</td>
<td>L5 PT</td>
<td>L5 PT VISp Chrna6</td>
<td>0.0095</td>
<td>0.822625</td>
<td>Core</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FYS4_171004_104_G01</td>
<td>2</td>
<td>1025804.0</td>
<td>8027</td>
<td>645142648</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>350650</td>
<td>M</td>
<td>51</td>
<td>Normal</td>
<td>...</td>
<td>8933</td>
<td>8146</td>
<td>127.77</td>
<td>127.77</td>
<td>GABAergic</td>
<td>Sst</td>
<td>Sst Calb2 Pdlim5</td>
<td>0.2852</td>
<td>0.856322</td>
<td>Core</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">FYS4_171004_104_H01</td>
<td>2</td>
<td>882435.0</td>
<td>6574</td>
<td>645142673</td>
<td>Cells</td>
<td>Mus musculus</td>
<td>350650</td>
<td>M</td>
<td>51</td>
<td>Normal</td>
<td>...</td>
<td>7393</td>
<td>6687</td>
<td>310.17</td>
<td>310.17</td>
<td>GABAergic</td>
<td>Pvalb</td>
<td>Pvalb Reln Tac1</td>
<td>0.6089</td>
<td>0.799198</td>
<td>Core</td>
</tr>
</tbody>
</table>

<p>14249 rows × 52 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata_cortex, target_sum<span class="op">=</span><span class="fl">1e5</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata_cortex)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata_cortex, min_mean<span class="op">=</span><span class="fl">0.0125</span>, max_mean<span class="op">=</span><span class="dv">3</span>, min_disp<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>sc.pp.scale(adata_cortex, max_value<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(adata_cortex, svd_solver<span class="op">=</span><span class="st">'arpack'</span>)</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_cortex, n_pcs<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata_cortex)</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_cortex, color<span class="op">=</span><span class="st">"subclass"</span>, legend_loc<span class="op">=</span><span class="st">'on data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>normalizing counts per cell
    finished (0:00:01)
extracting highly variable genes
    finished (0:00:06)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
... as `zero_center=True`, sparse input is densified and may lead to large memory consumption
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:10)
computing neighbors
    using 'X_pca' with n_pcs = 30
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:11)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:12)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-31-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="scanpy_08_spatial_files/figure-html/cell-31-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>adata_cortex.obs.subclass.value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="30">
<pre><code>subclass
L6 IT         1872
Sst           1741
Vip           1728
L4            1401
Pvalb         1337
Lamp5         1122
L2/3 IT        982
L6 CT          960
L5 IT          880
L5 PT          544
Astro          368
NP             362
L6b            358
Sncg           125
Endo            94
Oligo           91
VLMC            67
SMC             55
Macrophage      51
Meis2           45
Peri            32
Serpinf1        27
CR               7
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>For speed, and for a more fair comparison of the celltypes, we will subsample all celltypes to a maximum of 200 cells per class (<code>subclass</code>).</p>
<div class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>target_cells <span class="op">=</span> <span class="dv">200</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>adatas2 <span class="op">=</span> [adata_cortex[adata_cortex.obs.subclass <span class="op">==</span> clust] <span class="cf">for</span> clust <span class="kw">in</span> adata_cortex.obs.subclass.cat.categories]</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> dat <span class="kw">in</span> adatas2:</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> dat.n_obs <span class="op">&gt;</span> target_cells:</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>          sc.pp.subsample(dat, n_obs<span class="op">=</span>target_cells)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>adata_cortex <span class="op">=</span> adatas2[<span class="dv">0</span>].concatenate(<span class="op">*</span>adatas2[<span class="dv">1</span>:])</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>adata_cortex.obs.subclass.value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<pre><code>subclass
Astro         200
L6 IT         200
Sst           200
Pvalb         200
NP            200
Lamp5         200
L6b           200
Vip           200
L6 CT         200
L5 PT         200
L5 IT         200
L4            200
L2/3 IT       200
Sncg          125
Endo           94
Oligo          91
VLMC           67
SMC            55
Macrophage     51
Meis2          45
Peri           32
Serpinf1       27
CR              7
Name: count, dtype: int64</code></pre>
</div>
</div>
<div class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>    adata_cortex, color<span class="op">=</span>[<span class="st">"class"</span>, <span class="st">"subclass"</span>, <span class="st">"genotype"</span>, <span class="st">"brain_region"</span>], palette<span class="op">=</span>sc.pl.palettes.default_28</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Length of palette colors is smaller than the number of categories (palette length: 28, categories length: 61. Some categories will have the same color.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-34-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="scanpy_08_spatial_files/figure-html/cell-34-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_cortex, color<span class="op">=</span><span class="st">"subclass"</span>, legend_loc <span class="op">=</span> <span class="st">'on data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-35-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="scanpy_08_spatial_files/figure-html/cell-35-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-st_sub" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="meta-st_sub"><span class="header-section-number">5</span> Subset ST for cortex</h2>
<p>Since the scRNAseq dataset was generated from the mouse cortex, we will subset the visium dataset in order to select mainly the spots part of the cortex. Note that the integration can also be performed on the whole brain slice, but it would give rise to false positive cell type assignments and therefore it should be interpreted with more care.</p>
<p>For deconvolution we will need the counts data, so we will subset from the counts_adata object that we created earlier.</p>
<div class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>lib_a <span class="op">=</span> <span class="st">"V1_Mouse_Brain_Sagittal_Anterior"</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>counts_adata.obs[<span class="st">'clusters'</span>] <span class="op">=</span> adata.obs.clusters</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>adata_anterior_subset <span class="op">=</span> counts_adata[</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>    (counts_adata.obs.library_id <span class="op">==</span> lib_a) </span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>    <span class="op">&amp;</span> (counts_adata.obsm[<span class="st">"spatial"</span>][:, <span class="dv">1</span>] <span class="op">&lt;</span> <span class="dv">6000</span>), :</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>].copy()</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="co"># select also the cortex clusters</span></span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>adata_anterior_subset <span class="op">=</span> adata_anterior_subset[adata_anterior_subset.obs.clusters.isin([<span class="st">'3'</span>,<span class="st">'4'</span>,<span class="st">'6'</span>,<span class="st">'7'</span>]),:]</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a><span class="co"># plot to check that we have the correct regions</span></span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(</span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>    adata_anterior_subset,</span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a>    img_key<span class="op">=</span><span class="st">"hires"</span>,</span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a>    library_id <span class="op">=</span> lib_a,</span>
<span id="cb53-19"><a href="#cb53-19" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">'clusters'</span>],</span>
<span id="cb53-20"><a href="#cb53-20" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="fl">1.5</span></span>
<span id="cb53-21"><a href="#cb53-21" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-36-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="scanpy_08_spatial_files/figure-html/cell-36-output-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-st_deconv" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-st_deconv"><span class="header-section-number">6</span> Deconvolution</h2>
<p>Deconvolution is a method to estimate the abundance (or proportion) of different celltypes in a bulkRNAseq dataset using a single cell reference. As the Visium data can be seen as a small bulk, we can both use methods for traditional bulkRNAseq as well as methods especially developed for Visium data. Some methods for deconvolution are DWLS, cell2location, Tangram, Stereoscope, RCTD, SCDC and many more.</p>
<p>Here, we will use deconvolution with Stereoscope implemented in the SCVI-tools package. To read more about Stereoscope please check out this github page (https://github.com/almaan/stereoscope)</p>
<section id="meta-st_deconv_genes" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="meta-st_deconv_genes"><span class="header-section-number">6.1</span> Select genes for deconvolution</h3>
<p>Most deconvolution methods does a prior gene selection and there are different options that are used: - Use variable genes in the SC data. - Use variable genes in both SC and ST data - DE genes between clusters in the SC data.<br>
In this case we will use top DE genes per cluster, so first we have to run DGE detection on the scRNAseq data.</p>
<div class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata_cortex, <span class="st">'subclass'</span>, method <span class="op">=</span> <span class="st">"t-test"</span>, n_genes<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_dotplot(adata_cortex, n_genes<span class="op">=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ranking genes
WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.
    finished: added to `.uns['rank_genes_groups']`
    'names', sorted np.recarray to be indexed by group ids
    'scores', sorted np.recarray to be indexed by group ids
    'logfoldchanges', sorted np.recarray to be indexed by group ids
    'pvals', sorted np.recarray to be indexed by group ids
    'pvals_adj', sorted np.recarray to be indexed by group ids (0:00:03)
WARNING: dendrogram data not found (using key=dendrogram_subclass). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.
    using 'X_pca' with n_pcs = 50
Storing dendrogram info using `.uns['dendrogram_subclass']`</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="scanpy_08_spatial_files/figure-html/cell-37-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="scanpy_08_spatial_files/figure-html/cell-37-output-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>sc.tl.filter_rank_genes_groups(adata_cortex, min_fold_change<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>genes <span class="op">=</span> sc.get.rank_genes_groups_df(adata_cortex, group <span class="op">=</span> <span class="va">None</span>)</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtering genes using: min_in_group_fraction: 0.25 min_fold_change: 1, max_out_group_fraction: 0.5</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="36">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">group</th>
<th data-quarto-table-cell-role="th">names</th>
<th data-quarto-table-cell-role="th">scores</th>
<th data-quarto-table-cell-role="th">logfoldchanges</th>
<th data-quarto-table-cell-role="th">pvals</th>
<th data-quarto-table-cell-role="th">pvals_adj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Astro</td>
<td>24438</td>
<td>24.987505</td>
<td>NaN</td>
<td>3.887212e-64</td>
<td>7.446797e-63</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>Astro</td>
<td>25832</td>
<td>23.915390</td>
<td>inf</td>
<td>1.620941e-60</td>
<td>2.924028e-59</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>Astro</td>
<td>18190</td>
<td>21.893034</td>
<td>inf</td>
<td>5.711270e-55</td>
<td>9.514294e-54</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>Astro</td>
<td>25833</td>
<td>21.557812</td>
<td>inf</td>
<td>5.534148e-54</td>
<td>9.045118e-53</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>Astro</td>
<td>2882</td>
<td>20.950775</td>
<td>inf</td>
<td>2.972020e-53</td>
<td>4.796383e-52</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2295</td>
<td>Vip</td>
<td>24162</td>
<td>10.645898</td>
<td>193.386642</td>
<td>2.184734e-21</td>
<td>7.812906e-20</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2296</td>
<td>Vip</td>
<td>31218</td>
<td>10.593735</td>
<td>262.868835</td>
<td>2.901054e-21</td>
<td>1.025800e-19</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2297</td>
<td>Vip</td>
<td>18498</td>
<td>10.579262</td>
<td>NaN</td>
<td>1.337080e-21</td>
<td>4.856841e-20</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2298</td>
<td>Vip</td>
<td>24420</td>
<td>10.556447</td>
<td>NaN</td>
<td>3.184234e-21</td>
<td>1.123635e-19</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2299</td>
<td>Vip</td>
<td>29115</td>
<td>10.549688</td>
<td>322.383911</td>
<td>5.200959e-21</td>
<td>1.811284e-19</td>
</tr>
</tbody>
</table>

<p>2300 rows × 6 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>deg <span class="op">=</span> genes.names.unique().tolist()</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(deg))</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="co"># check that the genes are also present in the ST data</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>deg <span class="op">=</span> np.intersect1d(deg,adata_anterior_subset.var.index).tolist()</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(deg))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1624
0</code></pre>
</div>
</div>
<p>Train the model</p>
<p>First, train the model using scRNAseq data.</p>
<p>Stereoscope requires the data to be in counts, earlier in this tutorial we saved the spatial counts in a separate object counts_adata.</p>
<p>However, the single cell dataset that we dowloaded only has the lognormalized data in the adata.X slot, hence we will have to recalculate the count matrix.</p>
<div class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first do exponent and subtract pseudocount</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>E <span class="op">=</span> np.exp(adata_cortex.X)<span class="op">-</span><span class="dv">1</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>n <span class="op">=</span> np.<span class="bu">sum</span>(E,<span class="dv">1</span>)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(np.<span class="bu">min</span>(n), np.<span class="bu">max</span>(n))</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="co"># all sums to 1.7M</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>factor <span class="op">=</span> np.mean(n)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>nC <span class="op">=</span> np.array(adata_cortex.obs.nCount_RNA) <span class="co"># true number of counts</span></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>scaleF <span class="op">=</span> nC<span class="op">/</span>factor</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>C <span class="op">=</span> E <span class="op">*</span> scaleF[:,<span class="va">None</span>]</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>C <span class="op">=</span> C.astype(<span class="st">"int"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>151625.94316551366 10392533.010632813</code></pre>
</div>
</div>
<div class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>sc_adata <span class="op">=</span> adata_cortex.copy()</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>sc_adata.X <span class="op">=</span> C</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Setup the anndata, the implementation requires the counts matrix to be in the “counts” layer as a copy.</p>
<div class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scvi</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="co"># from scvi.data import register_tensor_from_anndata</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scvi.external <span class="im">import</span> RNAStereoscope, SpatialStereoscope</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add counts layer</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>sc_adata.layers[<span class="st">"counts"</span>] <span class="op">=</span> sc_adata.X.copy()</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a><span class="co"># subset for the selected genes</span></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>sc_adata <span class="op">=</span> sc_adata[:, deg].copy()</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a><span class="co"># create stereoscope object</span></span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>RNAStereoscope.setup_anndata(sc_adata, layer<span class="op">=</span><span class="st">"counts"</span>, labels_key<span class="op">=</span><span class="st">"subclass"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="co"># the model is saved to a file, so if is slow to run, you can simply reload it from disk by setting train = False</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>train <span class="op">=</span> <span class="va">True</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> train:</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>    sc_model <span class="op">=</span> RNAStereoscope(sc_adata)</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>    sc_model.train(max_epochs<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>    sc_model.history[<span class="st">"elbo_train"</span>][<span class="dv">10</span>:].plot()</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>    sc_model.save(<span class="st">"./data/spatial/visium/scmodel"</span>, overwrite<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>    sc_model <span class="op">=</span> RNAStereoscope.load(<span class="st">"./data/spatial/visium/scmodel"</span>, sc_adata)</span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Loaded RNA model from file!"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Predict propritions on the spatial data</p>
<p>First create a new st object with the correct genes and counts as a layer.</p>
<div class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>st_adata <span class="op">=</span> adata_anterior_subset.copy()</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>st_adata.layers[<span class="st">"counts"</span>] <span class="op">=</span> st_adata.X.copy()</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>st_adata <span class="op">=</span> st_adata[:, deg].copy()</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>SpatialStereoscope.setup_anndata(st_adata, layer<span class="op">=</span><span class="st">"counts"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>train<span class="op">=</span><span class="va">True</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> train:</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>    spatial_model <span class="op">=</span> SpatialStereoscope.from_rna_model(st_adata, sc_model)</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>    spatial_model.train(max_epochs <span class="op">=</span> <span class="dv">3000</span>)</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>    spatial_model.history[<span class="st">"elbo_train"</span>][<span class="dv">10</span>:].plot()</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>    spatial_model.save(<span class="st">"./data/spatial/stmodel"</span>, overwrite <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>    spatial_model <span class="op">=</span> SpatialStereoscope.load(<span class="st">"./data/spatial/stmodel"</span>, st_adata)</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Loaded Spatial model from file!"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get the results from the model, also put them in the .obs slot.</p>
<div class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>st_adata.obsm[<span class="st">"deconvolution"</span>] <span class="op">=</span> spatial_model.get_proportions()</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="co"># also copy to the obsm data frame</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ct <span class="kw">in</span> st_adata.obsm[<span class="st">"deconvolution"</span>].columns:</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>    st_adata.obs[ct] <span class="op">=</span> st_adata.obsm[<span class="st">"deconvolution"</span>][ct]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are then able to explore how cell types in the scRNA-seq dataset are predicted onto the visium dataset. Let’s first visualize the neurons cortical layers.</p>
<div class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    st_adata,</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    img_key<span class="op">=</span><span class="st">"hires"</span>,</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"L2/3 IT"</span>, <span class="st">"L4"</span>, <span class="st">"L5 PT"</span>, <span class="st">"L6 CT"</span>],</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>    library_id<span class="op">=</span>lib_a,</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="fl">1.5</span>,</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">2</span></span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can go ahead an visualize astrocytes and oligodendrocytes as well.</p>
<div class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>    st_adata, img_key<span class="op">=</span><span class="st">"hires"</span>, color<span class="op">=</span>[<span class="st">"Oligo"</span>, <span class="st">"Astro"</span>], size<span class="op">=</span><span class="fl">1.5</span>, library_id<span class="op">=</span>lib_a</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Keep in mind that the deconvolution results are just predictions, depending on how well your scRNAseq data covers the celltypes that are present in the ST data and on how parameters, gene selection etc. are tuned you may get different results.</p>
<div class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(st_adata, [<span class="st">"L2/3 IT"</span>, <span class="st">"L6 CT"</span>,<span class="st">"Oligo"</span>,<span class="st">"Astro"</span>],</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>            jitter<span class="op">=</span><span class="fl">0.4</span>, groupby <span class="op">=</span> <span class="st">'clusters'</span>, rotation<span class="op">=</span> <span class="dv">45</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Subset for another region that does not contain cortex cells and check what you get from the label transfer. Suggested region is the right end of the posterial section that you can select like this:</p>
<div class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this chunk has issues and therefore not evaluated</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>lib_p <span class="op">=</span> <span class="st">"V1_Mouse_Brain_Sagittal_Posterior"</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>adata_subregion <span class="op">=</span> adata[</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>    (adata.obs.library_id <span class="op">==</span> lib_p)</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>    <span class="op">&amp;</span> (adata.obsm[<span class="st">"spatial"</span>][:, <span class="dv">0</span>] <span class="op">&gt;</span> <span class="dv">6500</span>),</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>    :,</span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>].copy()</span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(</span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>    adata_subregion,</span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>    img_key<span class="op">=</span><span class="st">"hires"</span>,</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>    library_id<span class="op">=</span>lib_p,</span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">'n_genes_by_counts'</span>],</span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="fl">1.5</span></span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="meta-session" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">7</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>sc.logging.print_versions()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>-----
anndata     0.10.3
scanpy      1.9.6
-----
PIL                 10.0.0
annoy               NA
anyio               NA
asttokens           NA
attr                23.1.0
babel               2.12.1
backcall            0.2.0
certifi             2023.11.17
cffi                1.15.1
charset_normalizer  3.1.0
colorama            0.4.6
comm                0.1.3
cycler              0.12.1
cython_runtime      NA
dateutil            2.8.2
debugpy             1.6.7
decorator           5.1.1
defusedxml          0.7.1
exceptiongroup      1.2.0
executing           1.2.0
fastjsonschema      NA
fbpca               NA
gmpy2               2.1.2
h5py                3.9.0
idna                3.4
igraph              0.10.8
intervaltree        NA
ipykernel           6.23.1
ipython_genutils    0.2.0
jedi                0.18.2
jinja2              3.1.2
joblib              1.3.2
json5               NA
jsonpointer         2.0
jsonschema          4.17.3
jupyter_events      0.6.3
jupyter_server      2.6.0
jupyterlab_server   2.22.1
kiwisolver          1.4.5
leidenalg           0.10.1
llvmlite            0.41.1
louvain             0.8.1
markupsafe          2.1.2
matplotlib          3.8.0
matplotlib_inline   0.1.6
mpl_toolkits        NA
mpmath              1.3.0
natsort             8.4.0
nbformat            5.8.0
numba               0.58.1
numpy               1.26.2
opt_einsum          v3.3.0
overrides           NA
packaging           23.1
pandas              2.1.4
parso               0.8.3
patsy               0.5.5
pexpect             4.8.0
pickleshare         0.7.5
pkg_resources       NA
platformdirs        3.5.1
prometheus_client   NA
prompt_toolkit      3.0.38
psutil              5.9.5
ptyprocess          0.7.0
pure_eval           0.2.2
pvectorc            NA
pycparser           2.21
pydev_ipython       NA
pydevconsole        NA
pydevd              2.9.5
pydevd_file_utils   NA
pydevd_plugins      NA
pydevd_tracing      NA
pygments            2.15.1
pynndescent         0.5.11
pyparsing           3.1.1
pyrsistent          NA
pythonjsonlogger    NA
pytz                2023.3
requests            2.31.0
rfc3339_validator   0.1.4
rfc3986_validator   0.1.1
scanorama           1.7.4
scipy               1.11.4
seaborn             0.12.2
send2trash          NA
session_info        1.0.0
six                 1.16.0
sklearn             1.3.2
sniffio             1.3.0
socks               1.7.1
sortedcontainers    2.4.0
sparse              0.14.0
stack_data          0.6.2
statsmodels         0.14.1
sympy               1.12
texttable           1.7.0
threadpoolctl       3.2.0
torch               2.0.0
tornado             6.3.2
tqdm                4.65.0
traitlets           5.9.0
typing_extensions   NA
umap                0.5.5
urllib3             2.0.2
wcwidth             0.2.6
websocket           1.5.2
yaml                6.0
zmq                 25.0.2
zoneinfo            NA
zstandard           0.19.0
-----
IPython             8.13.2
jupyter_client      8.2.0
jupyter_core        5.3.0
jupyterlab          4.0.1
notebook            6.5.4
-----
Python 3.10.11 | packaged by conda-forge | (main, May 10 2023, 18:58:44) [GCC 11.3.0]
Linux-6.5.11-linuxkit-x86_64-with-glibc2.35
-----
Session information updated at 2024-01-16 23:30</code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2024 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"descPosition":"bottom","loop":true,"closeEffect":"zoom","openEffect":"zoom","selector":".lightbox"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>