---
title: "Seurat"
author: Åsa Björklund
date: August 22, 2019
output:
  html_notebook:
    toc: true
    toc_depth: 3
    toc_float: true
    highlight: tango
---

```{r setup, include=FALSE}
#define path to this script (wherever it is)
initial.options <- commandArgs(trailingOnly = FALSE)
script_path <- dirname(sub("--file=","",initial.options[grep("--file=",initial.options)]))
script_path <- "/Users/paulo.barenco/Box/repos/workshop-scRNAseq/labs/r_labs/seurat"
setwd(script_path)

text <- readLines("../Knit_labs.Rmd")
l <- function(x, text, lab="SEURAT"){  p = paste0("#",x,"_",lab,":")
                                       return( sub(p,"", text[grep( p ,text)] ))}
```
***
# `r l("DATA",text,"TITLE")`

`r l("DATA",text,"ALL1")`


```{bash, results='hide'}
mkdir data; cd data
curl -O http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v2/pbmc_1k_v2_filtered_feature_bc_matrix.h5
curl -O http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_filtered_feature_bc_matrix.h5
curl -O http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_protein_v3/pbmc_1k_protein_v3_filtered_feature_bc_matrix.h5
cd ..
```


```{python}
import os
import numpy as np
import pandas as pd
import scanpy as sc

os.chdir(os.getcwd())
print(os.getcwd())
```


```{python, message='hide',warnigns='hide'}
data_p3 = sc.read_10x_h5('data/pbmc_1k_protein_v3_filtered_feature_bc_matrix.h5', gex_only = True)  
# OBS! gex_only to only read gene expression data.
data_p3.var_names_make_unique()

data_v2 = sc.read_10x_h5('data/pbmc_1k_v2_filtered_feature_bc_matrix.h5')
data_v2.var_names_make_unique()

data_v3 = sc.read_10x_h5('data/pbmc_1k_v3_filtered_feature_bc_matrix.h5')
data_v3.var_names_make_unique()
```



```{python, message='hide',warnigns='hide'}
print("data_p3")
print(data_p3)
print("data_v2")
print(data_v2)
print("data_v3")
print(data_v3)
```

