---
#CSS_ALL:
editor_options:
  chunk_output_type: console
---

#CHUNK_OPT:

#DGE_TITLE:

#DGE_ALL:

#DGE_ALL2:

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(venn)
  library(dplyr)
  library(cowplot)
  library(ggplot2)
  library(pheatmap)
  library(enrichR)
})

alldata <- readRDS("data/results/covid_qc_dr_int_cl.rds")
```

```{r, fig.width=12, fig.height=6}
#Set the identity as louvain with resolution 0.3
sel.clust = "CCA_snn_res.0.3"

alldata <- SetIdent(alldata, value = sel.clust)
table(alldata@active.ident)

# plot this clustering
plot_grid(ncol = 3,
  DimPlot(alldata, label = T) + NoAxes(),
  DimPlot(alldata, group.by = "orig.ident") + NoAxes(),
  DimPlot(alldata, group.by = "type") + NoAxes() )


```

#DGE_TITLE1:

#DGE_ALL3:

```{r}
#Compute differentiall expression
markers_genes <- FindAllMarkers(alldata,
                               logfc.threshold = 0.2,
                               test.use = "wilcox",
                               min.pct = 0.1,
                               min.diff.pct = 0.2,
                               only.pos = TRUE,
                               max.cells.per.ident = 50,
                               assay = "RNA")
```

#DGE_ALL4:

```{r}
markers_genes %>% group_by(cluster)  %>% top_n(-25, p_val_adj) -> top25
top25
```

#DGE_ALL4:

```{r}
mypar(2,5,mar=c(4,6,3,1))
for(i in unique(top25$cluster)){
  barplot( sort( setNames(top25$avg_logFC, top25$gene) [top25$cluster == i], F),
           horiz = T,las=1 ,main=paste0(i," vs. rest"),border = "white", yaxs="i" )
  abline(v=c(0,0.25),lty=c(1,2))
}
```

#DGE_ALL4.1:

```{r}
markers_genes %>% group_by(cluster)  %>% top_n(-5, p_val_adj) -> top5

# create a scale.data slot for the selected genes
alldata <- ScaleData(alldata, features = as.character(unique(top5$gene)), assay = "RNA")
DoHeatmap(alldata, features = as.character(unique(top5$gene)),group.by = sel.clust, assay = "RNA")
```

#DGE_ALL4.2:

```{r, fig.height=10}
DotPlot(alldata, features = rev(as.character(unique(top5$gene))),group.by = sel.clust,assay = "RNA")+coord_flip()
```

#DGE_ALL4.3:

```{r,fig.height=10,fig.width=10}
# take top 3 genes per cluster/
top5 %>% group_by(cluster)  %>% top_n(-3, p_val) -> top3


# set pt.size to zero if you do not want all the points to hide the violin shapes, or to a small value like 0.1
VlnPlot(alldata, features = as.character(unique(top3$gene)), ncol = 5, group.by = sel.clust, assay = "RNA", pt.size = 0)
```

#DGE_ALL5:

#DGE_TITLE2:

#DGE_ALL6:

```{r,fig.height=10,fig.width=10}
# select all cells in cluster 1
cell_selection <- subset(alldata, cells = colnames(alldata)[ alldata@meta.data[,sel.clust] == 2])
cell_selection <- SetIdent(cell_selection, value = "type")
#Compute differentiall expression
DGE_cell_selection <- FindAllMarkers(cell_selection,
                               logfc.threshold = 0.2,
                               test.use = "wilcox",
                               min.pct = 0.1,
                               min.diff.pct = 0.2,
                               only.pos = TRUE,
                               max.cells.per.ident = 50,
                               assay = "RNA")
```

#DGE_ALL6.1:

```{r, fig.height=10}
DGE_cell_selection %>% group_by(cluster)  %>% top_n(-5, p_val) -> top5_cell_selection

VlnPlot(cell_selection, features = as.character(unique(top5_cell_selection$gene)),
        ncol = 5,group.by = "type",assay = "RNA", pt.size = .1)
```

#DGE_ALL6.1b:

```{r, fig.height=10}
VlnPlot(alldata, features = as.character(unique(top5_cell_selection$gene)),
        ncol = 5, split.by = "type",assay = "RNA", pt.size = 0)
```

#DGE_TITLE3:

#DGE_ALL7:

```{r}
# Load additional packages
library(enrichR)

# Check available databases to perform enrichment (then choose one)
enrichR::listEnrichrDbs()

# Perform enrichment
enrich_results <- enrichr(
 genes     = DGE_cell_selection$gene[DGE_cell_selection$cluster == "Covid"],
 databases = "GO_Biological_Process_2017b" )[[1]]

```

#DGE_ALL7.1:

```{r}
par(mar = c(3, 25, 2, 1))
barplot( height    = -log10(enrich_results$P.value)[10:1],
        names.arg = enrich_results$Term[10:1],
        horiz     = TRUE,
        las       = 1,
        border    = FALSE,
        cex.names = .6 )
abline(v = c(-log10(0.05)), lty = 2)
abline(v = 0, lty = 1)
```

#DGE_ALL7.2



```{r,fig.height=10,fig.width=10}
# select all cells in cluster 1
cell_selection <- subset(alldata, cells = colnames(alldata)[ alldata@meta.data[,sel.clust] == 2])
cell_selection <- SetIdent(cell_selection, value = "type")
#Compute differentiall expression
DGE_cell_selection <- FindMarkers(cell_selection,
                                  ident.1 = "Covid",
                               logfc.threshold = 0,
                               test.use = "wilcox",
                               min.pct = 0,
                               min.diff.pct = 0,
                               only.pos = TRUE,
                               max.cells.per.ident = 50,
                               return.thresh = 1, # <- IMPORTANT!!
                               assay = "RNA")
dim(DGE_cell_selection)
```



```{r,fig.height=10,fig.width=10}
# if folder and files doesn't exist, download it
if(!dir.exists("data/MSigDB_files/")) {
  dir.create("data/MSigDB_files")
  download.file(url="https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c2.cp.kegg.v6.2.symbols.gmt.txt","./data/MSigDB_files/c2.cp.kegg.v6.2.symbols.gmt.txt")
  download.file(url="https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c3.tft.v6.2.symbols.gmt.txt","./data/MSigDB_files/c3.tft.v6.2.symbols.gmt.txt")
  download.file(url="https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/h.all.v6.2.symbols.gmt.txt","./data/MSigDB_files/h.all.v6.2.symbols.gmt.txt")
  download.file(url="https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c2.cp.v6.2.symbols.gmt.txt","./data/MSigDB_files/c2.cp.v6.2.symbols.gmt.txt")
  download.file(url="https://raw.github.com/NBISweden/workshop-RNAseq/master/data/MSigDB_files/c5.bp.v6.2.symbols.gmt.txt","./data/MSigDB_files/c5.bp.v6.2.symbols.gmt.txt")
}
```



```{r,results='hide',block.title=TRUE,fig.height=5,fig.width=10}
library(fgsea)

# Create a gene rank based on the gene expression
gene_rank <- setNames( DGE_cell_selection$avg_logFC, casefold(rownames(DGE_cell_selection),upper=T) )

```

Once our list of genes are sorted (from highest to lowest expression in day 7), we can proceed with the enrichment itself. Here, we will be using the `fgsea()` function. This will result in a table containing information for several pathways. We can then sort and filter those pathways to visualize only the top ones. You can select/filter them by either `p-value` or normalized enrichemnet score (`NES`).

```{r,results='hide',block.title=TRUE,fig.height=5,fig.width=10}
# Load hallmark pathways
hallmark_pathways <- gmtPathways("data/MSigDB_files/c5.bp.v6.2.symbols.gmt.txt")

# Perform enrichemnt analysis
fgseaRes <- fgsea( pathways=hallmark_pathways, stats=gene_rank, minSize=15, maxSize=500, nperm=10000)

# Filter the results table to show only the top 10 UP and DOWN regulated processes (optional)
fgseaRes %>% top_n(20, NES) %>% top_n(-10, pval) -> top5_UP
fgseaRes %>% top_n(-20, NES) %>% top_n(-10, pval) -> top5_DOWN

topPathways <- c(top5_UP$pathway, rev(top5_DOWN$pathway))

# Nice summary table (shown as a plot)
dev.off()
plotGseaTable(hallmark_pathways[topPathways], gene_rank, fgseaRes, gseaParam = 0.5)
```


Which KEGG pathways are upregulated in
Which HALLMARK pathways are upregulated in
Which KEGG, GO and HALLMAK to cluster #####?
Which KEGG, GO and HALLMAK to cluster #####?

#DGE_ALL10:

```{r}
saveRDS(alldata,"data/3pbmc_qc_dr_int_cl_dge.rds")
write.csv(markers_genes)
```


#SESSION_INFO:

```{r}
sessionInfo()
```
