---
author: "Åsa Björklund  &  Paulo Czarnewski"
date: 'January 25, 2023'
output:
  html_document:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    keep_md: yes
    fig_caption: true
  html_notebook:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
editor_options:
  chunk_output_type: console
---


<style>
h1, .h1, h2, .h2, h3, .h3, h4, .h4 { margin-top: 50px }
p.caption {font-size: 0.9em;font-style: italic;color: grey;margin-right: 10%;margin-left: 10%;text-align: justify}
</style>

# Differential gene expression

In this tutorial we will cover about Differetial gene expression, which comprises an extensive range of topics and methods. In single cell, differential expresison can have multiple functionalities such as of identifying marker genes for cell populations, as well as differentially regulated genes across conditions (healthy vs control). We will also exercise on how to account the batch information in your test.

We can first load the data from the clustering session. Moreover, we can already decide which clustering resolution to use. First let's define using the `louvain` clustering to identifying differentially expressed genes.  


```r
suppressPackageStartupMessages({
    library(Seurat)
    library(dplyr)
    library(cowplot)
    library(ggplot2)
    library(pheatmap)
    library(enrichR)
    library(rafalib)
    library(Matrix)
    library(edgeR)
    library(MAST)
})

alldata <- readRDS("data/results/covid_qc_dr_int_cl.rds")
```


```r
# Set the identity as louvain with resolution 0.5
sel.clust = "CCA_snn_res.0.5"

alldata <- SetIdent(alldata, value = sel.clust)
table(alldata@active.ident)
```

```
## 
##    0    1    2    3    4    5    6    7    8    9 
## 1508 1026  573  447  430  425  423  246  237  217
```

```r
# plot this clustering
plot_grid(ncol = 3, DimPlot(alldata, label = T) + NoAxes(), DimPlot(alldata, group.by = "orig.ident") +
    NoAxes(), DimPlot(alldata, group.by = "type") + NoAxes())
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-2-1.png)<!-- -->

## Cell marker genes
***

Let us first compute a ranking for the highly differential genes in each cluster. There are many different tests and parameters to be chosen that can be used to refine your results. When looking for marker genes, we want genes that are positivelly expressed in a cell type and possibly not expressed in the others.


```r
# Compute differentiall expression
markers_genes <- FindAllMarkers(alldata, log2FC.threshold = 0.2, test.use = "wilcox",
    min.pct = 0.1, min.diff.pct = 0.2, only.pos = TRUE, max.cells.per.ident = 50,
    assay = "RNA")
```

We can now select the top 25 up regulated genes for plotting.


```r
markers_genes %>%
    group_by(cluster) %>%
    top_n(-25, p_val_adj) -> top25
top25
```

<div data-pagedtable="false">
  <script data-pagedtable-source type="application/json">
{"columns":[{"label":["p_val"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["avg_log2FC"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["pct.1"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["pct.2"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["p_val_adj"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["cluster"],"name":[6],"type":["fct"],"align":["left"]},{"label":["gene"],"name":[7],"type":["chr"],"align":["left"]}],"data":[{"1":"4.030538e-19","2":"3.6224690","3":"0.945","4":"0.055","5":"7.303738e-15","6":"0","7":"S100A12"},{"1":"8.250373e-19","2":"2.3517564","3":"0.862","4":"0.053","5":"1.495050e-14","6":"0","7":"PLBD1"},{"1":"3.796938e-18","2":"4.2581186","3":"0.999","4":"0.333","5":"6.880432e-14","6":"0","7":"S100A9"},{"1":"6.065190e-18","2":"4.5984341","3":"0.994","4":"0.247","5":"1.099073e-13","6":"0","7":"S100A8"},{"1":"7.847054e-18","2":"2.4286004","3":"0.971","4":"0.162","5":"1.421965e-13","6":"0","7":"GRN"},{"1":"1.386920e-17","2":"2.7159402","3":"0.967","4":"0.099","5":"2.513237e-13","6":"0","7":"MNDA"},{"1":"2.582759e-17","2":"3.3405302","3":"0.971","4":"0.057","5":"4.680218e-13","6":"0","7":"VCAN"},{"1":"1.180883e-16","2":"3.1490467","3":"0.987","4":"0.109","5":"2.139879e-12","6":"0","7":"AC020656.1"},{"1":"1.213828e-16","2":"1.7441296","3":"0.959","4":"0.344","5":"2.199577e-12","6":"0","7":"TALDO1"},{"1":"2.020234e-16","2":"2.2827850","3":"0.954","4":"0.069","5":"3.660866e-12","6":"0","7":"CSTA"},{"1":"3.524847e-16","2":"1.8562300","3":"0.884","4":"0.086","5":"6.387375e-12","6":"0","7":"MPEG1"},{"1":"1.922137e-15","2":"2.0356781","3":"0.930","4":"0.163","5":"3.483104e-11","6":"0","7":"CEBPD"},{"1":"2.057075e-15","2":"1.7628720","3":"0.987","4":"0.557","5":"3.727625e-11","6":"0","7":"TSPO"},{"1":"2.173184e-15","2":"2.8338407","3":"0.944","4":"0.044","5":"3.938026e-11","6":"0","7":"CD14"},{"1":"2.173184e-15","2":"2.1150104","3":"0.954","4":"0.148","5":"3.938026e-11","6":"0","7":"CYBB"},{"1":"2.361540e-15","2":"2.1277426","3":"0.975","4":"0.326","5":"4.279346e-11","6":"0","7":"APLP2"},{"1":"2.526877e-15","2":"3.0486276","3":"0.995","4":"0.122","5":"4.578954e-11","6":"0","7":"FCN1"},{"1":"2.590258e-15","2":"3.8150419","3":"1.000","4":"0.250","5":"4.693807e-11","6":"0","7":"LYZ"},{"1":"2.595774e-15","2":"1.8105266","3":"0.826","4":"0.032","5":"4.703802e-11","6":"0","7":"CD36"},{"1":"3.001071e-15","2":"1.6806881","3":"0.993","4":"0.292","5":"5.438241e-11","6":"0","7":"FCER1G"},{"1":"3.629664e-15","2":"2.3296482","3":"0.936","4":"0.049","5":"6.577315e-11","6":"0","7":"MS4A6A"},{"1":"1.014459e-14","2":"1.8965992","3":"0.889","4":"0.073","5":"1.838301e-10","6":"0","7":"TNFSF13B"},{"1":"1.755494e-14","2":"2.5046854","3":"0.999","4":"0.564","5":"3.181131e-10","6":"0","7":"CTSS"},{"1":"2.046643e-14","2":"1.9143997","3":"0.916","4":"0.085","5":"3.708722e-10","6":"0","7":"NCF2"},{"1":"2.221088e-14","2":"2.4596581","3":"0.867","4":"0.079","5":"4.024834e-10","6":"0","7":"MAFB"},{"1":"3.186372e-12","2":"1.7075141","3":"0.952","4":"0.271","5":"5.774025e-08","6":"1","7":"CD3E"},{"1":"3.146274e-11","2":"1.9714948","3":"0.991","4":"0.387","5":"5.701363e-07","6":"1","7":"CCL5"},{"1":"3.541748e-10","2":"1.4598310","3":"0.750","4":"0.213","5":"6.418001e-06","6":"1","7":"CD2"},{"1":"4.420552e-10","2":"1.9193203","3":"0.856","4":"0.162","5":"8.010482e-06","6":"1","7":"CD3G"},{"1":"4.686396e-10","2":"2.1164112","3":"0.935","4":"0.185","5":"8.492218e-06","6":"1","7":"CD3D"},{"1":"7.846456e-10","2":"1.7432483","3":"0.983","4":"0.324","5":"1.421856e-05","6":"1","7":"IL32"},{"1":"1.830217e-09","2":"1.0800652","3":"0.904","4":"0.614","5":"3.316536e-05","6":"1","7":"PPP2R5C"},{"1":"2.057523e-09","2":"1.6234940","3":"0.782","4":"0.241","5":"3.728437e-05","6":"1","7":"LYAR"},{"1":"3.250410e-09","2":"1.1832939","3":"0.758","4":"0.473","5":"5.890068e-05","6":"1","7":"PIK3R1"},{"1":"5.488231e-09","2":"1.4661147","3":"0.785","4":"0.341","5":"9.945224e-05","6":"1","7":"TRBC2"},{"1":"8.530113e-09","2":"1.5722483","3":"0.630","4":"0.073","5":"1.545742e-04","6":"1","7":"KLRG1"},{"1":"1.383278e-08","2":"1.3476845","3":"0.909","4":"0.229","5":"2.506638e-04","6":"1","7":"GZMA"},{"1":"2.562505e-08","2":"1.2047872","3":"0.821","4":"0.315","5":"4.643515e-04","6":"1","7":"SYNE2"},{"1":"4.065987e-08","2":"1.9645185","3":"0.837","4":"0.189","5":"7.367976e-04","6":"1","7":"GZMH"},{"1":"5.036795e-08","2":"1.7103344","3":"0.585","4":"0.056","5":"9.127177e-04","6":"1","7":"TRGC2"},{"1":"7.778883e-08","2":"2.2056428","3":"0.414","4":"0.043","5":"1.409611e-03","6":"1","7":"GZMK"},{"1":"2.337880e-07","2":"1.1509260","3":"0.749","4":"0.243","5":"4.236473e-03","6":"1","7":"C12orf75"},{"1":"3.884393e-07","2":"1.2338948","3":"0.955","4":"0.294","5":"7.038909e-03","6":"1","7":"CST7"},{"1":"5.121547e-07","2":"0.8341559","3":"0.553","4":"0.158","5":"9.280756e-03","6":"1","7":"MATK"},{"1":"5.863435e-07","2":"1.3591615","3":"0.983","4":"0.393","5":"1.062513e-02","6":"1","7":"NKG7"},{"1":"7.391237e-07","2":"2.1883772","3":"0.557","4":"0.070","5":"1.339366e-02","6":"1","7":"CD8A"},{"1":"9.429552e-07","2":"1.0192738","3":"0.934","4":"0.653","5":"1.708729e-02","6":"1","7":"CD99"},{"1":"9.905994e-07","2":"0.9960087","3":"0.951","4":"0.694","5":"1.795065e-02","6":"1","7":"HCST"},{"1":"1.296638e-06","2":"1.8199298","3":"0.466","4":"0.039","5":"2.349638e-02","6":"1","7":"CD8B"},{"1":"2.273294e-06","2":"1.5817532","3":"0.731","4":"0.169","5":"4.119437e-02","6":"1","7":"TRAC"},{"1":"9.291325e-20","2":"3.2847704","3":"0.974","4":"0.070","5":"1.683681e-15","6":"2","7":"CD79A"},{"1":"9.754727e-20","2":"4.2694030","3":"0.991","4":"0.061","5":"1.767654e-15","6":"2","7":"IGHM"},{"1":"1.259346e-19","2":"3.5433200","3":"0.951","4":"0.026","5":"2.282062e-15","6":"2","7":"IGHD"},{"1":"6.466082e-19","2":"2.4601181","3":"0.895","4":"0.066","5":"1.171719e-14","6":"2","7":"TNFRSF13C"},{"1":"3.530221e-18","2":"2.4936971","3":"0.908","4":"0.063","5":"6.397113e-14","6":"2","7":"MS4A1"},{"1":"3.150440e-16","2":"2.5725549","3":"0.885","4":"0.059","5":"5.708912e-12","6":"2","7":"LINC00926"},{"1":"1.168589e-15","2":"1.9915278","3":"0.974","4":"0.527","5":"2.117600e-11","6":"2","7":"HLA-DQB1"},{"1":"2.252741e-15","2":"2.1361773","3":"0.836","4":"0.065","5":"4.082192e-11","6":"2","7":"BANK1"},{"1":"2.320741e-15","2":"2.5835561","3":"0.747","4":"0.008","5":"4.205414e-11","6":"2","7":"TCL1A"},{"1":"6.215064e-15","2":"1.9036974","3":"0.991","4":"0.620","5":"1.126232e-10","6":"2","7":"HLA-DRA"},{"1":"7.321518e-15","2":"1.6054719","3":"0.974","4":"0.670","5":"1.326732e-10","6":"2","7":"HLA-DPB1"},{"1":"1.588360e-13","2":"1.2679710","3":"0.970","4":"0.686","5":"2.878268e-09","6":"2","7":"HLA-DPA1"},{"1":"6.881522e-13","2":"2.4709650","3":"0.878","4":"0.140","5":"1.247001e-08","6":"2","7":"CD79B"},{"1":"9.393499e-13","2":"2.3847103","3":"0.885","4":"0.287","5":"1.702196e-08","6":"2","7":"CD83"},{"1":"9.746224e-13","2":"2.0693421","3":"0.738","4":"0.074","5":"1.766113e-08","6":"2","7":"RALGPS2"},{"1":"3.984091e-12","2":"1.8642611","3":"0.649","4":"0.026","5":"7.219572e-08","6":"2","7":"FCER2"},{"1":"5.892288e-12","2":"1.6769173","3":"0.970","4":"0.636","5":"1.067741e-07","6":"2","7":"CXCR4"},{"1":"8.232620e-12","2":"1.5201152","3":"0.990","4":"0.710","5":"1.491833e-07","6":"2","7":"HLA-DRB1"},{"1":"2.371734e-11","2":"1.7348667","3":"0.651","4":"0.049","5":"4.297819e-07","6":"2","7":"CD22"},{"1":"7.243477e-11","2":"1.4948195","3":"0.531","4":"0.025","5":"1.312591e-06","6":"2","7":"PAX5"},{"1":"1.215491e-10","2":"1.4430905","3":"0.880","4":"0.306","5":"2.202591e-06","6":"2","7":"HLA-DQA1"},{"1":"4.375497e-10","2":"1.4093568","3":"0.799","4":"0.368","5":"7.928838e-06","6":"2","7":"TAGAP"},{"1":"4.579356e-10","2":"1.0867403","3":"0.428","4":"0.031","5":"8.298251e-06","6":"2","7":"HLA-DOB"},{"1":"6.169897e-10","2":"1.7766249","3":"0.693","4":"0.042","5":"1.118047e-05","6":"2","7":"AFF3"},{"1":"1.108248e-09","2":"0.9203767","3":"0.572","4":"0.227","5":"2.008257e-05","6":"2","7":"NCK2"},{"1":"5.817415e-15","2":"2.7361409","3":"0.966","4":"0.255","5":"1.054174e-10","6":"3","7":"PRF1"},{"1":"1.007764e-14","2":"2.4845106","3":"0.951","4":"0.309","5":"1.826170e-10","6":"3","7":"CD7"},{"1":"2.518696e-13","2":"2.1963425","3":"0.946","4":"0.327","5":"4.564130e-09","6":"3","7":"CD247"},{"1":"9.352265e-13","2":"2.1052394","3":"0.971","4":"0.313","5":"1.694724e-08","6":"3","7":"CTSW"},{"1":"9.715792e-13","2":"2.1081709","3":"0.998","4":"0.459","5":"1.760599e-08","6":"3","7":"NKG7"},{"1":"1.566427e-12","2":"2.6526719","3":"0.962","4":"0.247","5":"2.838522e-08","6":"3","7":"GZMB"},{"1":"3.585606e-12","2":"2.8877340","3":"0.861","4":"0.153","5":"6.497477e-08","6":"3","7":"SPON2"},{"1":"1.160474e-11","2":"1.6402033","3":"0.957","4":"0.514","5":"2.102896e-07","6":"3","7":"ARL4C"},{"1":"4.139793e-11","2":"1.7838602","3":"0.969","4":"0.368","5":"7.501720e-07","6":"3","7":"CST7"},{"1":"4.808598e-11","2":"2.0482418","3":"0.799","4":"0.100","5":"8.713660e-07","6":"3","7":"KLRF1"},{"1":"6.874122e-11","2":"2.7291248","3":"0.944","4":"0.300","5":"1.245660e-06","6":"3","7":"GNLY"},{"1":"1.023481e-10","2":"1.9361259","3":"0.906","4":"0.222","5":"1.854651e-06","6":"3","7":"KLRD1"},{"1":"2.433220e-10","2":"2.1184554","3":"0.828","4":"0.155","5":"4.409238e-06","6":"3","7":"KLRB1"},{"1":"6.757177e-10","2":"1.7371394","3":"0.848","4":"0.269","5":"1.224468e-05","6":"3","7":"HOPX"},{"1":"9.344156e-10","2":"1.9670123","3":"0.711","4":"0.109","5":"1.693255e-05","6":"3","7":"IL2RB"},{"1":"9.956854e-10","2":"1.7217385","3":"0.895","4":"0.328","5":"1.804281e-05","6":"3","7":"GZMM"},{"1":"1.940795e-09","2":"2.2812367","3":"0.779","4":"0.115","5":"3.516914e-05","6":"3","7":"CLIC3"},{"1":"3.658457e-09","2":"2.0644641","3":"0.852","4":"0.231","5":"6.629489e-05","6":"3","7":"FGFBP2"},{"1":"5.611335e-08","2":"1.2789362","3":"0.868","4":"0.499","5":"1.016830e-03","6":"3","7":"IFITM1"},{"1":"5.880818e-08","2":"1.1957754","3":"0.843","4":"0.504","5":"1.065663e-03","6":"3","7":"RUNX3"},{"1":"1.369476e-07","2":"1.6462541","3":"0.635","4":"0.177","5":"2.481627e-03","6":"3","7":"IGFBP7"},{"1":"2.093524e-07","2":"1.5009540","3":"0.718","4":"0.188","5":"3.793675e-03","6":"3","7":"MATK"},{"1":"3.180035e-07","2":"1.5517505","3":"0.660","4":"0.138","5":"5.762542e-03","6":"3","7":"S1PR5"},{"1":"3.298930e-07","2":"1.8149792","3":"0.915","4":"0.306","5":"5.977992e-03","6":"3","7":"GZMA"},{"1":"3.528957e-07","2":"1.1266333","3":"0.946","4":"0.669","5":"6.394822e-03","6":"3","7":"EFHD2"},{"1":"2.236156e-11","2":"2.0463232","3":"0.972","4":"0.362","5":"4.052138e-07","6":"4","7":"LTB"},{"1":"4.027791e-09","2":"2.2134488","3":"0.953","4":"0.198","5":"7.298761e-05","6":"4","7":"IL7R"},{"1":"6.418562e-09","2":"1.5587245","3":"0.953","4":"0.530","5":"1.163108e-04","6":"4","7":"LDHB"},{"1":"2.859897e-08","2":"1.5529894","3":"0.902","4":"0.220","5":"5.182419e-04","6":"4","7":"TRAC"},{"1":"6.122875e-08","2":"1.0822373","3":"0.986","4":"0.400","5":"1.109526e-03","6":"4","7":"IL32"},{"1":"1.526885e-07","2":"0.7590357","3":"0.758","4":"0.464","5":"2.766868e-03","6":"4","7":"EML4"},{"1":"1.009591e-06","2":"0.5823488","3":"0.737","4":"0.487","5":"1.829480e-02","6":"4","7":"ITGB1"},{"1":"1.078850e-06","2":"0.8115917","3":"0.512","4":"0.144","5":"1.954985e-02","6":"4","7":"SFXN1"},{"1":"1.123498e-06","2":"0.7566167","3":"0.826","4":"0.246","5":"2.035890e-02","6":"4","7":"CD3G"},{"1":"1.765159e-06","2":"0.8350560","3":"0.326","4":"0.023","5":"3.198645e-02","6":"4","7":"TNFRSF4"},{"1":"2.562025e-06","2":"0.6507726","3":"0.907","4":"0.671","5":"4.642645e-02","6":"4","7":"ANKRD12"},{"1":"2.802195e-06","2":"0.5532297","3":"0.484","4":"0.221","5":"5.077858e-02","6":"4","7":"ATF7IP2"},{"1":"3.560908e-06","2":"0.8540391","3":"0.909","4":"0.631","5":"6.452722e-02","6":"4","7":"GSTK1"},{"1":"4.141803e-06","2":"1.2020411","3":"0.653","4":"0.114","5":"7.505360e-02","6":"4","7":"RCAN3"},{"1":"4.282309e-06","2":"1.0154462","3":"0.442","4":"0.037","5":"7.759972e-02","6":"4","7":"AQP3"},{"1":"5.471898e-06","2":"0.8662807","3":"0.944","4":"0.351","5":"9.915626e-02","6":"4","7":"CD3E"},{"1":"1.423412e-05","2":"0.5305566","3":"0.449","4":"0.130","5":"2.579365e-01","6":"4","7":"BCL11B"},{"1":"2.321735e-05","2":"0.5933431","3":"0.495","4":"0.240","5":"4.207217e-01","6":"4","7":"CRYBG1"},{"1":"2.342179e-05","2":"0.7822442","3":"0.426","4":"0.067","5":"4.244262e-01","6":"4","7":"TNFRSF25"},{"1":"2.389055e-05","2":"0.4308161","3":"0.530","4":"0.239","5":"4.329206e-01","6":"4","7":"LAT"},{"1":"3.334832e-05","2":"0.9739107","3":"0.677","4":"0.193","5":"6.043048e-01","6":"4","7":"TCF7"},{"1":"4.140180e-05","2":"0.5807565","3":"0.651","4":"0.283","5":"7.502420e-01","6":"4","7":"PRDX2"},{"1":"4.356439e-05","2":"0.9701924","3":"0.972","4":"0.770","5":"7.894303e-01","6":"4","7":"KLF2"},{"1":"5.228216e-05","2":"1.1519699","3":"0.823","4":"0.390","5":"9.474050e-01","6":"4","7":"TRBC2"},{"1":"5.516405e-05","2":"0.8936803","3":"0.821","4":"0.328","5":"9.996277e-01","6":"4","7":"RORA"},{"1":"4.437526e-06","2":"0.2919710","3":"0.209","4":"0.526","5":"8.041241e-02","6":"5","7":"ENY2"},{"1":"9.118839e-06","2":"0.2938020","3":"0.122","4":"0.336","5":"1.652425e-01","6":"5","7":"MTRNR2L8"},{"1":"2.865683e-05","2":"0.4827007","3":"0.355","4":"0.820","5":"5.192903e-01","6":"5","7":"SSR4"},{"1":"3.565977e-05","2":"0.2951226","3":"0.214","4":"0.618","5":"6.461906e-01","6":"5","7":"RTRAF"},{"1":"6.441553e-05","2":"0.3287058","3":"0.311","4":"0.770","5":"1.000000e+00","6":"5","7":"GUK1"},{"1":"1.312838e-04","2":"0.2674065","3":"0.188","4":"0.499","5":"1.000000e+00","6":"5","7":"COX14"},{"1":"1.342544e-04","2":"0.6345142","3":"0.285","4":"0.676","5":"1.000000e+00","6":"5","7":"ARL6IP5"},{"1":"1.701177e-04","2":"0.3999925","3":"0.282","4":"0.515","5":"1.000000e+00","6":"5","7":"NPC2"},{"1":"1.998318e-04","2":"0.3048495","3":"0.165","4":"0.475","5":"1.000000e+00","6":"5","7":"RALY"},{"1":"2.106093e-04","2":"0.3113956","3":"0.202","4":"0.586","5":"1.000000e+00","6":"5","7":"UBE2L3"},{"1":"2.117490e-04","2":"0.4825479","3":"0.242","4":"0.518","5":"1.000000e+00","6":"5","7":"GDI2"},{"1":"2.182623e-04","2":"0.2533031","3":"0.155","4":"0.503","5":"1.000000e+00","6":"5","7":"PHB2"},{"1":"2.343720e-04","2":"0.2803755","3":"0.238","4":"0.618","5":"1.000000e+00","6":"5","7":"COPE"},{"1":"2.538858e-04","2":"0.3218451","3":"0.113","4":"0.326","5":"1.000000e+00","6":"5","7":"SCAMP2"},{"1":"2.990571e-04","2":"0.3895540","3":"0.216","4":"0.535","5":"1.000000e+00","6":"5","7":"GNB1"},{"1":"3.118270e-04","2":"0.4055566","3":"0.289","4":"0.657","5":"1.000000e+00","6":"5","7":"BRK1"},{"1":"3.367555e-04","2":"0.3834210","3":"0.184","4":"0.459","5":"1.000000e+00","6":"5","7":"TRAM1"},{"1":"4.554685e-04","2":"0.4037718","3":"0.148","4":"0.379","5":"1.000000e+00","6":"5","7":"ATG3"},{"1":"4.922118e-04","2":"0.3549792","3":"0.240","4":"0.625","5":"1.000000e+00","6":"5","7":"UFC1"},{"1":"5.599402e-04","2":"0.3381469","3":"0.113","4":"0.341","5":"1.000000e+00","6":"5","7":"JARID2"},{"1":"5.674472e-04","2":"0.3098839","3":"0.271","4":"0.721","5":"1.000000e+00","6":"5","7":"PSMB1"},{"1":"6.904868e-04","2":"0.3551764","3":"0.167","4":"0.427","5":"1.000000e+00","6":"5","7":"UBE2J1"},{"1":"7.520536e-04","2":"0.2721731","3":"0.111","4":"0.333","5":"1.000000e+00","6":"5","7":"HMGA1"},{"1":"8.023394e-04","2":"0.2755823","3":"0.120","4":"0.422","5":"1.000000e+00","6":"5","7":"TANK"},{"1":"8.640894e-04","2":"0.3666120","3":"0.240","4":"0.541","5":"1.000000e+00","6":"5","7":"RTN4"},{"1":"1.056958e-03","2":"0.2947947","3":"0.261","4":"0.464","5":"1.000000e+00","6":"5","7":"PYCARD"},{"1":"1.105767e-03","2":"0.3231236","3":"0.238","4":"0.567","5":"1.000000e+00","6":"5","7":"SELENOW"},{"1":"1.213090e-03","2":"0.2880647","3":"0.134","4":"0.384","5":"1.000000e+00","6":"5","7":"ZFAND6"},{"1":"1.237938e-03","2":"0.2876243","3":"0.148","4":"0.411","5":"1.000000e+00","6":"5","7":"SSR3"},{"1":"1.280987e-03","2":"0.4715397","3":"0.327","4":"0.700","5":"1.000000e+00","6":"5","7":"PSME2"},{"1":"1.593714e-03","2":"0.4506568","3":"0.176","4":"0.440","5":"1.000000e+00","6":"5","7":"PCNP"},{"1":"1.662503e-03","2":"0.4481783","3":"0.358","4":"0.738","5":"1.000000e+00","6":"5","7":"FLNA"},{"1":"1.664278e-03","2":"0.7486715","3":"0.184","4":"0.438","5":"1.000000e+00","6":"5","7":"DNAJC15"},{"1":"1.726300e-03","2":"0.5778727","3":"0.162","4":"0.447","5":"1.000000e+00","6":"5","7":"PFDN2"},{"1":"1.785180e-03","2":"0.5127436","3":"0.226","4":"0.456","5":"1.000000e+00","6":"5","7":"GLRX"},{"1":"1.852096e-03","2":"0.2631331","3":"0.181","4":"0.452","5":"1.000000e+00","6":"5","7":"GLIPR2"},{"1":"1.953455e-03","2":"0.2885995","3":"0.289","4":"0.521","5":"1.000000e+00","6":"5","7":"APLP2"},{"1":"2.041627e-03","2":"0.2506980","3":"0.416","4":"0.890","5":"1.000000e+00","6":"5","7":"MYL12B"},{"1":"2.101798e-03","2":"0.3942688","3":"0.266","4":"0.683","5":"1.000000e+00","6":"5","7":"ATP5MC3"},{"1":"2.142603e-03","2":"1.0510202","3":"0.240","4":"0.548","5":"1.000000e+00","6":"5","7":"RPS27L"},{"1":"2.181678e-03","2":"0.2896366","3":"0.125","4":"0.375","5":"1.000000e+00","6":"5","7":"VDAC1"},{"1":"2.318256e-03","2":"0.5066689","3":"0.245","4":"0.578","5":"1.000000e+00","6":"5","7":"ATP5PD"},{"1":"2.445409e-03","2":"0.3058299","3":"0.224","4":"0.629","5":"1.000000e+00","6":"5","7":"SRSF2"},{"1":"2.656454e-03","2":"0.4942823","3":"0.188","4":"0.472","5":"1.000000e+00","6":"5","7":"PSMA4"},{"1":"2.752139e-03","2":"0.3951306","3":"0.139","4":"0.359","5":"1.000000e+00","6":"5","7":"SLC25A37"},{"1":"3.045022e-03","2":"0.2664375","3":"0.136","4":"0.378","5":"1.000000e+00","6":"5","7":"ATP5MC1"},{"1":"3.320210e-03","2":"0.2795745","3":"0.226","4":"0.569","5":"1.000000e+00","6":"5","7":"ATF4"},{"1":"3.662275e-03","2":"0.7141439","3":"0.278","4":"0.601","5":"1.000000e+00","6":"5","7":"TRAPPC1"},{"1":"4.280046e-03","2":"0.3044374","3":"0.132","4":"0.355","5":"1.000000e+00","6":"5","7":"ACADVL"},{"1":"4.315331e-03","2":"0.3938197","3":"0.336","4":"0.706","5":"1.000000e+00","6":"5","7":"CAP1"},{"1":"4.368616e-03","2":"1.1604329","3":"0.249","4":"0.621","5":"1.000000e+00","6":"5","7":"EIF1B"},{"1":"4.429675e-03","2":"0.4817489","3":"0.212","4":"0.504","5":"1.000000e+00","6":"5","7":"BCAP31"},{"1":"4.452050e-03","2":"0.3210853","3":"0.141","4":"0.383","5":"1.000000e+00","6":"5","7":"PSMB7"},{"1":"4.522130e-03","2":"0.2622762","3":"0.169","4":"0.498","5":"1.000000e+00","6":"5","7":"RNPEPL1"},{"1":"4.800301e-03","2":"0.4851875","3":"0.165","4":"0.404","5":"1.000000e+00","6":"5","7":"C1orf43"},{"1":"4.873645e-03","2":"0.3014470","3":"0.315","4":"0.711","5":"1.000000e+00","6":"5","7":"LY6E"},{"1":"4.892906e-03","2":"0.3181010","3":"0.162","4":"0.435","5":"1.000000e+00","6":"5","7":"MMP24OS"},{"1":"5.022478e-03","2":"0.4838211","3":"0.172","4":"0.392","5":"1.000000e+00","6":"5","7":"ZYX"},{"1":"5.187567e-03","2":"0.3077304","3":"0.202","4":"0.517","5":"1.000000e+00","6":"5","7":"LENG8"},{"1":"5.378745e-03","2":"0.4856319","3":"0.289","4":"0.654","5":"1.000000e+00","6":"5","7":"SNX3"},{"1":"5.492468e-03","2":"0.3415987","3":"0.167","4":"0.441","5":"1.000000e+00","6":"5","7":"RAB8A"},{"1":"5.535551e-03","2":"0.3867997","3":"0.176","4":"0.377","5":"1.000000e+00","6":"5","7":"HSBP1"},{"1":"5.639706e-03","2":"0.2996742","3":"0.120","4":"0.410","5":"1.000000e+00","6":"5","7":"PSMG2"},{"1":"6.112307e-03","2":"0.2842820","3":"0.287","4":"0.696","5":"1.000000e+00","6":"5","7":"SLC25A5"},{"1":"6.236756e-03","2":"0.4296866","3":"0.304","4":"0.693","5":"1.000000e+00","6":"5","7":"ARPC4"},{"1":"6.275851e-03","2":"0.4796176","3":"0.200","4":"0.537","5":"1.000000e+00","6":"5","7":"NDUFV2"},{"1":"6.285963e-03","2":"0.3914180","3":"0.224","4":"0.538","5":"1.000000e+00","6":"5","7":"CDC37"},{"1":"6.342945e-03","2":"0.4195533","3":"0.264","4":"0.607","5":"1.000000e+00","6":"5","7":"ATP5MF"},{"1":"6.477289e-03","2":"0.6171572","3":"0.273","4":"0.618","5":"1.000000e+00","6":"5","7":"ATP5MD"},{"1":"7.217976e-03","2":"0.5332049","3":"0.186","4":"0.482","5":"1.000000e+00","6":"5","7":"GTF2A2"},{"1":"7.711745e-03","2":"0.2955819","3":"0.148","4":"0.462","5":"1.000000e+00","6":"5","7":"SLC3A2"},{"1":"7.836979e-03","2":"0.4081263","3":"0.160","4":"0.372","5":"1.000000e+00","6":"5","7":"POLR2G"},{"1":"8.024732e-03","2":"0.4496095","3":"0.318","4":"0.691","5":"1.000000e+00","6":"5","7":"ATP6V0E1"},{"1":"8.431749e-03","2":"0.3338597","3":"0.271","4":"0.542","5":"1.000000e+00","6":"5","7":"MT2A"},{"1":"8.526934e-03","2":"0.5661826","3":"0.501","4":"0.829","5":"1.000000e+00","6":"5","7":"RHOA"},{"1":"8.697043e-03","2":"0.5146182","3":"0.205","4":"0.519","5":"1.000000e+00","6":"5","7":"CBX3"},{"1":"8.697043e-03","2":"0.3269341","3":"0.155","4":"0.418","5":"1.000000e+00","6":"5","7":"PSMB2"},{"1":"8.708088e-03","2":"0.4240587","3":"0.226","4":"0.699","5":"1.000000e+00","6":"5","7":"RAN"},{"1":"8.916632e-03","2":"0.3550228","3":"0.155","4":"0.356","5":"1.000000e+00","6":"5","7":"COMT"},{"1":"9.158027e-03","2":"0.4490588","3":"0.193","4":"0.434","5":"1.000000e+00","6":"5","7":"TCIRG1"},{"1":"9.402540e-03","2":"0.3519091","3":"0.179","4":"0.496","5":"1.000000e+00","6":"5","7":"CNN2"},{"1":"9.988135e-03","2":"0.2904221","3":"0.155","4":"0.432","5":"1.000000e+00","6":"5","7":"SNF8"},{"1":"3.385047e-15","2":"2.2451038","3":"1.000","4":"0.462","5":"6.134043e-11","6":"6","7":"NKG7"},{"1":"4.802489e-15","2":"2.1234417","3":"0.991","4":"0.369","5":"8.702591e-11","6":"6","7":"CST7"},{"1":"5.374537e-15","2":"2.0813713","3":"0.986","4":"0.248","5":"9.739199e-11","6":"6","7":"GZMB"},{"1":"2.752900e-14","2":"1.8580003","3":"0.920","4":"0.224","5":"4.988531e-10","6":"6","7":"KLRD1"},{"1":"2.271558e-13","2":"2.1419871","3":"0.853","4":"0.108","5":"4.116291e-09","6":"6","7":"TRDC"},{"1":"3.242041e-13","2":"2.0620864","3":"0.967","4":"0.317","5":"5.874903e-09","6":"6","7":"CTSW"},{"1":"5.864718e-13","2":"2.4467018","3":"0.962","4":"0.301","5":"1.062745e-08","6":"6","7":"GNLY"},{"1":"1.043732e-12","2":"2.4757850","3":"0.946","4":"0.226","5":"1.891347e-08","6":"6","7":"FGFBP2"},{"1":"2.675995e-11","2":"1.8713424","3":"0.917","4":"0.262","5":"4.849171e-07","6":"6","7":"PRF1"},{"1":"4.124042e-11","2":"1.7439545","3":"0.943","4":"0.330","5":"7.473177e-07","6":"6","7":"CD247"},{"1":"9.781405e-11","2":"1.7523923","3":"0.943","4":"0.306","5":"1.772488e-06","6":"6","7":"GZMA"},{"1":"1.863545e-10","2":"1.9346992","3":"0.995","4":"0.458","5":"3.376930e-06","6":"6","7":"CCL5"},{"1":"2.142262e-10","2":"2.0502761","3":"0.976","4":"0.254","5":"3.881993e-06","6":"6","7":"GZMH"},{"1":"1.022898e-09","2":"1.4586756","3":"0.877","4":"0.253","5":"1.853594e-05","6":"6","7":"FCGR3A"},{"1":"2.002225e-09","2":"1.6461091","3":"0.872","4":"0.270","5":"3.628232e-05","6":"6","7":"HOPX"},{"1":"2.464828e-09","2":"1.3358857","3":"0.801","4":"0.298","5":"4.466515e-05","6":"6","7":"C12orf75"},{"1":"2.650588e-09","2":"1.3741495","3":"0.652","4":"0.173","5":"4.803130e-05","6":"6","7":"SPON2"},{"1":"4.561523e-09","2":"1.2120891","3":"0.868","4":"0.566","5":"8.265936e-05","6":"6","7":"ABHD17A"},{"1":"1.050265e-08","2":"1.9324099","3":"0.759","4":"0.107","5":"1.903185e-04","6":"6","7":"KLRF1"},{"1":"1.611637e-08","2":"1.2270629","3":"0.636","4":"0.142","5":"2.920447e-04","6":"6","7":"S1PR5"},{"1":"4.047382e-08","2":"1.1300278","3":"0.939","4":"0.671","5":"7.334260e-04","6":"6","7":"EFHD2"},{"1":"5.628031e-08","2":"1.0894544","3":"0.950","4":"0.517","5":"1.019855e-03","6":"6","7":"ARL4C"},{"1":"9.313276e-08","2":"1.5377245","3":"0.655","4":"0.166","5":"1.687659e-03","6":"6","7":"SYNGR1"},{"1":"1.098802e-07","2":"1.1885720","3":"0.922","4":"0.626","5":"1.991139e-03","6":"6","7":"CD81"},{"1":"1.182468e-07","2":"1.2995801","3":"0.837","4":"0.233","5":"2.142750e-03","6":"6","7":"TRBC1"},{"1":"1.915495e-15","2":"2.0018200","3":"0.793","4":"0.121","5":"3.471069e-11","6":"7","7":"CCR7"},{"1":"3.118902e-14","2":"1.4263067","3":"0.850","4":"0.550","5":"5.651762e-10","6":"7","7":"LDHB"},{"1":"2.240386e-13","2":"1.6476642","3":"0.943","4":"0.385","5":"4.059803e-09","6":"7","7":"LTB"},{"1":"2.722159e-13","2":"1.8946276","3":"0.951","4":"0.224","5":"4.932824e-09","6":"7","7":"IL7R"},{"1":"4.479717e-13","2":"1.3354267","3":"0.854","4":"0.482","5":"8.117695e-09","6":"7","7":"LEPROTL1"},{"1":"1.270833e-11","2":"1.6731276","3":"0.915","4":"0.407","5":"2.302877e-07","6":"7","7":"PIK3IP1"},{"1":"5.833127e-11","2":"1.7869818","3":"0.768","4":"0.205","5":"1.057021e-06","6":"7","7":"TCF7"},{"1":"1.837198e-10","2":"1.5941113","3":"0.553","4":"0.056","5":"3.329187e-06","6":"7","7":"LEF1"},{"1":"4.579356e-10","2":"1.3735322","3":"0.516","4":"0.051","5":"8.298251e-06","6":"7","7":"MAL"},{"1":"4.695111e-10","2":"1.3831783","3":"0.561","4":"0.137","5":"8.508011e-06","6":"7","7":"RCAN3"},{"1":"1.266033e-09","2":"1.3823776","3":"0.801","4":"0.377","5":"2.294178e-05","6":"7","7":"GIMAP7"},{"1":"3.069007e-09","2":"1.2668890","3":"0.472","4":"0.076","5":"5.561347e-05","6":"7","7":"CAMK4"},{"1":"3.336130e-09","2":"1.0434694","3":"0.703","4":"0.253","5":"6.045401e-05","6":"7","7":"TRAC"},{"1":"7.779858e-09","2":"0.7927264","3":"0.341","4":"0.083","5":"1.409788e-04","6":"7","7":"OXNAD1"},{"1":"9.506278e-09","2":"1.0453187","3":"0.793","4":"0.520","5":"1.722633e-04","6":"7","7":"FOXP1"},{"1":"1.059525e-08","2":"1.4457249","3":"0.512","4":"0.046","5":"1.919965e-04","6":"7","7":"TRABD2A"},{"1":"3.437082e-08","2":"1.7218195","3":"0.463","4":"0.018","5":"6.228336e-04","6":"7","7":"TSHZ2"},{"1":"7.886416e-08","2":"1.2711996","3":"0.549","4":"0.152","5":"1.429098e-03","6":"7","7":"PDE3B"},{"1":"1.841884e-07","2":"0.7580581","3":"0.675","4":"0.457","5":"3.337679e-03","6":"7","7":"SEPT6"},{"1":"3.087850e-07","2":"0.8817861","3":"0.431","4":"0.075","5":"5.595492e-03","6":"7","7":"CD27"},{"1":"3.197791e-07","2":"1.1384444","3":"0.488","4":"0.140","5":"5.794717e-03","6":"7","7":"BCL11B"},{"1":"3.419550e-07","2":"0.7769008","3":"0.520","4":"0.251","5":"6.196567e-03","6":"7","7":"NCK2"},{"1":"5.170199e-07","2":"0.9564802","3":"0.724","4":"0.510","5":"9.368917e-03","6":"7","7":"GCC2"},{"1":"8.258497e-07","2":"0.7774203","3":"0.321","4":"0.089","5":"1.496522e-02","6":"7","7":"RIC3"},{"1":"1.175099e-06","2":"1.2346668","3":"0.520","4":"0.121","5":"2.129396e-02","6":"7","7":"SERINC5"},{"1":"8.252986e-12","2":"2.6328602","3":"0.996","4":"0.126","5":"1.495524e-07","6":"8","7":"CD79A"},{"1":"6.436950e-11","2":"2.4252605","3":"0.932","4":"0.116","5":"1.166440e-06","6":"8","7":"MS4A1"},{"1":"6.436950e-11","2":"2.1678914","3":"0.949","4":"0.109","5":"1.166440e-06","6":"8","7":"BANK1"},{"1":"5.001381e-10","2":"1.2799810","3":"0.595","4":"0.008","5":"9.063003e-06","6":"8","7":"TNFRSF13B"},{"1":"8.830772e-10","2":"1.4558265","3":"0.591","4":"0.042","5":"1.600224e-05","6":"8","7":"IGHG3"},{"1":"1.050088e-09","2":"1.1520357","3":"0.620","4":"0.140","5":"1.902865e-05","6":"8","7":"CCDC50"},{"1":"3.462245e-09","2":"1.7854300","3":"0.958","4":"0.339","5":"6.273935e-05","6":"8","7":"HLA-DQA1"},{"1":"3.463759e-09","2":"1.0859628","3":"0.679","4":"0.171","5":"6.276677e-05","6":"8","7":"SWAP70"},{"1":"5.662033e-09","2":"1.8842464","3":"0.916","4":"0.117","5":"1.026017e-04","6":"8","7":"TNFRSF13C"},{"1":"1.928859e-08","2":"1.4027635","3":"0.658","4":"0.045","5":"3.495285e-04","6":"8","7":"SPIB"},{"1":"3.815461e-08","2":"1.4287424","3":"0.840","4":"0.188","5":"6.913997e-04","6":"8","7":"CD79B"},{"1":"4.792339e-08","2":"1.5450282","3":"0.882","4":"0.388","5":"8.684198e-04","6":"8","7":"LTB"},{"1":"5.286240e-08","2":"1.1620896","3":"0.565","4":"0.041","5":"9.579196e-04","6":"8","7":"POU2AF1"},{"1":"8.632895e-08","2":"1.0324477","3":"0.578","4":"0.081","5":"1.564367e-03","6":"8","7":"GNG7"},{"1":"8.670423e-08","2":"1.4640806","3":"0.722","4":"0.057","5":"1.571167e-03","6":"8","7":"BLK"},{"1":"1.802599e-07","2":"1.5545337","3":"0.523","4":"0.026","5":"3.266490e-03","6":"8","7":"IGHG1"},{"1":"3.259620e-07","2":"0.6634841","3":"0.730","4":"0.172","5":"5.906757e-03","6":"8","7":"PDLIM1"},{"1":"3.646974e-07","2":"1.3765510","3":"0.831","4":"0.114","5":"6.608682e-03","6":"8","7":"LINC00926"},{"1":"4.431302e-07","2":"1.2596012","3":"0.506","4":"0.038","5":"8.029963e-03","6":"8","7":"COBLL1"},{"1":"7.747689e-07","2":"1.3761180","3":"0.937","4":"0.574","5":"1.403959e-02","6":"8","7":"EZR"},{"1":"1.464671e-06","2":"1.0001136","3":"0.456","4":"0.035","5":"2.654130e-02","6":"8","7":"CD24"},{"1":"1.500868e-06","2":"0.8108956","3":"0.380","4":"0.033","5":"2.719723e-02","6":"8","7":"FCRL2"},{"1":"1.578954e-06","2":"0.4807777","3":"0.338","4":"0.080","5":"2.861223e-02","6":"8","7":"CD27"},{"1":"1.739731e-06","2":"1.5813280","3":"0.987","4":"0.689","5":"3.152566e-02","6":"8","7":"HLA-DPB1"},{"1":"1.984374e-06","2":"0.9253058","3":"0.515","4":"0.083","5":"3.595884e-02","6":"8","7":"ADAM28"},{"1":"2.561561e-17","2":"3.0410594","3":"1.000","4":"0.385","5":"4.641805e-13","6":"9","7":"LST1"},{"1":"8.659509e-16","2":"1.9995177","3":"0.945","4":"0.218","5":"1.569190e-11","6":"9","7":"LRRC25"},{"1":"1.589283e-15","2":"2.3557871","3":"1.000","4":"0.643","5":"2.879940e-11","6":"9","7":"COTL1"},{"1":"2.750892e-15","2":"1.7138645","3":"0.908","4":"0.181","5":"4.984892e-11","6":"9","7":"RRAS"},{"1":"3.097556e-15","2":"2.2362868","3":"0.908","4":"0.151","5":"5.613081e-11","6":"9","7":"MS4A7"},{"1":"4.716134e-15","2":"2.1474430","3":"0.949","4":"0.154","5":"8.546106e-11","6":"9","7":"SMIM25"},{"1":"5.938452e-15","2":"2.5692811","3":"0.982","4":"0.273","5":"1.076107e-10","6":"9","7":"FCGR3A"},{"1":"1.256327e-14","2":"3.2206138","3":"0.899","4":"0.037","5":"2.276591e-10","6":"9","7":"CDKN1C"},{"1":"1.272906e-14","2":"2.1881493","3":"0.986","4":"0.313","5":"2.306633e-10","6":"9","7":"SERPINA1"},{"1":"1.900311e-14","2":"1.9601161","3":"1.000","4":"0.611","5":"3.443553e-10","6":"9","7":"PSAP"},{"1":"3.623702e-14","2":"1.9164757","3":"0.885","4":"0.109","5":"6.566510e-10","6":"9","7":"CSF1R"},{"1":"4.186014e-14","2":"1.6071376","3":"0.935","4":"0.226","5":"7.585476e-10","6":"9","7":"PILRA"},{"1":"5.766298e-14","2":"1.9513618","3":"1.000","4":"0.391","5":"1.044911e-09","6":"9","7":"CST3"},{"1":"1.250180e-13","2":"1.7846489","3":"0.922","4":"0.275","5":"2.265451e-09","6":"9","7":"PECAM1"},{"1":"2.017684e-13","2":"1.4923717","3":"0.972","4":"0.463","5":"3.656246e-09","6":"9","7":"TNFRSF1B"},{"1":"2.366411e-13","2":"1.6142855","3":"0.853","4":"0.105","5":"4.288173e-09","6":"9","7":"TCF7L2"},{"1":"2.400518e-13","2":"1.0800473","3":"0.862","4":"0.231","5":"4.349979e-09","6":"9","7":"TBXAS1"},{"1":"3.245009e-13","2":"1.1558180","3":"0.737","4":"0.106","5":"5.880280e-09","6":"9","7":"CAMK1"},{"1":"4.157467e-13","2":"1.5670518","3":"0.857","4":"0.171","5":"7.533746e-09","6":"9","7":"HMOX1"},{"1":"7.429236e-13","2":"1.9319518","3":"0.982","4":"0.360","5":"1.346252e-08","6":"9","7":"SPI1"},{"1":"1.184499e-12","2":"1.6773960","3":"0.972","4":"0.245","5":"2.146431e-08","6":"9","7":"LILRB2"},{"1":"1.789602e-12","2":"1.7092166","3":"0.963","4":"0.309","5":"3.242937e-08","6":"9","7":"CD68"},{"1":"1.802095e-12","2":"1.4988401","3":"0.903","4":"0.219","5":"3.265577e-08","6":"9","7":"LMO2"},{"1":"3.527285e-12","2":"1.6812206","3":"0.986","4":"0.477","5":"6.391794e-08","6":"9","7":"NPC2"},{"1":"3.597583e-12","2":"1.6991356","3":"0.982","4":"0.500","5":"6.519181e-08","6":"9","7":"VSIR"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>

We can now select the top 25 up regulated genes for plotting.


```r
mypar(2, 5, mar = c(4, 6, 3, 1))
for (i in unique(top25$cluster)) {
    barplot(sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], F),
        horiz = T, las = 1, main = paste0(i, " vs. rest"), border = "white", yaxs = "i")
    abline(v = c(0, 0.25), lty = c(1, 2))
}
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-5-1.png)<!-- -->

We can visualize them as a heatmap. Here we are selecting the top 5.


```r
markers_genes %>%
    group_by(cluster) %>%
    top_n(-5, p_val_adj) -> top5

# create a scale.data slot for the selected genes
alldata <- ScaleData(alldata, features = as.character(unique(top5$gene)), assay = "RNA")
DoHeatmap(alldata, features = as.character(unique(top5$gene)), group.by = sel.clust,
    assay = "RNA")
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-6-1.png)<!-- -->

Another way is by representing the overal group expression and detection rates in a dot-plot.


```r
DotPlot(alldata, features = rev(as.character(unique(top5$gene))), group.by = sel.clust,
    assay = "RNA") + coord_flip()
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-7-1.png)<!-- -->

We can also plot a violin plot for each gene.


```r
# take top 3 genes per cluster/
top5 %>%
    group_by(cluster) %>%
    top_n(-3, p_val) -> top3


# set pt.size to zero if you do not want all the points to hide the violin
# shapes, or to a small value like 0.1
VlnPlot(alldata, features = as.character(unique(top3$gene)), ncol = 5, group.by = sel.clust,
    assay = "RNA", pt.size = 0)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-8-1.png)<!-- -->

<style>
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 10px;}
</style>
<div class = "blue">
**Your turn**

Take a screen shot of those results and re-run the same code above with another test: "wilcox" (Wilcoxon Rank Sum test), "bimod" (Likelihood-ratio test), "roc" (Identifies 'markers' of gene expression using ROC analysis),"t" (Student's t-test),"negbinom" (negative binomial generalized linear model),"poisson" (poisson generalized linear model), "LR" (logistic regression), "MAST" (hurdle model), "DESeq2" (negative binomial distribution).
</div>

## Differential expression across conditions
***

The second way of computing differential expression is to answer which genes are differentially expressed within a cluster. For example, in our case we have libraries comming from patients and controls and we would like to know which genes are influenced the most in a particular cell type.

For this end, we will first subset our data for the desired cell cluster, then change the cell identities to the variable of comparison (which now in our case is the "type", e.g. Covid/Ctrl).


```r
# select all cells in cluster 1
cell_selection <- subset(alldata, cells = colnames(alldata)[alldata@meta.data[, sel.clust] ==
    2])
cell_selection <- SetIdent(cell_selection, value = "type")
# Compute differentiall expression
DGE_cell_selection <- FindAllMarkers(cell_selection, log2FC.threshold = 0.2, test.use = "wilcox",
    min.pct = 0.1, min.diff.pct = 0.2, only.pos = TRUE, max.cells.per.ident = 50,
    assay = "RNA")
```

We can now plot the expression across the "type".


```r
DGE_cell_selection %>%
    group_by(cluster) %>%
    top_n(-5, p_val) -> top5_cell_selection

VlnPlot(cell_selection, features = as.character(unique(top5_cell_selection$gene)),
    ncol = 5, group.by = "type", assay = "RNA", pt.size = 0.1)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-10-1.png)<!-- -->

We can also plot these genes across all clusters, but split by "type", to check if the genes are also up/downregulated in other celltypes.


```r
VlnPlot(alldata, features = as.character(unique(top5_cell_selection$gene)), ncol = 5,
    split.by = "type", assay = "RNA", pt.size = 0)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-11-1.png)<!-- -->

### Remove sex chromosome genes

To remove some of the bias due to inbalanced sex in the subjects we can remove the sex chromosome related genes.



```r
gene.info = read.csv("data/results/genes.table.csv")  #was created in the QC exercise

auto.genes = gene.info$external_gene_name[!(gene.info$chromosome_name %in% c("X",
    "Y"))]

cell_selection@active.assay = "RNA"
keep.genes = intersect(rownames(cell_selection), auto.genes)
cell_selection = cell_selection[keep.genes, ]

# then renormalize the data
cell_selection = NormalizeData(cell_selection)
```

Rerun differential expression:


```r
# Compute differentiall expression
DGE_cell_selection <- FindMarkers(cell_selection, ident.1 = "Covid", ident.2 = "Ctrl",
    logfc.threshold = 0.2, test.use = "wilcox", min.pct = 0.1, min.diff.pct = 0.2,
    assay = "RNA")

# Define as Covid or Ctrl in the df and add a gene column
DGE_cell_selection$direction = ifelse(DGE_cell_selection$avg_log2FC > 0, "Covid",
    "Ctrl")
DGE_cell_selection$gene = rownames(DGE_cell_selection)


DGE_cell_selection %>%
    group_by(direction) %>%
    top_n(-5, p_val) %>%
    arrange(direction) -> top5_cell_selection

VlnPlot(cell_selection, features = as.character(unique(top5_cell_selection$gene)),
    ncol = 5, group.by = "type", assay = "RNA", pt.size = 0.1)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-13-1.png)<!-- -->

We can also plot these genes across all clusters, but split by "type", to check if the genes are also up/downregulated in other celltypes/clusters.


```r
VlnPlot(alldata, features = as.character(unique(top5_cell_selection$gene)), ncol = 5,
    split.by = "type", assay = "RNA", pt.size = 0)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-14-1.png)<!-- -->




## Patient Batch effects

When we are testing for Covid vs Control we are running a DGE test for 3 vs 3 individuals. That will be very sensitive to sample differences unless we find a way to control for it. So first, lets check how the top DGEs are expressed across the individuals:


```r
VlnPlot(cell_selection, group.by = "orig.ident", features = as.character(unique(top5_cell_selection$gene)),
    ncol = 5, assay = "RNA", pt.size = 0)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-15-1.png)<!-- -->

As you can see, many of the genes detected as DGE in Covid are unique to one or 2 patients.

We can examine more genes with a DotPlot:


```r
DGE_cell_selection %>%
    group_by(direction) %>%
    top_n(-20, p_val) -> top20_cell_selection
DotPlot(cell_selection, features = rev(as.character(unique(top20_cell_selection$gene))),
    group.by = "orig.ident", assay = "RNA") + coord_flip()
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-16-1.png)<!-- -->

As you can see, most of the DGEs are driven by the `covid_17` patient.

But it is also the sample with the highest number of cells:


```r
table(cell_selection$orig.ident)
```

```
## 
##  covid_1 covid_15 covid_17  ctrl_13  ctrl_14   ctrl_5 
##       91       38      173       67       63      141
```

## Subsample

So one obvious thing to consider is an equal amount of cells per individual so that the DGE results are not dominated by a single sample.

So we will use the `downsample` option in the Seurat function `WhichCells` to select 30 cells per cluster:


```r
cell_selection <- SetIdent(cell_selection, value = "orig.ident")
sub_data <- subset(cell_selection, cells = WhichCells(cell_selection, downsample = 30))

table(sub_data$orig.ident)
```

```
## 
##  covid_1 covid_15 covid_17  ctrl_13  ctrl_14   ctrl_5 
##       30       30       30       30       30       30
```

And now we run DGE analysis again:


```r
sub_data <- SetIdent(sub_data, value = "type")

# Compute differentiall expression
DGE_sub <- FindMarkers(sub_data, ident.1 = "Covid", ident.2 = "Ctrl", logfc.threshold = 0.2,
    test.use = "wilcox", min.pct = 0.1, min.diff.pct = 0.2, assay = "RNA")

# Define as Covid or Ctrl in the df and add a gene column
DGE_sub$direction = ifelse(DGE_sub$avg_log2FC > 0, "Covid", "Ctrl")
DGE_sub$gene = rownames(DGE_sub)


DGE_sub %>%
    group_by(direction) %>%
    top_n(-5, p_val) %>%
    arrange(direction) -> top5_sub

VlnPlot(sub_data, features = as.character(unique(top5_sub$gene)), ncol = 5, group.by = "type",
    assay = "RNA", pt.size = 0.1)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-19-1.png)<!-- -->

Plot as dotplot, but in the full dataset:


```r
DGE_sub %>%
    group_by(direction) %>%
    top_n(-20, p_val) -> top20_sub
DotPlot(cell_selection, features = rev(as.character(unique(top20_sub$gene))), group.by = "orig.ident",
    assay = "RNA") + coord_flip()
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-20-1.png)<!-- -->

It looks much better now. But if we look per patient you can see that we still have some genes that are dominated by a single patient.


Why do you think this is?

## Pseudobulk

One option is to treat the samples as pseudobulks and do differential expression for the 3 patients vs 3 controls. You do lose some information about cell variability within each patient, but instead you gain the advantage of mainly looking for effects that are seen in multiple patients.

However, having only 3 patients is probably too low, with many more patients it will work better to run pseudobulk analysis.

For a fair comparison we should have equal number of cells per sample when we create the pseudobulk, so we will use the subsampled object.


```r
# get the count matrix for all cells
DGE_DATA <- sub_data@assays$RNA@counts

# Compute pseudobulk
mm <- Matrix::sparse.model.matrix(~0 + sub_data$orig.ident)
pseudobulk <- DGE_DATA %*% mm
```

Then run edgeR:


```r
# define the groups
bulk.labels = c("Covid", "Covid", "Covid", "Ctrl", "Ctrl", "Ctrl")

dge.list <- DGEList(counts = pseudobulk, group = factor(bulk.labels))
keep <- filterByExpr(dge.list)
dge.list <- dge.list[keep, , keep.lib.sizes = FALSE]

dge.list <- calcNormFactors(dge.list)
design = model.matrix(~bulk.labels)

dge.list <- estimateDisp(dge.list, design)

fit <- glmQLFit(dge.list, design)
qlf <- glmQLFTest(fit, coef = 2)
topTags(qlf)
```

```
## Coefficient:  bulk.labelsCtrl 
##            logFC    logCPM        F       PValue          FDR
## S100A9 -3.933901  8.028860 95.32177 3.084725e-08 5.179253e-05
## S100A8 -4.451499  7.477650 60.81560 6.590074e-07 5.532367e-04
## IGHA1  -3.477088  6.421735 22.63779 2.009392e-04 1.124590e-01
## CD69   -2.142911 10.825342 18.68613 4.992765e-04 2.095713e-01
## CCND3  -1.188752  7.869797 15.17994 1.234315e-03 4.144830e-01
## AREG   -1.870247  6.533761 13.97595 1.727960e-03 4.835407e-01
## RELB    1.490767  6.982100 11.81355 3.287908e-03 6.020757e-01
## AHNAK   1.089398  7.977803 11.79535 3.306522e-03 6.020757e-01
## KDM6B  -1.567158  7.677377 11.51378 3.610379e-03 6.020757e-01
## NFKBIA -1.223306 10.575741 11.10059 4.115171e-03 6.020757e-01
```

As you can see, we have very few significant genes, actually only 2 with FDR < 0.1. Since we only have 3 vs 3 samples, we should not expect too many genes with this method.

Again as dotplot including all genes with FDR < 1:


```r
res.edgeR <- topTags(qlf, 100)$table
res.edgeR$dir = ifelse(res.edgeR$logFC > 0, "Covid", "Ctrl")
res.edgeR$gene = rownames(res.edgeR)

res.edgeR %>%
    group_by(dir) %>%
    top_n(-10, PValue) %>%
    arrange(dir) -> top.edgeR



DotPlot(cell_selection, features = as.character(unique(top.edgeR$gene)), group.by = "orig.ident",
    assay = "RNA") + coord_flip() + ggtitle("EdgeR pseudobulk") + RotatedAxis()
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-23-1.png)<!-- -->

As you can see, even if we get few genes detected the seem to make sense across all the patients.

## MAST random effect

MAST has the option to add a random effect for the patient when running DGE analysis. It is quite slow, even with this small dataset, so it may not be practical for a larger dataset unless you have access to a compute cluster.

We will run MAST with and without patient info as random effect and compare the results

First, filter genes in part to speed up the process but also to avoid too many warnings in the model fitting step of MAST. We will keep genes that are expressed with at least 2 reads in 2 covid patients or 2 controls.



```r
# select genes that are expressed in at least 2 patients or 2 ctrls with > 2
# reads.
nPatient = sapply(unique(cell_selection$orig.ident), function(x) rowSums(cell_selection@assays$RNA@counts[,
    cell_selection$orig.ident == x] > 2))
nCovid = rowSums(nPatient[, 1:3] > 2)
nCtrl = rowSums(nPatient[, 4:6] > 2)

sel = nCovid >= 2 | nCtrl >= 2
cell_selection_sub = cell_selection[sel, ]
```

Set up the MAST object.


```r
# create the feature data
fData <- data.frame(primerid = rownames(cell_selection_sub))
m = cell_selection_sub@meta.data
m$wellKey = rownames(m)

# make sure type and orig.ident are factors
m$orig.ident = factor(m$orig.ident)
m$type = factor(m$type)

sca <- MAST::FromMatrix(exprsArray = as.matrix(x = cell_selection_sub@assays$RNA@data),
    check_sanity = FALSE, cData = m, fData = fData)
```

First, run the regular MAST analysis without random effects


```r
# takes a while to run, so save a file to tmpdir in case you have to rerun the
# code
tmpdir = "tmp_dge"
dir.create(tmpdir, showWarnings = F)

tmpfile1 = file.path(tmpdir, "mast_bayesglm_cl1.Rds")
if (file.exists(tmpfile1)) {
    fcHurdle1 = readRDS(tmpfile1)
} else {
    zlmCond <- suppressMessages(MAST::zlm(~type, sca, method = "bayesglm", ebayes = T))
    summaryCond <- suppressMessages(MAST::summary(zlmCond, doLRT = "typeCtrl"))
    summaryDt <- summaryCond$datatable
    fcHurdle <- merge(summaryDt[summaryDt$contrast == "typeCtrl" & summaryDt$component ==
        "logFC", c(1, 7, 5, 6, 8)], summaryDt[summaryDt$contrast == "typeCtrl" &
        summaryDt$component == "H", c(1, 4)], by = "primerid")
    fcHurdle1 <- stats::na.omit(as.data.frame(fcHurdle))
    saveRDS(fcHurdle1, tmpfile1)
}
```

Then run MAST with glmer and random effect.


```r
library(lme4)

tmpfile2 = file.path(tmpdir, "mast_glme_cl1.Rds")
if (file.exists(tmpfile2)) {
    fcHurdle2 = readRDS(tmpfile2)
} else {
    zlmCond <- suppressMessages(MAST::zlm(~type + (1 | orig.ident), sca, method = "glmer",
        ebayes = F, strictConvergence = FALSE))

    summaryCond <- suppressMessages(MAST::summary(zlmCond, doLRT = "typeCtrl"))
    summaryDt <- summaryCond$datatable
    fcHurdle <- merge(summaryDt[summaryDt$contrast == "typeCtrl" & summaryDt$component ==
        "logFC", c(1, 7, 5, 6, 8)], summaryDt[summaryDt$contrast == "typeCtrl" &
        summaryDt$component == "H", c(1, 4)], by = "primerid")
    fcHurdle2 <- stats::na.omit(as.data.frame(fcHurdle))
    saveRDS(fcHurdle2, tmpfile2)
}
```

Top genes with normal MAST:


```r
top1 = head(fcHurdle1[order(fcHurdle1$`Pr(>Chisq)`), ], 10)
top1
```

<div data-pagedtable="false">
  <script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["primerid"],"name":[1],"type":["chr"],"align":["left"]},{"label":["coef"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["ci.hi"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["ci.lo"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["z"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["Pr(>Chisq)"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"HLA-DRB5","2":"-0.7404644","3":"-0.5648164","4":"-0.9161124","5":"-8.262453","6":"8.208167e-50","_rn_":"362"},{"1":"RPS26","2":"1.2170354","3":"1.3749446","4":"1.0591262","5":"15.105805","6":"9.707574e-48","_rn_":"728"},{"1":"CD74","2":"0.4991497","3":"0.5605379","4":"0.4377615","5":"15.936531","6":"2.385815e-45","_rn_":"145"},{"1":"CD69","2":"-1.2992695","3":"-1.1123603","4":"-1.4861788","5":"-13.624374","6":"5.872770e-42","_rn_":"143"},{"1":"ISG20","2":"-0.8202045","3":"-0.6723129","4":"-0.9680962","5":"-10.869928","6":"2.888473e-30","_rn_":"415"},{"1":"NFKBIA","2":"-0.7205167","3":"-0.5725089","4":"-0.8685245","5":"-9.541301","6":"3.122769e-29","_rn_":"512"},{"1":"LY6E","2":"-0.7199797","3":"-0.5834347","4":"-0.8565246","5":"-10.334575","6":"1.428674e-26","_rn_":"454"},{"1":"EEF1A1","2":"0.4437284","3":"0.5194865","4":"0.3679702","5":"11.479844","6":"7.609188e-26","_rn_":"246"},{"1":"RPL4","2":"0.6255466","3":"0.7427171","4":"0.5083761","5":"10.463802","6":"2.163788e-25","_rn_":"700"},{"1":"S100A9","2":"-0.5905734","3":"-0.4786486","4":"-0.7024981","5":"-10.341793","6":"3.689851e-23","_rn_":"751"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>

```r
fcHurdle1$pval = fcHurdle1$`Pr(>Chisq)`
fcHurdle1$dir = ifelse(fcHurdle1$z > 0, "up", "down")
fcHurdle1 %>%
    group_by(dir) %>%
    top_n(-10, pval) %>%
    arrange(z) -> mastN

mastN = mastN$primerid
```

Top genes with random effect:


```r
top2 = head(fcHurdle2[order(fcHurdle2$`Pr(>Chisq)`), ], 10)
top2
```

<div data-pagedtable="false">
  <script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["primerid"],"name":[1],"type":["chr"],"align":["left"]},{"label":["coef"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["ci.hi"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["ci.lo"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["z"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["Pr(>Chisq)"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"HLA-DRA","2":"0.3684288","3":"0.4639804","4":"0.2728771","5":"7.557244","6":"6.067723e-05","_rn_":"360"},{"1":"S100A9","2":"-0.5936154","3":"-0.4807224","4":"-0.7065084","5":"-10.305905","6":"6.583353e-05","_rn_":"751"},{"1":"HLA-DPB1","2":"0.3614619","3":"0.4667714","4":"0.2561524","5":"6.727336","6":"8.568929e-05","_rn_":"357"},{"1":"S100A8","2":"-0.5187687","3":"-0.3497358","4":"-0.6878015","5":"-6.015209","6":"1.003802e-04","_rn_":"750"},{"1":"SF1","2":"0.5228481","3":"0.7760301","4":"0.2696660","5":"4.047536","6":"1.757736e-04","_rn_":"768"},{"1":"PHACTR1","2":"-0.4871676","3":"-0.3518684","4":"-0.6224668","5":"-7.057183","6":"1.763375e-04","_rn_":"561"},{"1":"PCBP2","2":"0.2890742","3":"0.4130207","4":"0.1651276","5":"4.571123","6":"2.553222e-04","_rn_":"548"},{"1":"JCHAIN","2":"0.2787417","3":"0.3950091","4":"0.1624743","5":"4.698855","6":"4.164025e-04","_rn_":"421"},{"1":"ZFAS1","2":"0.2795446","3":"0.4116192","4":"0.1474699","5":"4.148390","6":"8.384734e-04","_rn_":"964"},{"1":"HLA-DPA1","2":"0.2913176","3":"0.4160733","4":"0.1665618","5":"4.576719","6":"9.761835e-04","_rn_":"356"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>

```r
fcHurdle2$pval = fcHurdle2$`Pr(>Chisq)`
fcHurdle2$dir = ifelse(fcHurdle2$z > 0, "up", "down")
fcHurdle2 %>%
    group_by(dir) %>%
    top_n(-10, pval) %>%
    arrange(z) -> mastR

mastR = mastR$primerid
```

As you can see, we have lower significance for the genes with the random effect added.

Dotplot for top10 genes in each direction:


```r
p1 = DotPlot(cell_selection, features = mastN, group.by = "orig.ident", assay = "RNA") +
    coord_flip() + RotatedAxis() + ggtitle("Regular MAST")

p2 = DotPlot(cell_selection, features = mastR, group.by = "orig.ident", assay = "RNA") +
    coord_flip() + RotatedAxis() + ggtitle("With random effect")


p1 + p2
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-30-1.png)<!-- -->




## Gene Set Analysis
***

Hypergeometric enrichment test

Having a defined list of differentially expressed genes, you can now look for their combined function using hypergeometric test:


```r
# Load additional packages
library(enrichR)

# Check available databases to perform enrichment (then choose one)
enrichR::listEnrichrDbs()
```

<div data-pagedtable="false">
  <script data-pagedtable-source type="application/json">
{"columns":[{"label":["geneCoverage"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["genesPerTerm"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["libraryName"],"name":[3],"type":["chr"],"align":["left"]},{"label":["link"],"name":[4],"type":["chr"],"align":["left"]},{"label":["numTerms"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["appyter"],"name":[6],"type":["chr"],"align":["left"]},{"label":["categoryId"],"name":[7],"type":["dbl"],"align":["right"]}],"data":[{"1":"13362","2":"275","3":"Genome_Browser_PWMs","4":"http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/","5":"615","6":"ea115789fcbf12797fd692cec6df0ab4dbc79c6a","7":"1"},{"1":"27884","2":"1284","3":"TRANSFAC_and_JASPAR_PWMs","4":"http://jaspar.genereg.net/html/DOWNLOAD/","5":"326","6":"7d42eb43a64a4e3b20d721fc7148f685b53b6b30","7":"1"},{"1":"6002","2":"77","3":"Transcription_Factor_PPIs","4":"","5":"290","6":"849f222220618e2599d925b6b51868cf1dab3763","7":"1"},{"1":"47172","2":"1370","3":"ChEA_2013","4":"http://amp.pharm.mssm.edu/lib/cheadownload.jsp","5":"353","6":"7ebe772afb55b63b41b79dd8d06ea0fdd9fa2630","7":"7"},{"1":"47107","2":"509","3":"Drug_Perturbations_from_GEO_2014","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"701","6":"ad270a6876534b7cb063e004289dcd4d3164f342","7":"7"},{"1":"21493","2":"3713","3":"ENCODE_TF_ChIP-seq_2014","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"498","6":"497787ebc418d308045efb63b8586f10c526af51","7":"7"},{"1":"1295","2":"18","3":"BioCarta_2013","4":"https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","5":"249","6":"4a293326037a5229aedb1ad7b2867283573d8bcd","7":"7"},{"1":"3185","2":"73","3":"Reactome_2013","4":"http://www.reactome.org/download/index.html","5":"78","6":"b343994a1b68483b0122b08650201c9b313d5c66","7":"7"},{"1":"2854","2":"34","3":"WikiPathways_2013","4":"http://www.wikipathways.org/index.php/Download_Pathways","5":"199","6":"5c307674c8b97e098f8399c92f451c0ff21cbf68","7":"7"},{"1":"15057","2":"300","3":"Disease_Signatures_from_GEO_up_2014","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"142","6":"248c4ed8ea28352795190214713c86a39fd7afab","7":"7"},{"1":"4128","2":"48","3":"KEGG_2013","4":"http://www.kegg.jp/kegg/download/","5":"200","6":"eb26f55d3904cb0ea471998b6a932a9bf65d8e50","7":"7"},{"1":"34061","2":"641","3":"TF-LOF_Expression_from_GEO","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"269","6":"","7":"1"},{"1":"7504","2":"155","3":"TargetScan_microRNA","4":"http://www.targetscan.org/cgi-bin/targetscan/data_download.cgi?db=vert_61","5":"222","6":"f4029bf6a62c91ab29401348e51df23b8c44c90f","7":"7"},{"1":"16399","2":"247","3":"PPI_Hub_Proteins","4":"http://amp.pharm.mssm.edu/X2K","5":"385","6":"69c0cfe07d86f230a7ef01b365abcc7f6e52f138","7":"2"},{"1":"12753","2":"57","3":"GO_Molecular_Function_2015","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"1136","6":"f531ac2b6acdf7587a54b79b465a5f4aab8f00f9","7":"7"},{"1":"23726","2":"127","3":"GeneSigDB","4":"https://pubmed.ncbi.nlm.nih.gov/22110038/","5":"2139","6":"6d655e0aa3408a7accb3311fbda9b108681a8486","7":"4"},{"1":"32740","2":"85","3":"Chromosome_Location","4":"http://software.broadinstitute.org/gsea/msigdb/index.jsp","5":"386","6":"8dab0f96078977223646ff63eb6187e0813f1433","7":"7"},{"1":"13373","2":"258","3":"Human_Gene_Atlas","4":"http://biogps.org/downloads/","5":"84","6":"0741451470203d7c40a06274442f25f74b345c9c","7":"5"},{"1":"19270","2":"388","3":"Mouse_Gene_Atlas","4":"http://biogps.org/downloads/","5":"96","6":"31191bfadded5f96983f93b2a113cf2110ff5ddb","7":"5"},{"1":"13236","2":"82","3":"GO_Cellular_Component_2015","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"641","6":"e1d004d5797cbd2363ef54b1c3b361adb68795c6","7":"7"},{"1":"14264","2":"58","3":"GO_Biological_Process_2015","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"5192","6":"bf120b6e11242b1a64c80910d8e89f87e618e235","7":"7"},{"1":"3096","2":"31","3":"Human_Phenotype_Ontology","4":"http://www.human-phenotype-ontology.org/","5":"1779","6":"17a138b0b70aa0e143fe63c14f82afb70bc3ed0a","7":"3"},{"1":"22288","2":"4368","3":"Epigenomics_Roadmap_HM_ChIP-seq","4":"http://www.roadmapepigenomics.org/","5":"383","6":"e1bc8a398e9b21f9675fb11bef18087eda21b1bf","7":"1"},{"1":"4533","2":"37","3":"KEA_2013","4":"http://amp.pharm.mssm.edu/lib/keacommandline.jsp","5":"474","6":"462045609440fa1e628a75716b81a1baa5bd9145","7":"7"},{"1":"10231","2":"158","3":"NURSA_Human_Endogenous_Complexome","4":"https://www.nursa.org/nursa/index.jsf","5":"1796","6":"7d3566b12ebc23dd23d9ca9bb97650f826377b16","7":"2"},{"1":"2741","2":"5","3":"CORUM","4":"http://mips.helmholtz-muenchen.de/genre/proj/corum/","5":"1658","6":"d047f6ead7831b00566d5da7a3b027ed9196e104","7":"2"},{"1":"5655","2":"342","3":"SILAC_Phosphoproteomics","4":"http://amp.pharm.mssm.edu/lib/keacommandline.jsp","5":"84","6":"54dcd9438b33301deb219866e162b0f9da7e63a0","7":"2"},{"1":"10406","2":"715","3":"MGI_Mammalian_Phenotype_Level_3","4":"http://www.informatics.jax.org/","5":"71","6":"c3bfc90796cfca8f60cba830642a728e23a53565","7":"7"},{"1":"10493","2":"200","3":"MGI_Mammalian_Phenotype_Level_4","4":"http://www.informatics.jax.org/","5":"476","6":"0b09a9a1aa0af4fc7ea22d34a9ae644d45864bd6","7":"7"},{"1":"11251","2":"100","3":"Old_CMAP_up","4":"http://www.broadinstitute.org/cmap/","5":"6100","6":"9041f90cccbc18479138330228b336265e09021c","7":"7"},{"1":"8695","2":"100","3":"Old_CMAP_down","4":"http://www.broadinstitute.org/cmap/","5":"6100","6":"ebc0d905b3b3142f936d400c5f2a4ff926c81c37","7":"7"},{"1":"1759","2":"25","3":"OMIM_Disease","4":"http://www.omim.org/downloads","5":"90","6":"cb2b92578a91e023d0498a334923ee84add34eca","7":"4"},{"1":"2178","2":"89","3":"OMIM_Expanded","4":"http://www.omim.org/downloads","5":"187","6":"27eca242904d8e12a38cf8881395bc50d57a03e1","7":"4"},{"1":"851","2":"15","3":"VirusMINT","4":"http://mint.bio.uniroma2.it/download.html","5":"85","6":"5abad1fc36216222b0420cadcd9be805a0dda63e","7":"4"},{"1":"10061","2":"106","3":"MSigDB_Computational","4":"http://www.broadinstitute.org/gsea/msigdb/collections.jsp","5":"858","6":"e4cdcc7e259788fdf9b25586cce3403255637064","7":"4"},{"1":"11250","2":"166","3":"MSigDB_Oncogenic_Signatures","4":"http://www.broadinstitute.org/gsea/msigdb/collections.jsp","5":"189","6":"c76f5319c33c4833c71db86a30d7e33cd63ff8cf","7":"4"},{"1":"15406","2":"300","3":"Disease_Signatures_from_GEO_down_2014","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"142","6":"aabdf7017ae55ae75a004270924bcd336653b986","7":"7"},{"1":"17711","2":"300","3":"Virus_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"323","6":"45268b7fc680d05dd9a29743c2f2b2840a7620bf","7":"4"},{"1":"17576","2":"300","3":"Virus_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"323","6":"5f531580ccd168ee4acc18b02c6bdf8200e19d08","7":"4"},{"1":"15797","2":"176","3":"Cancer_Cell_Line_Encyclopedia","4":"https://portals.broadinstitute.org/ccle/home\\n","5":"967","6":"eb38dbc3fb20adafa9d6f9f0b0e36f378e75284f","7":"5"},{"1":"12232","2":"343","3":"NCI-60_Cancer_Cell_Lines","4":"http://biogps.org/downloads/","5":"93","6":"75c81676d8d6d99d262c9660edc024b78cfb07c9","7":"5"},{"1":"13572","2":"301","3":"Tissue_Protein_Expression_from_ProteomicsDB","4":"https://www.proteomicsdb.org/","5":"207","6":"","7":"7"},{"1":"6454","2":"301","3":"Tissue_Protein_Expression_from_Human_Proteome_Map","4":"http://www.humanproteomemap.org/index.php","5":"30","6":"49351dc989f9e6ca97c55f8aca7778aa3bfb84b9","7":"5"},{"1":"3723","2":"47","3":"HMDB_Metabolites","4":"http://www.hmdb.ca/downloads","5":"3906","6":"1905132115d22e4119bce543bdacaab074edb363","7":"6"},{"1":"7588","2":"35","3":"Pfam_InterPro_Domains","4":"ftp://ftp.ebi.ac.uk/pub/databases/interpro/","5":"311","6":"e2b4912cfb799b70d87977808c54501544e4cdc9","7":"6"},{"1":"7682","2":"78","3":"GO_Biological_Process_2013","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"941","6":"5216d1ade194ffa5a6c00f105e2b1899f64f45fe","7":"7"},{"1":"7324","2":"172","3":"GO_Cellular_Component_2013","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"205","6":"fd1332a42395e0bc1dba82868b39be7983a48cc5","7":"7"},{"1":"8469","2":"122","3":"GO_Molecular_Function_2013","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"402","6":"7e3e99e5aae02437f80b0697b197113ce3209ab0","7":"7"},{"1":"13121","2":"305","3":"Allen_Brain_Atlas_up","4":"http://www.brain-map.org/","5":"2192","6":"3804715a63a308570e47aa1a7877f01147ca6202","7":"5"},{"1":"26382","2":"1811","3":"ENCODE_TF_ChIP-seq_2015","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"816","6":"56b6adb4dc8a2f540357ef992d6cd93dfa2907e5","7":"1"},{"1":"29065","2":"2123","3":"ENCODE_Histone_Modifications_2015","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"412","6":"55b56cd8cf2ff04b26a09b9f92904008b82f3a6f","7":"1"},{"1":"280","2":"9","3":"Phosphatase_Substrates_from_DEPOD","4":"http://www.koehn.embl.de/depod/","5":"59","6":"d40701e21092b999f4720d1d2b644dd0257b6259","7":"2"},{"1":"13877","2":"304","3":"Allen_Brain_Atlas_down","4":"http://www.brain-map.org/","5":"2192","6":"ea67371adec290599ddf484ced2658cfae259304","7":"5"},{"1":"15852","2":"912","3":"ENCODE_Histone_Modifications_2013","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"109","6":"c209ae527bc8e98e4ccd27a668d36cd2c80b35b4","7":"7"},{"1":"4320","2":"129","3":"Achilles_fitness_increase","4":"http://www.broadinstitute.org/achilles","5":"216","6":"98366496a75f163164106e72439fb2bf2f77de4e","7":"4"},{"1":"4271","2":"128","3":"Achilles_fitness_decrease","4":"http://www.broadinstitute.org/achilles","5":"216","6":"83a710c1ff67fd6b8af0d80fa6148c40dbd9bc64","7":"4"},{"1":"10496","2":"201","3":"MGI_Mammalian_Phenotype_2013","4":"http://www.informatics.jax.org/","5":"476","6":"a4c6e217a81a4a58ff5a1c9fc102b70beab298e9","7":"7"},{"1":"1678","2":"21","3":"BioCarta_2015","4":"https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","5":"239","6":"70e4eb538daa7688691acfe5d9c3c19022be832b","7":"7"},{"1":"756","2":"12","3":"HumanCyc_2015","4":"http://humancyc.org/","5":"125","6":"711f0c02b23f5e02a01207174943cfeee9d3ea9c","7":"7"},{"1":"3800","2":"48","3":"KEGG_2015","4":"http://www.kegg.jp/kegg/download/","5":"179","6":"e80d25c56de53c704791ddfdc6ab5eec28ae7243","7":"7"},{"1":"2541","2":"39","3":"NCI-Nature_2015","4":"http://pid.nci.nih.gov/","5":"209","6":"47edfc012bcbb368a10b717d8dca103f7814b5a4","7":"7"},{"1":"1918","2":"39","3":"Panther_2015","4":"http://www.pantherdb.org/","5":"104","6":"ab824aeeff0712bab61f372e43aebb870d1677a9","7":"7"},{"1":"5863","2":"51","3":"WikiPathways_2015","4":"http://www.wikipathways.org/index.php/Download_Pathways","5":"404","6":"1f7eea2f339f37856522c1f1c70ec74c7b25325f","7":"7"},{"1":"6768","2":"47","3":"Reactome_2015","4":"http://www.reactome.org/download/index.html","5":"1389","6":"36e541bee015eddb8d53827579549e30fe7a3286","7":"7"},{"1":"25651","2":"807","3":"ESCAPE","4":"http://www.maayanlab.net/ESCAPE/","5":"315","6":"a7acc741440264717ff77751a7e5fed723307835","7":"5"},{"1":"19129","2":"1594","3":"HomoloGene","4":"http://www.ncbi.nlm.nih.gov/homologene","5":"12","6":"663b665b75a804ef98add689f838b68e612f0d2a","7":"6"},{"1":"23939","2":"293","3":"Disease_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"839","6":"0f412e0802d76efa0374504c2c9f5e0624ff7f09","7":"8"},{"1":"23561","2":"307","3":"Disease_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"839","6":"9ddc3902fb01fb9eaf1a2a7c2ff3acacbb48d37e","7":"8"},{"1":"23877","2":"302","3":"Drug_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"906","6":"068623a05ecef3e4a5e0b4f8db64bb8faa3c897f","7":"8"},{"1":"15886","2":"9","3":"Genes_Associated_with_NIH_Grants","4":"https://grants.nih.gov/grants/oer.htm\\n","5":"32876","6":"76fc5ec6735130e287e62bae6770a3c5ee068645","7":"6"},{"1":"24350","2":"299","3":"Drug_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"906","6":"c9c2155b5ac81ac496854fa61ba566dcae06cc80","7":"8"},{"1":"3102","2":"25","3":"KEA_2015","4":"http://amp.pharm.mssm.edu/Enrichr","5":"428","6":"18a081774e6e0aaf60b1a4be7fd20afcf9e08399","7":"2"},{"1":"31132","2":"298","3":"Gene_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"2460","6":"53dedc29ce3100930d68e506f941ef59de05dc6b","7":"8"},{"1":"30832","2":"302","3":"Gene_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"2460","6":"499882af09c62dd6da545c15cb51c1dc5e234f78","7":"8"},{"1":"48230","2":"1429","3":"ChEA_2015","4":"http://amp.pharm.mssm.edu/Enrichr","5":"395","6":"712eb7b6edab04658df153605ec6079fa89fb5c7","7":"7"},{"1":"5613","2":"36","3":"dbGaP","4":"http://www.ncbi.nlm.nih.gov/gap","5":"345","6":"010f1267055b1a1cb036e560ea525911c007a666","7":"4"},{"1":"9559","2":"73","3":"LINCS_L1000_Chem_Pert_up","4":"https://clue.io/","5":"33132","6":"5e678b3debe8d8ea95187d0cd35c914017af5eb3","7":"7"},{"1":"9448","2":"63","3":"LINCS_L1000_Chem_Pert_down","4":"https://clue.io/","5":"33132","6":"fedbf5e221f45ee60ebd944f92569b5eda7f2330","7":"7"},{"1":"16725","2":"1443","3":"GTEx_Tissue_Expression_Down","4":"http://www.gtexportal.org/","5":"2918","6":"74b818bd299a9c42c1750ffe43616aa9f7929f02","7":"5"},{"1":"19249","2":"1443","3":"GTEx_Tissue_Expression_Up","4":"http://www.gtexportal.org/","5":"2918","6":"103738763d89cae894bec9f145ac28167a90e611","7":"5"},{"1":"15090","2":"282","3":"Ligand_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"261","6":"1eb3c0426140340527155fd0ef67029db2a72191","7":"8"},{"1":"16129","2":"292","3":"Aging_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"286","6":"cd95fe1b505ba6f28cd722cfba50fdea979d3b4c","7":"8"},{"1":"15309","2":"308","3":"Aging_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"286","6":"74c4f0a0447777005b2a5c00c9882a56dfc62d7c","7":"8"},{"1":"15103","2":"318","3":"Ligand_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"261","6":"31baa39da2931ddd5f7aedf2d0bbba77d2ba7b46","7":"8"},{"1":"15022","2":"290","3":"MCF7_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"401","6":"555f68aef0a29a67b614a0d7e20b6303df9069c6","7":"8"},{"1":"15676","2":"310","3":"MCF7_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"401","6":"1bc2ba607f1ff0dda44e2a15f32a2c04767da18c","7":"8"},{"1":"15854","2":"279","3":"Microbe_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"312","6":"9e613dba78ef7e60676b13493a9dc49ccd3c8b3f","7":"8"},{"1":"15015","2":"321","3":"Microbe_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"312","6":"d0c3e2a68e8c611c669098df2c87b530cec3e132","7":"8"},{"1":"3788","2":"159","3":"LINCS_L1000_Ligand_Perturbations_down","4":"https://clue.io/","5":"96","6":"957846cb05ef31fc8514120516b73cc65af7980e","7":"7"},{"1":"3357","2":"153","3":"LINCS_L1000_Ligand_Perturbations_up","4":"https://clue.io/","5":"96","6":"3bd494146c98d8189898a947f5ef5710f1b7c4b2","7":"7"},{"1":"12668","2":"300","3":"L1000_Kinase_and_GPCR_Perturbations_down","4":"https://clue.io/","5":"3644","6":"1ccc5bce553e0c2279f8e3f4ddcfbabcf566623b","7":"7"},{"1":"12638","2":"300","3":"L1000_Kinase_and_GPCR_Perturbations_up","4":"https://clue.io/","5":"3644","6":"b54a0d4ba525eac4055c7314ca9d9312adcb220c","7":"7"},{"1":"8973","2":"64","3":"Reactome_2016","4":"http://www.reactome.org/download/index.html","5":"1530","6":"1f54638e8f45075fb79489f0e0ef906594cb0678","7":"7"},{"1":"7010","2":"87","3":"KEGG_2016","4":"http://www.kegg.jp/kegg/download/","5":"293","6":"43f56da7540195ba3c94eb6e34c522a699b36da9","7":"7"},{"1":"5966","2":"51","3":"WikiPathways_2016","4":"http://www.wikipathways.org/index.php/Download_Pathways","5":"437","6":"340be98b444cad50bb974df69018fd598e23e5e1","7":"7"},{"1":"15562","2":"887","3":"ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X","4":"","5":"104","6":"5426f7747965c23ef32cff46fabf906e2cd76bfa","7":"1"},{"1":"17850","2":"300","3":"Kinase_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"285","6":"bb9682d78b8fc43be842455e076166fcd02cefc3","7":"2"},{"1":"17660","2":"300","3":"Kinase_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"285","6":"78618915009cac3a0663d6f99d359e39a31b6660","7":"2"},{"1":"1348","2":"19","3":"BioCarta_2016","4":"http://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","5":"237","6":"13d9ab18921d5314a5b2b366f6142b78ab0ff6aa","7":"2"},{"1":"934","2":"13","3":"HumanCyc_2016","4":"http://humancyc.org/","5":"152","6":"d6a502ef9b4c789ed5e73ca5a8de372796e5c72a","7":"2"},{"1":"2541","2":"39","3":"NCI-Nature_2016","4":"http://pid.nci.nih.gov/","5":"209","6":"3c1e1f7d1a651d9aaa198e73704030716fc09431","7":"2"},{"1":"2041","2":"42","3":"Panther_2016","4":"http://www.pantherdb.org/pathway/","5":"112","6":"ca5f6abf7f75d9baae03396e84d07300bf1fd051","7":"2"},{"1":"5209","2":"300","3":"DrugMatrix","4":"https://ntp.niehs.nih.gov/drugmatrix/","5":"7876","6":"255c3db820d612f34310f22a6985dad50e9fe1fe","7":"4"},{"1":"49238","2":"1550","3":"ChEA_2016","4":"http://amp.pharm.mssm.edu/Enrichr","5":"645","6":"af271913344aa08e6a755af1d433ef15768d749a","7":"7"},{"1":"2243","2":"19","3":"huMAP","4":"http://proteincomplexes.org/","5":"995","6":"249247d2f686d3eb4b9e4eb976c51159fac80a89","7":"2"},{"1":"19586","2":"545","3":"Jensen_TISSUES","4":"http://tissues.jensenlab.org/","5":"1842","6":"e8879ab9534794721614d78fe2883e9e564d7759","7":"3"},{"1":"22440","2":"505","3":"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"1302","6":"f0752e4d7f5198f86446678966b260c530d19d78","7":"8"},{"1":"8184","2":"24","3":"MGI_Mammalian_Phenotype_2017","4":"http://www.informatics.jax.org/","5":"5231","6":"0705e59bff98deda6e9cbe00cfcdd871c85e7d04","7":"7"},{"1":"18329","2":"161","3":"Jensen_COMPARTMENTS","4":"http://compartments.jensenlab.org/","5":"2283","6":"56ec68c32d4e83edc2ee83bea0e9f6a3829b2279","7":"3"},{"1":"15755","2":"28","3":"Jensen_DISEASES","4":"http://diseases.jensenlab.org/","5":"1811","6":"3045dff8181367c1421627bb8e4c5a32c6d67f98","7":"3"},{"1":"10271","2":"22","3":"BioPlex_2017","4":"http://bioplex.hms.harvard.edu/","5":"3915","6":"b8620b1a9d0d271d1a2747d8cfc63589dba39991","7":"2"},{"1":"10427","2":"38","3":"GO_Cellular_Component_2017","4":"http://www.geneontology.org/","5":"636","6":"8fed21d22dfcc3015c05b31d942fdfc851cc8e04","7":"7"},{"1":"10601","2":"25","3":"GO_Molecular_Function_2017","4":"http://www.geneontology.org/","5":"972","6":"b4018906e0a8b4e81a1b1afc51e0a2e7655403eb","7":"7"},{"1":"13822","2":"21","3":"GO_Biological_Process_2017","4":"http://www.geneontology.org/","5":"3166","6":"d9da4dba4a3eb84d4a28a3835c06dfbbe5811f92","7":"7"},{"1":"8002","2":"143","3":"GO_Cellular_Component_2017b","4":"http://www.geneontology.org/","5":"816","6":"ecf39c41fa5bc7deb625a2b5761a708676e9db7c","7":"7"},{"1":"10089","2":"45","3":"GO_Molecular_Function_2017b","4":"http://www.geneontology.org/","5":"3271","6":"8d8340361dd36a458f1f0a401f1a3141de1f3200","7":"7"},{"1":"13247","2":"49","3":"GO_Biological_Process_2017b","4":"http://www.geneontology.org/","5":"10125","6":"6404c38bffc2b3732de4e3fbe417b5043009fe34","7":"7"},{"1":"21809","2":"2316","3":"ARCHS4_Tissues","4":"http://amp.pharm.mssm.edu/archs4","5":"108","6":"4126374338235650ab158ba2c61cd2e2383b70df","7":"5"},{"1":"23601","2":"2395","3":"ARCHS4_Cell-lines","4":"http://amp.pharm.mssm.edu/archs4","5":"125","6":"5496ef9c9ae9429184d0b9485c23ba468ee522a8","7":"5"},{"1":"20883","2":"299","3":"ARCHS4_IDG_Coexp","4":"http://amp.pharm.mssm.edu/archs4","5":"352","6":"ce60be284fdd5a9fc6240a355421a9e12b1ee84a","7":"4"},{"1":"19612","2":"299","3":"ARCHS4_Kinases_Coexp","4":"http://amp.pharm.mssm.edu/archs4","5":"498","6":"6721c5ed97b7772e4a19fdc3f797110df0164b75","7":"2"},{"1":"25983","2":"299","3":"ARCHS4_TFs_Coexp","4":"http://amp.pharm.mssm.edu/archs4","5":"1724","6":"8a468c3ae29fa68724f744cbef018f4f3b61c5ab","7":"1"},{"1":"19500","2":"137","3":"SysMyo_Muscle_Gene_Sets","4":"http://sys-myo.rhcloud.com/","5":"1135","6":"","7":"8"},{"1":"14893","2":"128","3":"miRTarBase_2017","4":"http://mirtarbase.mbc.nctu.edu.tw/","5":"3240","6":"6b7c7fe2a97b19aecbfba12d8644af6875ad99c4","7":"1"},{"1":"17598","2":"1208","3":"TargetScan_microRNA_2017","4":"http://www.targetscan.org/","5":"683","6":"79d13fb03d2fa6403f9be45c90eeda0f6822e269","7":"1"},{"1":"5902","2":"109","3":"Enrichr_Libraries_Most_Popular_Genes","4":"http://amp.pharm.mssm.edu/Enrichr","5":"121","6":"e9b7d8ee237d0a690bd79d970a23a9fa849901ed","7":"6"},{"1":"12486","2":"299","3":"Enrichr_Submissions_TF-Gene_Coocurrence","4":"http://amp.pharm.mssm.edu/Enrichr","5":"1722","6":"be2ca8ef5a8c8e17d7e7bd290e7cbfe0951396c0","7":"1"},{"1":"1073","2":"100","3":"Data_Acquisition_Method_Most_Popular_Genes","4":"http://amp.pharm.mssm.edu/Enrichr","5":"12","6":"17ce5192b9eba7d109b6d228772ea8ab222e01ef","7":"6"},{"1":"19513","2":"117","3":"DSigDB","4":"http://tanlab.ucdenver.edu/DSigDB/DSigDBv1.0/","5":"4026","6":"287476538ab98337dbe727b3985a436feb6d192a","7":"4"},{"1":"14433","2":"36","3":"GO_Biological_Process_2018","4":"http://www.geneontology.org/","5":"5103","6":"b5b77681c46ac58cd050e60bcd4ad5041a9ab0a9","7":"7"},{"1":"8655","2":"61","3":"GO_Cellular_Component_2018","4":"http://www.geneontology.org/","5":"446","6":"e9ebe46188efacbe1056d82987ff1c70218fa7ae","7":"7"},{"1":"11459","2":"39","3":"GO_Molecular_Function_2018","4":"http://www.geneontology.org/","5":"1151","6":"79ff80ae9a69dd00796e52569e41422466fa0bee","7":"7"},{"1":"19741","2":"270","3":"TF_Perturbations_Followed_by_Expression","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"1958","6":"34d08a4878c19584aaf180377f2ea96faa6a6eb1","7":"1"},{"1":"27360","2":"802","3":"Chromosome_Location_hg19","4":"http://hgdownload.cse.ucsc.edu/downloads.html","5":"36","6":"fdab39c467ba6b0fb0288df1176d7dfddd7196d5","7":"6"},{"1":"13072","2":"26","3":"NIH_Funded_PIs_2017_Human_GeneRIF","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"5687","6":"859b100fac3ca774ad84450b1fbb65a78fcc6b12","7":"6"},{"1":"13464","2":"45","3":"NIH_Funded_PIs_2017_Human_AutoRIF","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"12558","6":"fc5bf033b932cf173633e783fc8c6228114211f8","7":"6"},{"1":"13787","2":"200","3":"Rare_Diseases_AutoRIF_ARCHS4_Predictions","4":"https://amp.pharm.mssm.edu/geneshot/","5":"3725","6":"375ff8cdd64275a916fa24707a67968a910329bb","7":"4"},{"1":"13929","2":"200","3":"Rare_Diseases_GeneRIF_ARCHS4_Predictions","4":"https://www.ncbi.nlm.nih.gov/gene/about-generif","5":"2244","6":"0f7fb7f347534779ecc6c87498e96b5460a8d652","7":"4"},{"1":"16964","2":"200","3":"NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"12558","6":"f77de51aaf0979dd6f56381cf67ba399b4640d28","7":"6"},{"1":"17258","2":"200","3":"NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"5684","6":"25fa899b715cd6a9137f6656499f89cd25144029","7":"6"},{"1":"10352","2":"58","3":"Rare_Diseases_GeneRIF_Gene_Lists","4":"https://www.ncbi.nlm.nih.gov/gene/about-generif","5":"2244","6":"0fb9ac92dbe52024661c088f71a1134f00567a8b","7":"4"},{"1":"10471","2":"76","3":"Rare_Diseases_AutoRIF_Gene_Lists","4":"https://amp.pharm.mssm.edu/geneshot/","5":"3725","6":"ee3adbac2da389959410260b280e7df1fd3730df","7":"4"},{"1":"12419","2":"491","3":"SubCell_BarCode","4":"http://www.subcellbarcode.org/","5":"104","6":"b50bb9480d8a77103fb75b331fd9dd927246939a","7":"2"},{"1":"19378","2":"37","3":"GWAS_Catalog_2019","4":"https://www.ebi.ac.uk/gwas","5":"1737","6":"fef3864bcb5dd9e60cee27357eff30226116c49b","7":"4"},{"1":"6201","2":"45","3":"WikiPathways_2019_Human","4":"https://www.wikipathways.org/","5":"472","6":"b0c9e9ebb9014f14561e896008087725a2db24b7","7":"7"},{"1":"4558","2":"54","3":"WikiPathways_2019_Mouse","4":"https://www.wikipathways.org/","5":"176","6":"e7750958da20f585c8b6d5bc4451a5a4305514ba","7":"7"},{"1":"3264","2":"22","3":"TRRUST_Transcription_Factors_2019","4":"https://www.grnpedia.org/trrust/","5":"571","6":"5f8cf93e193d2bcefa5a37ccdf0eefac576861b0","7":"1"},{"1":"7802","2":"92","3":"KEGG_2019_Human","4":"https://www.kegg.jp/","5":"308","6":"3477bc578c4ea5d851dcb934fe2a41e9fd789bb4","7":"7"},{"1":"8551","2":"98","3":"KEGG_2019_Mouse","4":"https://www.kegg.jp/","5":"303","6":"187eb44b2d6fa154ebf628eba1f18537f64e797c","7":"7"},{"1":"12444","2":"23","3":"InterPro_Domains_2019","4":"https://www.ebi.ac.uk/interpro/","5":"1071","6":"18dd5ec520fdf589a93d6a7911289c205e1ddf22","7":"6"},{"1":"9000","2":"20","3":"Pfam_Domains_2019","4":"https://pfam.xfam.org/","5":"608","6":"a6325ed264f9ac9e6518796076c46a1d885cca7a","7":"6"},{"1":"7744","2":"363","3":"DepMap_WG_CRISPR_Screens_Broad_CellLines_2019","4":"https://depmap.org/","5":"558","6":"0b08b32b20854ac8a738458728a9ea50c2e04800","7":"4"},{"1":"6204","2":"387","3":"DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019","4":"https://depmap.org/","5":"325","6":"b7c4ead26d0eb64f1697c030d31682b581c8bb56","7":"4"},{"1":"13420","2":"32","3":"MGI_Mammalian_Phenotype_Level_4_2019","4":"http://www.informatics.jax.org/","5":"5261","6":"f1bed632e89ebc054da44236c4815cdce03ef5ee","7":"7"},{"1":"14148","2":"122","3":"UK_Biobank_GWAS_v1","4":"https://www.ukbiobank.ac.uk/tag/gwas/","5":"857","6":"958fb52e6215626673a5acf6e9289a1b84d11b4a","7":"4"},{"1":"9813","2":"49","3":"BioPlanet_2019","4":"https://tripod.nih.gov/bioplanet/","5":"1510","6":"e110851dfc763d30946f2abedcc2cd571ac357a0","7":"2"},{"1":"1397","2":"13","3":"ClinVar_2019","4":"https://www.ncbi.nlm.nih.gov/clinvar/","5":"182","6":"0a95303f8059bec08836ecfe02ce3da951150547","7":"4"},{"1":"9116","2":"22","3":"PheWeb_2019","4":"http://pheweb.sph.umich.edu/","5":"1161","6":"6a7c7321b6b72c5285b722f7902d26a2611117cb","7":"4"},{"1":"17464","2":"63","3":"DisGeNET","4":"https://www.disgenet.org","5":"9828","6":"3c261626478ce9e6bf2c7f0a8014c5e901d43dc0","7":"4"},{"1":"394","2":"73","3":"HMS_LINCS_KinomeScan","4":"http://lincs.hms.harvard.edu/kinomescan/","5":"148","6":"47ba06cdc92469ac79400fc57acd84ba343ba616","7":"2"},{"1":"11851","2":"586","3":"CCLE_Proteomics_2020","4":"https://portals.broadinstitute.org/ccle","5":"378","6":"7094b097ae2301a1d6a5bd856a193b084cca993d","7":"5"},{"1":"8189","2":"421","3":"ProteomicsDB_2020","4":"https://www.proteomicsdb.org/","5":"913","6":"8c87c8346167bac2ba68195a32458aba9b1acfd1","7":"5"},{"1":"18704","2":"100","3":"lncHUB_lncRNA_Co-Expression","4":"https://amp.pharm.mssm.edu/lnchub/","5":"3729","6":"45b597d7efa5693b7e4172b09c0ed2dda3305582","7":"1"},{"1":"5605","2":"39","3":"Virus-Host_PPI_P-HIPSTer_2020","4":"http://phipster.org/","5":"6715","6":"a592eed13e8e9496aedbab63003b965574e46a65","7":"2"},{"1":"5718","2":"31","3":"Elsevier_Pathway_Collection","4":"http://www.transgene.ru/disease-pathways/","5":"1721","6":"9196c760e3bcae9c9de1e3f87ad81f96bde24325","7":"2"},{"1":"14156","2":"40","3":"Table_Mining_of_CRISPR_Studies","4":"","5":"802","6":"ad580f3864fa8ff69eaca11f6d2e7f9b86378d08","7":"6"},{"1":"16979","2":"295","3":"COVID-19_Related_Gene_Sets","4":"https://amp.pharm.mssm.edu/covid19","5":"205","6":"72b0346849570f66a77a6856722601e711596cb4","7":"7"},{"1":"4383","2":"146","3":"MSigDB_Hallmark_2020","4":"https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp","5":"50","6":"6952efda94663d4bd8db09bf6eeb4e67d21ef58c","7":"2"},{"1":"54974","2":"483","3":"Enrichr_Users_Contributed_Lists_2020","4":"https://maayanlab.cloud/Enrichr","5":"1482","6":"8dc362703b38b30ac3b68b6401a9b20a58e7d3ef","7":"6"},{"1":"12118","2":"448","3":"TG_GATES_2020","4":"https://toxico.nibiohn.go.jp/english/","5":"1190","6":"9e32560437b11b4628b00ccf3d584360f7f7daee","7":"4"},{"1":"12361","2":"124","3":"Allen_Brain_Atlas_10x_scRNA_2021","4":"https://portal.brain-map.org/","5":"766","6":"46f8235cb585829331799a71aec3f7c082170219","7":"5"},{"1":"9763","2":"139","3":"Descartes_Cell_Types_and_Tissue_2021","4":"https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/","5":"172","6":"","7":"5"},{"1":"8078","2":"102","3":"KEGG_2021_Human","4":"https://www.kegg.jp/","5":"320","6":"","7":"2"},{"1":"7173","2":"43","3":"WikiPathway_2021_Human","4":"https://www.wikipathways.org/","5":"622","6":"","7":"2"},{"1":"5833","2":"100","3":"HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression","4":"https://hubmapconsortium.github.io/ccf-asct-reporter/","5":"344","6":"","7":"5"},{"1":"14937","2":"33","3":"GO_Biological_Process_2021","4":"http://www.geneontology.org/","5":"6036","6":"","7":"3"},{"1":"11497","2":"80","3":"GO_Cellular_Component_2021","4":"http://www.geneontology.org/","5":"511","6":"","7":"3"},{"1":"11936","2":"34","3":"GO_Molecular_Function_2021","4":"http://www.geneontology.org/","5":"1274","6":"","7":"3"},{"1":"9767","2":"33","3":"MGI_Mammalian_Phenotype_Level_4_2021","4":"http://www.informatics.jax.org/","5":"4601","6":"","7":"3"},{"1":"14167","2":"80","3":"CellMarker_Augmented_2021","4":"http://biocc.hrbmu.edu.cn/CellMarker/","5":"1097","6":"","7":"5"},{"1":"17851","2":"102","3":"Orphanet_Augmented_2021","4":"http://www.orphadata.org/","5":"3774","6":"","7":"4"},{"1":"16853","2":"360","3":"COVID-19_Related_Gene_Sets_2021","4":"https://maayanlab.cloud/covid19/","5":"478","6":"","7":"4"},{"1":"6654","2":"136","3":"PanglaoDB_Augmented_2021","4":"https://panglaodb.se/","5":"178","6":"","7":"5"},{"1":"1683","2":"10","3":"Azimuth_Cell_Types_2021","4":"https://azimuth.hubmapconsortium.org/","5":"341","6":"","7":"5"},{"1":"20414","2":"112","3":"PhenGenI_Association_2021","4":"https://www.ncbi.nlm.nih.gov/gap/phegeni","5":"950","6":"","7":"4"},{"1":"26076","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Human_Down","4":"https://maayanlab.cloud/archs4/","5":"4269","6":"","7":"8"},{"1":"26338","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Human_Up","4":"https://maayanlab.cloud/archs4/","5":"4269","6":"","7":"8"},{"1":"25381","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Mouse_Down","4":"https://maayanlab.cloud/archs4/","5":"4216","6":"","7":"8"},{"1":"25409","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Mouse_Up","4":"https://maayanlab.cloud/archs4/","5":"4216","6":"","7":"8"},{"1":"11980","2":"250","3":"GTEx_Aging_Signatures_2021","4":"https://gtexportal.org/","5":"270","6":"","7":"4"},{"1":"31158","2":"805","3":"HDSigDB_Human_2021","4":"https://www.hdinhd.org/","5":"2564","6":"","7":"4"},{"1":"30006","2":"815","3":"HDSigDB_Mouse_2021","4":"https://www.hdinhd.org/","5":"2579","6":"","7":"4"},{"1":"13370","2":"103","3":"HuBMAP_ASCTplusB_augmented_2022","4":"https://hubmapconsortium.github.io/ccf-asct-reporter/","5":"777","6":"","7":"5"},{"1":"13697","2":"343","3":"FANTOM6_lncRNA_KD_DEGs","4":"https://fantom.gsc.riken.jp/6/","5":"350","6":"","7":"1"},{"1":"2183","2":"18","3":"MAGMA_Drugs_and_Diseases","4":"https://github.com/nybell/drugsets/tree/main/DATA/GENESETS","5":"1395","6":"","7":"4"},{"1":"12765","2":"13","3":"PFOCR_Pathways","4":"https://www.wikipathways.org/index.php/Portal:PFOCR","5":"17326","6":"","7":"2"},{"1":"1509","2":"100","3":"Tabula_Sapiens","4":"https://tabula-sapiens-portal.ds.czbiohub.org/","5":"469","6":"","7":"5"},{"1":"18365","2":"1214","3":"ChEA_2022","4":"https://maayanlab.cloud/chea3/","5":"757","6":"","7":"1"},{"1":"13525","2":"175","3":"Diabetes_Perturbations_GEO_2022","4":"https://appyters.maayanlab.cloud/#/Gene_Expression_T2D_Signatures","5":"601","6":"","7":"4"},{"1":"9525","2":"245","3":"LINCS_L1000_Chem_Pert_Consensus_Sigs","4":"https://maayanlab.cloud/sigcom-lincs/#/Download","5":"10850","6":"","7":"4"},{"1":"9440","2":"245","3":"LINCS_L1000_CRISPR_KO_Consensus_Sigs","4":"https://maayanlab.cloud/sigcom-lincs/#/Download","5":"10424","6":"","7":"4"},{"1":"3857","2":"80","3":"Tabula_Muris","4":"https://tabula-muris.ds.czbiohub.org/","5":"106","6":"","7":"5"},{"1":"10489","2":"61","3":"Reactome_2022","4":"https://reactome.org/download-data","5":"1818","6":"","7":"2"},{"1":"1198","2":"23","3":"SynGO_2022","4":"https://www.syngoportal.org/","5":"118","6":"","7":"3"},{"1":"1882","2":"47","3":"GlyGen_Glycosylated_Proteins_2022","4":"https://www.glygen.org/","5":"338","6":"","7":"2"},{"1":"1552","2":"16","3":"IDG_Drug_Targets_2022","4":"https://drugcentral.org/","5":"888","6":"","7":"4"},{"1":"6713","2":"68","3":"KOMP2_Mouse_Phenotypes_2022","4":"https://www.mousephenotype.org/","5":"529","6":"","7":"3"},{"1":"936","2":"15","3":"Metabolomics_Workbench_Metabolites_2022","4":"https://www.metabolomicsworkbench.org/","5":"233","6":"","7":"2"},{"1":"8220","2":"146","3":"Proteomics_Drug_Atlas_2023","4":"https://www.nature.com/articles/s41587-022-01539-0","5":"1748","6":"","7":"4"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>

```r
# Perform enrichment
enrich_results <- enrichr(genes = DGE_cell_selection$gene[DGE_cell_selection$avg_log2FC >
    0], databases = "GO_Biological_Process_2017b")[[1]]
```

```
## Uploading data to Enrichr... Done.
##   Querying GO_Biological_Process_2017b... Done.
## Parsing results... Done.
```


Some databases of interest:

* `GO_Biological_Process_2017b`
* `KEGG_2019_Human`
* `KEGG_2019_Mouse`
* `WikiPathways_2019_Human`
* `WikiPathways_2019_Mouse`

You visualize your results using a simple barplot, for example:


```r
par(mfrow = c(1, 1), mar = c(3, 25, 2, 1))
barplot(height = -log10(enrich_results$P.value)[10:1], names.arg = enrich_results$Term[10:1],
    horiz = TRUE, las = 1, border = FALSE, cex.names = 0.6)
abline(v = c(-log10(0.05)), lty = 2)
abline(v = 0, lty = 1)
```

![](seurat_05_dge_files/figure-html/unnamed-chunk-32-1.png)<!-- -->

## Gene Set Enrichment Analysis (GSEA)

Besides the enrichment using hypergeometric test, we can also perform gene set enrichment analysis (GSEA), which scores ranked genes list (usually based on fold changes) and computes permutation test to check if a particular gene set is more present in the Up-regulated genes, amongthe DOWN_regulated genes or not differentially regulated.


```r
cell_selection = SetIdent(cell_selection, value = "type")

DGE_cell_selection2 <- FindMarkers(cell_selection, ident.1 = "Covid", log2FC.threshold = -Inf,
    test.use = "wilcox", min.pct = 0.1, min.diff.pct = 0, only.pos = FALSE, max.cells.per.ident = 50,
    assay = "RNA")

# Create a gene rank based on the gene expression fold change
gene_rank <- setNames(DGE_cell_selection2$avg_log2FC, casefold(rownames(DGE_cell_selection2),
    upper = T))
```

 Once our list of genes are sorted, we can proceed with the enrichment itself. We can use the package to get gene set from the Molecular Signature Database (MSigDB) and select KEGG pathways as an example.


```r
library(msigdbr)

# Download gene sets
msigdbgmt <- msigdbr::msigdbr("Homo sapiens")
msigdbgmt <- as.data.frame(msigdbgmt)

# List available gene sets
unique(msigdbgmt$gs_subcat)
```

```
##  [1] "MIR:MIR_Legacy"  "TFT:TFT_Legacy"  "CGP"             "TFT:GTRD"       
##  [5] ""                "VAX"             "CP:BIOCARTA"     "CGN"            
##  [9] "GO:BP"           "GO:CC"           "IMMUNESIGDB"     "GO:MF"          
## [13] "HPO"             "CP:KEGG"         "MIR:MIRDB"       "CM"             
## [17] "CP"              "CP:PID"          "CP:REACTOME"     "CP:WIKIPATHWAYS"
```

```r
# Subset which gene set you want to use.
msigdbgmt_subset <- msigdbgmt[msigdbgmt$gs_subcat == "CP:WIKIPATHWAYS", ]
gmt <- lapply(unique(msigdbgmt_subset$gs_name), function(x) {
    msigdbgmt_subset[msigdbgmt_subset$gs_name == x, "gene_symbol"]
})
names(gmt) <- unique(paste0(msigdbgmt_subset$gs_name, "_", msigdbgmt_subset$gs_exact_source))
```

 Next, we will be using the GSEA. This will result in a table containing information for several pathways. We can then sort and filter those pathways to visualize only the top ones. You can select/filter them by either `p-value` or normalized enrichemnet score (`NES`).


```r
library(fgsea)

# Perform enrichemnt analysis
fgseaRes <- fgsea(pathways = gmt, stats = gene_rank, minSize = 15, maxSize = 500)
fgseaRes <- fgseaRes[order(fgseaRes$pval, decreasing = T), ]

# Filter the results table to show only the top 10 UP or DOWN regulated
# processes (optional)
top10_UP <- fgseaRes$pathway[1:10]

# Nice summary table (shown as a plot)
dev.off()
plotGseaTable(gmt[top10_UP], gene_rank, fgseaRes, gseaParam = 0.5)
```

<style>
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 10px;}
</style>
<div class = "blue">
**Your turn**

Which KEGG pathways are upregulated in this cluster?
Which KEGG pathways are dowregulated in this cluster?
Change the pathway source to another gene set (e.g. "CP:WIKIPATHWAYS" or "CP:REACTOME" or "CP:BIOCARTA" or "GO:BP") and check the if you get simmilar results?
</div>

Finally, lets save the integrated data for further analysis.


```r
saveRDS(alldata, "data/3pbmc_qc_dr_int_cl_dge.rds")
# aslo, save the list of DGE results to a file.
write.csv(markers_genes, file = "data/3pbmc_qc_dr_int_cl_dge.csv")
```


### Session Info
***


```r
sessionInfo()
```

```
## R version 4.1.3 (2022-03-10)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Big Sur/Monterey 10.16
## 
## Matrix products: default
## BLAS/LAPACK: /Users/asabjor/miniconda3/envs/scRNAseq2023/lib/libopenblasp-r0.3.21.dylib
## 
## locale:
## [1] C/UTF-8/C/C/C/C
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] fgsea_1.20.0                msigdbr_7.5.1              
##  [3] lme4_1.1-31                 MAST_1.20.0                
##  [5] SingleCellExperiment_1.16.0 SummarizedExperiment_1.24.0
##  [7] Biobase_2.54.0              GenomicRanges_1.46.1       
##  [9] GenomeInfoDb_1.30.1         IRanges_2.28.0             
## [11] S4Vectors_0.32.4            BiocGenerics_0.40.0        
## [13] MatrixGenerics_1.6.0        matrixStats_0.63.0         
## [15] edgeR_3.36.0                limma_3.50.3               
## [17] Matrix_1.5-3                rafalib_1.0.0              
## [19] enrichR_3.1                 pheatmap_1.0.12            
## [21] ggplot2_3.4.0               cowplot_1.1.1              
## [23] dplyr_1.0.10                SeuratObject_4.1.3         
## [25] Seurat_4.3.0                RJSONIO_1.3-1.7            
## [27] optparse_1.7.3             
## 
## loaded via a namespace (and not attached):
##   [1] fastmatch_1.1-3        plyr_1.8.8             igraph_1.3.5          
##   [4] lazyeval_0.2.2         sp_1.6-0               splines_4.1.3         
##   [7] BiocParallel_1.28.3    listenv_0.9.0          scattermore_0.8       
##  [10] digest_0.6.31          htmltools_0.5.4        fansi_1.0.4           
##  [13] magrittr_2.0.3         tensor_1.5             cluster_2.1.4         
##  [16] ROCR_1.0-11            globals_0.16.2         spatstat.sparse_3.0-0 
##  [19] colorspace_2.1-0       ggrepel_0.9.2          xfun_0.36             
##  [22] RCurl_1.98-1.9         jsonlite_1.8.4         progressr_0.13.0      
##  [25] spatstat.data_3.0-0    survival_3.5-0         zoo_1.8-11            
##  [28] glue_1.6.2             polyclip_1.10-4        gtable_0.3.1          
##  [31] zlibbioc_1.40.0        XVector_0.34.0         leiden_0.4.3          
##  [34] DelayedArray_0.20.0    future.apply_1.10.0    abind_1.4-5           
##  [37] scales_1.2.1           DBI_1.1.3              spatstat.random_3.0-1 
##  [40] miniUI_0.1.1.1         Rcpp_1.0.10            viridisLite_0.4.1     
##  [43] xtable_1.8-4           reticulate_1.27        htmlwidgets_1.6.1     
##  [46] httr_1.4.4             getopt_1.20.3          RColorBrewer_1.1-3    
##  [49] ellipsis_0.3.2         ica_1.0-3              farver_2.1.1          
##  [52] pkgconfig_2.0.3        sass_0.4.5             uwot_0.1.14           
##  [55] deldir_1.0-6           locfit_1.5-9.7         utf8_1.2.2            
##  [58] labeling_0.4.2         tidyselect_1.2.0       rlang_1.0.6           
##  [61] reshape2_1.4.4         later_1.3.0            munsell_0.5.0         
##  [64] tools_4.1.3            cachem_1.0.6           cli_3.6.0             
##  [67] generics_0.1.3         ggridges_0.5.4         evaluate_0.20         
##  [70] stringr_1.5.0          fastmap_1.1.0          yaml_2.3.7            
##  [73] goftest_1.2-3          babelgene_22.9         knitr_1.41            
##  [76] fitdistrplus_1.1-8     purrr_1.0.1            RANN_2.6.1            
##  [79] pbapply_1.7-0          future_1.30.0          nlme_3.1-161          
##  [82] mime_0.12              formatR_1.14           compiler_4.1.3        
##  [85] plotly_4.10.1          curl_4.3.3             png_0.1-8             
##  [88] spatstat.utils_3.0-1   tibble_3.1.8           bslib_0.4.2           
##  [91] stringi_1.7.12         highr_0.10             lattice_0.20-45       
##  [94] nloptr_2.0.3           vctrs_0.5.2            pillar_1.8.1          
##  [97] lifecycle_1.0.3        spatstat.geom_3.0-5    lmtest_0.9-40         
## [100] jquerylib_0.1.4        RcppAnnoy_0.0.20       data.table_1.14.6     
## [103] bitops_1.0-7           irlba_2.3.5.1          httpuv_1.6.8          
## [106] patchwork_1.1.2        R6_2.5.1               promises_1.2.0.1      
## [109] KernSmooth_2.23-20     gridExtra_2.3          parallelly_1.34.0     
## [112] codetools_0.2-18       boot_1.3-28.1          MASS_7.3-58.2         
## [115] assertthat_0.2.1       rjson_0.2.21           withr_2.5.0           
## [118] sctransform_0.3.5      GenomeInfoDbData_1.2.7 parallel_4.1.3        
## [121] grid_4.1.3             minqa_1.2.5            tidyr_1.2.1           
## [124] rmarkdown_2.20         Rtsne_0.16             spatstat.explore_3.0-5
## [127] shiny_1.7.4
```
