---
author: "Åsa Björklund  &  Paulo Czarnewski"
date: "Sept 13, 2019"
output:
  html_document:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    keep_md: yes
    fig_caption: true
  html_notebook:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message='hide', warning='hide', result='hold',fig.width=12)
```
<style>
h1, .h1, h2, .h2, h3, .h3, h4, .h4 { margin-top: 50px }
p.caption {font-size: 0.9em;font-style: italic;color: grey;margin-right: 10%;margin-left: 10%;text-align: justify}
</style>

# Differential gene expression

In this tutorial we will cover about Differetial gene expression, which comprises an extensive range of topics and methods. In single cell, differential expresison can have multiple functionalities such as of identifying marker genes for cell populations, as well as differentially regulated genes across conditions (healthy vs control). We will also exercise on how to account the batch information in your test.

We can first load the data from the clustering session. Moreover, we can already decide which clustering resolution to use. First let's define using the `louvain_2` clustering to identifying differentially expressed genes.  

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(dplyr)
  library(cowplot)
  library(ggplot2)
  library(pheatmap)
  library(rafalib)
})

alldata <- readRDS("data/3pbmc_qc_dr_int_cl.rds")

#Set the identity as louvain_2 clustering
print(alldata@active.ident[1:10])
alldata <- SetIdent(alldata, value = "kmeans_5")
print(alldata@active.ident[1:10])
```

## Cell marker genes
***

Let us first compute a ranking for the highly differential genes in each cluster. There are many different tests and parameters to be chosen that can be used to refine your results. When looking for marker genes, we want genes that are positivelly expressed in a cell type and possibly not expressed in the others.

```{r}
#Compute differentiall expression
markers_genes <- FindAllMarkers(alldata,
                               logfc.threshold = 0.2,
                               test.use = "wilcox",
                               min.pct = 0.1,
                               min.diff.pct = 0.2,
                               only.pos = TRUE,
                               max.cells.per.ident = 50,
                               assay = "RNA")
```

We can now select the top 25 up regulated genes for plotting.

```{r}
markers_genes %>% group_by(cluster)  %>% top_n(-25, p_val_adj) -> top25
top25
```

We can now select the top 25 up regulated genes for plotting.

```{r}
mypar(1,5,mar=c(4,6,3,1))
for(i in unique(top25$cluster)){
  barplot( sort( setNames(top25$avg_logFC, top25$gene) [top25$cluster == i], F),
           horiz = T,las=1 ,main=paste0(i," vs. rest"),border = "white", yaxs="i" )
  abline(v=c(0,0.25),lty=c(1,2))
}
```

We can visualize them as a heatmap. Here we are selecting the top 5.

```{r}
markers_genes %>% group_by(cluster)  %>% top_n(-5, p_val_adj) -> top5

alldata <- ScaleData(alldata, features = as.character(unique(top5$gene)), assay = "RNA")
DoHeatmap(alldata, features = as.character(unique(top5$gene)),group.by = "kmeans_5", assay = "RNA")
```

Another way is by representing the overal group expression and detection rates in a dot-plot.

```{r}
DotPlot(alldata, features = as.character(unique(top5$gene)),group.by = "kmeans_5",assay = "RNA")+coord_flip()
```

We can also plot a violin plot for each gene.

```{r,fig.height=10,fig.width=10}
VlnPlot(alldata, features = as.character(unique(top5$gene)), ncol = 5, group.by = "kmeans_5",assay = "RNA")
```

<style>
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 10px;}
</style>
<div class = "blue">
**Your turn**

Take a screen shot of those results and re-run the same code above with another test: "wilcox" (Wilcoxon Rank Sum test), "bimod" (Likelihood-ratio test), "roc" (Identifies 'markers' of gene expression using ROC analysis),"t" (Student's t-test),"negbinom" (negative binomial generalized linear model),"poisson" (poisson generalized linear model), "LR" (logistic regression), "MAST" (hurdle model), "DESeq2" (negative binomial distribution).
</div>

## Differential expression across conditions
***

The second way of computing differential expression is to answer which genes are differentially expressed within a cluster. For example, in our case we have libraries comming from 2 different library preparation methods (batches) and we would like to know which genes are influenced the most in a particular cell type. The same concenpt applies if you have instead two or more biological groups (control vs treated, time#0 vs time#1 vs time#2, etc).

For this end, we will first subset our data for the desired cell cluster, then change the cell identities to the variable of comparison (which now in our case is the "Chemistry").

```{r,fig.height=10,fig.width=10}
cell_selection <- subset(alldata, cells = colnames(alldata)[ alldata$kmeans_5 == 4 ])
cell_selection <- SetIdent(cell_selection, value = "Chemistry")
#Compute differentiall expression
DGE_cell_selection <- FindAllMarkers(cell_selection,
                               logfc.threshold = 0.2,
                               test.use = "wilcox",
                               min.pct = 0.1,
                               min.diff.pct = 0.2,
                               only.pos = TRUE,
                               max.cells.per.ident = 50,
                               assay = "RNA")
```

We can now plot the expression across the "Chemistry".

```{r}
DGE_cell_selection %>% group_by(cluster)  %>% top_n(-5, p_val_adj) -> top5_cell_selection

VlnPlot(cell_selection, features = as.character(unique(top5_cell_selection$gene)),
        ncol = 5,group.by = "Chemistry",assay = "RNA")
```

We can clearly see some patterns across them. Those are the genes that impact the most on your batches (see the dimensionality reduction and integration exercises for more details). We can plot those genes using the integrated and non-integrated UMAP for ilustration.

```{r,fig.asp=.28}
FeaturePlot(alldata, reduction = "UMAP_on_CCA",dims = 1:2,
            features = c("JUND","RPS17","CD81"),order = T,ncol = 3)
```


Finally, lets save the integrated data for further analysis.

```{r}
saveRDS(alldata,"data/3pbmc_qc_dr_int_cl.rds")
```


### Session Info
***

```{r}
sessionInfo()
```



















