<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund &amp; Paulo Czarnewski">
<meta name="description" content="Assignment of cell identities based on gene expression patterns using reference data.">

<title> Celltype prediction</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background: #E9F2D1;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=JetBrains+Mono&amp;family=Nunito:ital,wght@0,400;0,600;1,400;1,600&amp;display=swap" rel="stylesheet">
    

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo." class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-bacterium" aria-label="bacterium"></i> Celltype prediction</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> SEURAT TOOLKIT</p>
                  <div>
        <div class="description">
          Assignment of cell identities based on gene expression patterns using reference data.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund &amp; Paulo Czarnewski </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">02-Nov-2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#meta-ct_read" id="toc-meta-ct_read" class="nav-link active" data-scroll-target="#meta-ct_read"><span class="header-section-number">1</span> Read data</a></li>
  <li><a href="#meta-ct_ref" id="toc-meta-ct_ref" class="nav-link" data-scroll-target="#meta-ct_ref"><span class="header-section-number">2</span> Reference data</a></li>
  <li><a href="#label-transfer" id="toc-label-transfer" class="nav-link" data-scroll-target="#label-transfer"><span class="header-section-number">3</span> Label transfer</a></li>
  <li><a href="#meta-ct_scpred" id="toc-meta-ct_scpred" class="nav-link" data-scroll-target="#meta-ct_scpred"><span class="header-section-number">4</span> scPred</a></li>
  <li><a href="#meta-ct_compare" id="toc-meta-ct_compare" class="nav-link" data-scroll-target="#meta-ct_compare"><span class="header-section-number">5</span> Compare results</a></li>
  <li><a href="#meta-ct_gsea" id="toc-meta-ct_gsea" class="nav-link" data-scroll-target="#meta-ct_gsea"><span class="header-section-number">6</span> GSEA with celltype markers</a>
  <ul>
  <li><a href="#meta-ct_gsea_deg" id="toc-meta-ct_gsea_deg" class="nav-link" data-scroll-target="#meta-ct_gsea_deg"><span class="header-section-number">6.1</span> DEG overlap</a></li>
  <li><a href="#meta-ct_gsea_annot" id="toc-meta-ct_gsea_annot" class="nav-link" data-scroll-target="#meta-ct_gsea_annot"><span class="header-section-number">6.2</span> With annotated gene sets</a></li>
  </ul></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">7</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run R commands unless otherwise specified.</p>
</div>
</div>
<p>Celltype prediction can either be performed on indiviudal cells where each cell gets a predicted celltype label, or on the level of clusters. All methods are based on similarity to other datasets, single cell or sorted bulk RNAseq, or uses known marker genes for each celltype.<br>
We will select one sample from the Covid data, <code>ctrl_13</code> and predict celltype by cell on that sample.<br>
Some methods will predict a celltype to each cell based on what it is most similar to even if the celltype of that cell is not included in the reference. Other methods include an uncertainty so that cells with low similarity scores will be unclassified.<br>
There are multiple different methods to predict celltypes, here we will just cover a few of those.</p>
<p>Here we will use a reference PBMC dataset from the <code>scPred</code> package which is provided as a Seurat object with counts. And we will test classification based on the <code>scPred</code> and <code>scMap</code> methods. Finally we will use gene set enrichment predict celltype based on the DEGs of each cluster.</p>
<section id="meta-ct_read" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="meta-ct_read"><span class="header-section-number">1</span> Read data</h2>
<p>First, lets load required libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(cowplot)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pheatmap)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(rafalib)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remotes::install_github("powellgenomicslab/scPred")</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(scPred)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s read in the saved Covid-19 data object from the clustering step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load the data and select 'ctrl_13` sample</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>alldata <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/results/covid_qc_dr_int_cl.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Subset one patient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">=</span> alldata[, alldata<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">'ctrl_13'</span>]</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set active assay to RNA and remove the CCA assay</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>ctrl<span class="sc">@</span>active.assay <span class="ot">=</span> <span class="st">'RNA'</span> </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>ctrl[[<span class="st">'CCA'</span>]] <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>ctrl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
18121 features across 1139 samples within 1 assay 
Active assay: RNA (18121 features, 0 variable features)
 6 dimensional reductions calculated: umap, tsne, harmony, umap_harmony, scanorama, umap_scanorama</code></pre>
</div>
</div>
</section>
<section id="meta-ct_ref" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="meta-ct_ref"><span class="header-section-number">2</span> Reference data</h2>
<p>Load the reference dataset with annotated labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> scPred<span class="sc">::</span>pbmc_1</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>reference</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
32838 features across 3500 samples within 1 assay 
Active assay: RNA (32838 features, 0 variable features)</code></pre>
</div>
</div>
<p>Rerun analysis pipeline. Run normalization, feature selection and dimensionality reduction</p>
<p>Here, we will run all the steps that we did in previous labs in one go using the <code>magittr</code> package with the pipe-operator <code>%&gt;%</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> reference <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunPCA</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunUMAP</span>(<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(reference, <span class="at">group.by =</span> <span class="st">"cell_type"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-6-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Run all steps of the analysis for the ctrl sample as well. Use the clustering from the integration lab with resolution 0.5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Set the identity as louvain with resolution 0.3</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(ctrl, <span class="at">value =</span> <span class="st">"CCA_snn_res.0.5"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> ctrl <span class="sc">%&gt;%</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunPCA</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunUMAP</span>(<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl,  <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-8-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="label-transfer" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="label-transfer"><span class="header-section-number">3</span> Label transfer</h2>
<p>First we will run label transfer using a similar method as in the integration exercise. But, instad of CCA the default for the ’FindTransferAnchors` function is to use “pcaproject”, e.g.&nbsp;the query dataset is projected onto the PCA of the reference dataset. Then, the labels of the reference data are predicted.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>transfer.anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(<span class="at">reference =</span> reference, <span class="at">query =</span> ctrl, </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> transfer.anchors, <span class="at">refdata =</span> reference<span class="sc">$</span>cell_type, </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(<span class="at">object =</span> ctrl, <span class="at">metadata =</span> predictions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl, <span class="at">group.by =</span> <span class="st">"predicted.id"</span>, <span class="at">label =</span> T, <span class="at">repel =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-10-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Now plot how many cells of each celltypes can be found in each cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>CCA_snn_res.<span class="fl">0.5</span>, <span class="at">fill =</span> predicted.id)) <span class="sc">+</span> <span class="fu">geom_bar</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-11-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_scpred" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="meta-ct_scpred"><span class="header-section-number">4</span> scPred</h2>
<p>scPred will train a classifier based on all principal components. First, <code>getFeatureSpace</code> will create a scPred object stored in the <code>@misc</code> slot where it extracts the PCs that best separates the different celltypes. Then <code>trainModel</code> will do the actual training for each celltype.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">getFeatureSpace</span>(reference, <span class="st">"cell_type"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">trainModel</span>(reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Extracting feature space for each cell type...
DONE!
●  Training models for each cell type...
maximum number of iterations reached 0.0001152056 -0.0001143117DONE!</code></pre>
</div>
</div>
<p>We can then print how well the training worked for the different celltypes by printing the number of PCs used for each, the ROC value and Sensitivity/Specificity. Which celltypes do you think are harder to classify based on this dataset?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get_scpred</span>(reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'scPred' object
✔  Prediction variable = cell_type 
✔  Discriminant features per cell type
✔  Training model(s)
Summary

|Cell type   |    n| Features|Method    |   ROC|  Sens|  Spec|
|:-----------|----:|--------:|:---------|-----:|-----:|-----:|
|B cell      |  280|       50|svmRadial | 1.000| 0.964| 1.000|
|CD4 T cell  | 1620|       50|svmRadial | 0.997| 0.972| 0.975|
|CD8 T cell  |  945|       50|svmRadial | 0.985| 0.899| 0.978|
|cDC         |   26|       50|svmRadial | 0.995| 0.547| 1.000|
|cMono       |  212|       50|svmRadial | 0.994| 0.958| 0.970|
|ncMono      |   79|       50|svmRadial | 0.998| 0.570| 1.000|
|NK cell     |  312|       50|svmRadial | 0.999| 0.933| 0.996|
|pDC         |   20|       50|svmRadial | 1.000| 0.700| 1.000|
|Plasma cell |    6|       50|svmRadial | 1.000| 0.800| 1.000|</code></pre>
</div>
</div>
<p>You can optimize parameters for each dataset by chaning parameters and testing different types of models, see more at: https://powellgenomicslab.github.io/scPred/articles/introduction.html. But for now, we will continue with this model. Now, lets predict celltypes on our data, where scPred will align the two datasets with Harmony and then perform classification.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">scPredict</span>(ctrl, reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Matching reference with new dataset...
     ─ 2000 features present in reference loadings
     ─ 1774 features shared between reference and new dataset
     ─ 88.7% of features in the reference are present in new dataset
●  Aligning new data to reference...
●  Classifying cells...
DONE!</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl, <span class="at">group.by =</span> <span class="st">"scpred_prediction"</span>, <span class="at">label =</span> T, <span class="at">repel =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-15-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Now plot how many cells of each celltypes can be found in each cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>CCA_snn_res.<span class="fl">0.5</span>, <span class="at">fill =</span> scpred_prediction)) <span class="sc">+</span> <span class="fu">geom_bar</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-16-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_compare" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="meta-ct_compare"><span class="header-section-number">5</span> Compare results</h2>
<p>Now we will compare the output of the two methods using the convenient function in scPred <code>crossTab</code> that prints the overlap between two metadata slots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">crossTab</span>(ctrl, <span class="st">"predicted.id"</span>, <span class="st">"scpred_prediction"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 14%">
<col style="width: 8%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 4%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 9%">
<col style="width: 4%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">B cell</th>
<th style="text-align: right;">CD4 T cell</th>
<th style="text-align: right;">CD8 T cell</th>
<th style="text-align: right;">cDC</th>
<th style="text-align: right;">cMono</th>
<th style="text-align: right;">ncMono</th>
<th style="text-align: right;">NK cell</th>
<th style="text-align: right;">pDC</th>
<th style="text-align: right;">Plasma cell</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">B cell</td>
<td style="text-align: right;">103</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4 T cell</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">200</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD8 T cell</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">225</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">cDC</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">cMono</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">198</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">ncMono</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">101</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NK cell</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">153</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">pDC</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Plasma cell</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">unassigned</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">55</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</section>
<section id="meta-ct_gsea" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-ct_gsea"><span class="header-section-number">6</span> GSEA with celltype markers</h2>
<p>Another option, where celltype can be classified on cluster level is to use gene set enrichment among the DEGs with known markers for different celltypes. Similar to how we did functional enrichment for the DEGs in the Differential expression exercise. There are some resources for celltype gene sets that can be used. Such as <a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/">CellMarker</a>, <a href="https://panglaodb.se/">PanglaoDB</a> or celltype gene sets at <a href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a>. We can also look at overlap between DEGs in a reference dataset and the dataset you are analysing.</p>
<section id="meta-ct_gsea_deg" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="meta-ct_gsea_deg"><span class="header-section-number">6.1</span> DEG overlap</h3>
<p>First, lets extract top DEGs for our Covid-19 dataset and the reference dataset. When we run differential expression for our dataset, we want to report as many genes as possible, hence we set the cutoffs quite lenient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run differential expression in our dataset, using clustering at resolution 0.5</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>alldata <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(alldata,<span class="at">value =</span> <span class="st">"CCA_snn_res.0.5"</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>DGE_table <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  alldata,</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">logfc.threshold =</span> <span class="dv">0</span>,</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">test.use =</span> <span class="st">"wilcox"</span>,</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.pct =</span> <span class="fl">0.1</span>,</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.diff.pct =</span> <span class="dv">0</span>,</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">max.cells.per.ident =</span> <span class="dv">20</span>,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">return.thresh =</span> <span class="dv">1</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"RNA"</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="co"># split into a list</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>DGE_list <span class="ot">&lt;-</span> <span class="fu">split</span>(DGE_table, DGE_table<span class="sc">$</span>cluster)</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="fu">unlist</span>(<span class="fu">lapply</span>(DGE_list, nrow))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   0    1    2    3    4    5    6    7    8    9 
3166 3314 2499 4072 2556 2036 3861 2218 2479 3343 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute differential gene expression in reference dataset (that has cell annotation)</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>( reference, <span class="at">value =</span> <span class="st">"cell_type"</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>reference_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  reference ,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.pct =</span> .<span class="dv">1</span> , </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.diff.pct =</span> .<span class="dv">2</span>,</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">only.pos =</span> T, </span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">max.cells.per.ident =</span> <span class="dv">20</span> ,</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">return.thresh =</span> <span class="dv">1</span> </span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the top cell marker genes in reference dataset</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="co"># select top 50 with hihgest foldchange among top 100 signifcant genes.</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>reference_markers <span class="ot">&lt;-</span> reference_markers [ <span class="fu">order</span>(reference_markers<span class="sc">$</span>avg_log2FC,<span class="at">decreasing =</span> T), ]</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>reference_markers <span class="sc">%&gt;%</span> </span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> </span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="sc">-</span><span class="dv">100</span>, p_val) <span class="sc">%&gt;%</span> </span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="dv">50</span>, avg_log2FC) <span class="ot">-&gt;</span> top50_cell_selection</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform the markers into a list</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>ref_list <span class="ot">=</span> <span class="fu">split</span>(top50_cell_selection<span class="sc">$</span>gene, top50_cell_selection<span class="sc">$</span>cluster)</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="fu">unlist</span>(<span class="fu">lapply</span>(ref_list, length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> CD8 T cell  CD4 T cell       cMono      B cell     NK cell         pDC 
         30          15          50          50          50          50 
     ncMono         cDC Plasma cell 
         50          50          50 </code></pre>
</div>
</div>
<p>Now we can run GSEA for the DEGs from our dataset and check for enrichment of top DEGs in the reference dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(fgsea))</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># run fgsea for each of the clusters in the list</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(DGE_list, <span class="cf">function</span>(x){</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  gene_rank <span class="ot">&lt;-</span> <span class="fu">setNames</span>(x<span class="sc">$</span>avg_log2FC, x<span class="sc">$</span>gene)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>( <span class="at">pathways=</span>ref_list, <span class="at">stats=</span>gene_rank,<span class="at">nperm=</span><span class="dv">10000</span>)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fgseaRes)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">names</span>(DGE_list)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="co"># You can filter and resort the table based on ES, NES or pvalue</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>pval <span class="sc">&lt;</span> <span class="fl">0.1</span> , ]} )</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>size <span class="sc">&gt;</span> <span class="dv">2</span> , ]} )</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[<span class="fu">order</span>(x<span class="sc">$</span>NES,<span class="at">decreasing =</span> T), ]} )</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`0`
   pathway        pval        padj        ES      NES nMoreExtreme size
1:   cMono 0.000099990 0.000299970 0.9604823 2.082162            0   48
2:  ncMono 0.000099990 0.000299970 0.8400708 1.817122            0   46
3:     cDC 0.000099990 0.000299970 0.8137532 1.753954            0   43
4:     pDC 0.001002506 0.002255639 0.7623781 1.564625            9   21
5:  B cell 0.007857359 0.014143246 0.7385015 1.483444           77   16
6: NK cell 0.016469377 0.024704066 0.7615592 1.466565          159   11
                                    leadingEdge
1:      S100A8,S100A9,LYZ,S100A12,VCAN,FCN1,...
2:     CTSS,TYMP,CST3,S100A11,AIF1,SERPINA1,...
3:              LYZ,GRN,TYMP,CST3,AIF1,SPI1,...
4:         GRN,MS4A6A,CST3,MPEG1,CTSB,TGFBI,...
5: NCF1,LY86,MARCH1,HLA-DRB5,POU2F2,PHACTR1,...
6:       TYROBP,FCER1G,SRGN,CCL3,CD63,MYO1F,...

$`1`
       pathway         pval         padj        ES      NES nMoreExtreme size
1:  CD8 T cell 0.0001000901 0.0003503153 0.9472460 2.204879            0   29
2:     NK cell 0.0001000400 0.0003503153 0.8485185 1.988182            0   32
3:  CD4 T cell 0.0015444015 0.0036036036 0.8753374 1.705631           13    7
4: Plasma cell 0.0866606625 0.1172265547 0.5478953 1.278696          865   30
                                  leadingEdge
1:          CD8A,CD3D,GZMH,CCL5,CD3G,CD8B,...
2:          CCL5,NKG7,GZMA,GZMM,CST7,CCL4,...
3:                CD3D,CD3G,CD3E,IL7R,PIK3IP1
4: FKBP11,PRDM1,PEBP1,SEC11C,PPIB,SELENOS,...

$`2`
       pathway         pval        padj        ES      NES nMoreExtreme size
1:      B cell 0.0001000000 0.000800000 0.8988279 2.015474            0   46
2:         cDC 0.0003053746 0.001221498 0.8772318 1.797614            2   14
3:         pDC 0.0011067512 0.002951336 0.7753617 1.628458           10   18
4: Plasma cell 0.0835030550 0.118339040 0.7150909 1.361282          778    8
5:      ncMono 0.0887542797 0.118339040 0.8425481 1.344170          673    3
                                            leadingEdge
1:      CD79A,TCL1A,LINC00926,MS4A1,CD79B,TNFRSF13C,...
2: CD74,HLA-DQB1,HLA-DRA,HLA-DPB1,HLA-DRB1,HLA-DQA1,...
3:               CD74,TCF4,BCL11A,IRF8,HERPUD1,SPIB,...
4:                PLPP5,ISG20,HERPUD1,MZB1,ITM2C,JCHAIN
5:                                  HLA-DPA1,POU2F2,LYN

$`3`
       pathway         pval         padj        ES      NES nMoreExtreme size
1:     NK cell 0.0001000000 0.0004019697 0.9400385 2.445529            0   50
2:  CD8 T cell 0.0001004924 0.0004019697 0.9179690 2.251090            0   24
3:         pDC 0.0006280750 0.0016748665 0.8051689 1.784358            5   11
4:      ncMono 0.0197270329 0.0394540658 0.7970971 1.570700          171    6
5: Plasma cell 0.0248422318 0.0397475709 0.5710832 1.429502          247   30
                                  leadingEdge
1:         SPON2,PRF1,GNLY,GZMB,CD7,CLIC3,...
2:        PRF1,GNLY,GZMB,CTSW,NKG7,FGFBP2,...
3:  GZMB,PLAC8,C12orf75,RRBP1,ALOX5AP,HSP90B1
4:                    FCGR3A,IFITM2,RHOC,HES4
5: CD38,FKBP11,SLAMF7,HSP90B1,SDF2L1,PPIB,...

$`4`
   pathway       pval      padj        ES      NES nMoreExtreme size
1:  B cell 0.03324453 0.1496004 0.8571390 1.470997          286    5
2:  ncMono 0.01910191 0.1496004 0.6490433 1.366874          190   35
                                leadingEdge
1:                PDLIM1,HLA-DRB5,STX7,NCF1
2: OAZ1,TIMP1,FKBP1A,CST3,IFITM3,FCER1G,...

$`5`
      pathway         pval         padj        ES      NES nMoreExtreme size
1: CD4 T cell 0.0001016984 0.0007118885 0.9134477 1.799050            0   14
2: CD8 T cell 0.0014178209 0.0049623732 0.8957908 1.612853           12    7
                         leadingEdge
1: IL7R,LTB,LDHB,RCAN3,NOSIP,MAL,...
2:      IL32,CD3D,CD3E,CD2,CD3G,CD8B

$`6`
       pathway         pval         padj        ES      NES nMoreExtreme size
1:     NK cell 0.0001000000 0.0004010829 0.9241341 2.407067            0   45
2:  CD8 T cell 0.0001002707 0.0004010829 0.9155580 2.278493            0   27
3:      ncMono 0.0029816514 0.0079510703 0.8599246 1.697683           25    6
4:         pDC 0.0243642941 0.0487285881 0.7067908 1.543200          228   10
5:  CD4 T cell 0.0422912206 0.0676659529 0.8575282 1.445832          315    3
6: Plasma cell 0.0991694186 0.1322258915 0.5055153 1.281763          990   32
                                  leadingEdge
1:        FGFBP2,GNLY,NKG7,CST7,GZMB,CTSW,...
2:        FGFBP2,GNLY,NKG7,CST7,GZMB,CTSW,...
3:                         FCGR3A,IFITM2,RHOC
4:  GZMB,C12orf75,HSP90B1,ALOX5AP,PLAC8,RRBP1
5:                                  CD3E,CD3D
6: PRDM1,FKBP11,HSP90B1,PPIB,SPCS2,SLAMF7,...

$`7`
      pathway         pval         padj        ES      NES nMoreExtreme size
1: CD4 T cell 0.0001016467 0.0006098801 0.9187840 1.953882            0   14
2: CD8 T cell 0.0744383728 0.1488767456 0.7949413 1.357119          612    4
                            leadingEdge
1: IL7R,TCF7,LTB,TSHZ2,PIK3IP1,LEF1,...
2:                   CD3E,CD3G,CD3D,CD2

$`8`
       pathway         pval        padj        ES      NES nMoreExtreme size
1:      B cell 0.0000999900 0.000304909 0.8990233 1.910958            0   45
2:         cDC 0.0001016363 0.000304909 0.8963109 1.763319            0   14
3:         pDC 0.0001011122 0.000304909 0.8409385 1.676958            0   16
4: Plasma cell 0.0010163635 0.002286818 0.8262721 1.625531            9   14
5:      ncMono 0.0046046573 0.008288383 0.9587620 1.511378           34    3
                                            leadingEdge
1:        CD79A,MS4A1,BANK1,CD74,TNFRSF13C,HLA-DQA1,...
2: CD74,HLA-DQA1,HLA-DRA,HLA-DPB1,HLA-DQB1,HLA-DPA1,...
3:             CD74,JCHAIN,SPIB,HERPUD1,TCF4,CCDC50,...
4:                JCHAIN,HERPUD1,ISG20,ITM2C,MZB1,PEBP1
5:                                      HLA-DPA1,POU2F2

$`9`
   pathway        pval         padj        ES      NES nMoreExtreme size
1:  ncMono 0.000099990 0.0002666933 0.9569912 2.036762            0   49
2:   cMono 0.000100010 0.0002666933 0.8930249 1.861037            0   33
3:     cDC 0.000100000 0.0002666933 0.8435239 1.767796            0   37
4: NK cell 0.006547315 0.0130946292 0.7881824 1.522179           63   12
5:  B cell 0.031278378 0.0500454041 0.6998417 1.386118          309   16
6:     pDC 0.054854026 0.0731387009 0.6815870 1.342377          542   15
                                              leadingEdge
1:                CDKN1C,LST1,FCGR3A,COTL1,AIF1,MS4A7,...
2:               LST1,COTL1,AIF1,SERPINA1,FCER1G,PSAP,...
3:                   LST1,COTL1,AIF1,FCER1G,SPI1,CST3,...
4:             FCGR3A,RHOC,FCER1G,TYROBP,IFITM2,MYO1F,...
5: HLA-DPA1,POU2F2,HLA-DRB5,HLA-DPB1,HLA-DRA,HLA-DRB1,...
6:                    CST3,NPC2,PLD4,MPEG1,TGFBI,CTSB,...</code></pre>
</div>
</div>
<p>Selecing top significant overlap per cluster, we can now rename the clusters according to the predicted labels. OBS! Be aware that if you have some clusters that have non-significant p-values for all the gene sets, the cluster label will not be very reliable. Also, the gene sets you are using may not cover all the celltypes you have in your dataset and hence predictions may just be the most similar celltype. Also, some of the clusters have very similar p-values to multiple celltypes, for instance the ncMono and cMono celltypes are equally good for some clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(res,<span class="cf">function</span>(x){<span class="fu">as.data.frame</span>(x)[<span class="dv">1</span>,<span class="dv">1</span>]}))</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>alldata<span class="sc">$</span>ref_gsea <span class="ot">&lt;-</span> new.cluster.ids[<span class="fu">as.character</span>(alldata<span class="sc">@</span>active.ident)]</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>( <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(alldata,<span class="at">label =</span> T,<span class="at">group.by =</span> <span class="st">"CCA_snn_res.0.5"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(alldata,<span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"ref_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-21-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Compare to results with the other celltype prediction methods in the ctrl_13 sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>ctrl<span class="sc">$</span>ref_gsea <span class="ot">=</span> alldata<span class="sc">$</span>ref_gsea[alldata<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"ctrl_13"</span>]</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>( <span class="at">ncol =</span> <span class="dv">3</span>,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl,<span class="at">label =</span> T,<span class="at">group.by =</span> <span class="st">"ref_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"GSEA"</span>),</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl,<span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"predicted.id"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"LabelTransfer"</span>),</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl,<span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"scpred_prediction"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"scPred"</span>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-22-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="1536"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_gsea_annot" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="meta-ct_gsea_annot"><span class="header-section-number">6.2</span> With annotated gene sets</h3>
<p>Read in the gene lists and do some filtering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the human marker table</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"data/cellmarker/Human_cell_markers.txt"</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers [ markers<span class="sc">$</span>speciesType <span class="sc">==</span> <span class="st">"Human"</span>, ]</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers [ markers<span class="sc">$</span>cancerType <span class="sc">==</span> <span class="st">"Normal"</span>, ]</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Filter by tissue (to reduce computational time and have tissue-specific classification)</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(<span class="fu">unique</span>(markers<span class="sc">$</span>tissueType))</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">"blood"</span>,<span class="fu">unique</span>(markers<span class="sc">$</span>tissueType),<span class="at">value =</span> T)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers [ markers<span class="sc">$</span>tissueType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Blood"</span>,<span class="st">"Venous blood"</span>,</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>                                                <span class="st">"Serum"</span>,<span class="st">"Plasma"</span>,</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>                                                <span class="st">"Spleen"</span>,<span class="st">"Bone marrow"</span>,<span class="st">"Lymph node"</span>), ]</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="co"># remove strange characters etc.</span></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>( <span class="fu">unique</span>(markers<span class="sc">$</span>cellName) , <span class="cf">function</span>(x){</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">paste</span>(markers<span class="sc">$</span>geneSymbol[markers<span class="sc">$</span>cellName <span class="sc">==</span> x],<span class="at">sep=</span><span class="st">","</span>)</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"[[]|[]]| |-"</span>,<span class="st">","</span>,x)</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>( x , <span class="at">split =</span> <span class="st">","</span>))</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">unique</span>(x [ <span class="sc">!</span> x <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">""</span>,<span class="st">"NA"</span>,<span class="st">"family"</span>) ])</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">casefold</span>(x,<span class="at">upper =</span> T)</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(celltype_list) <span class="ot">&lt;-</span> <span class="fu">unique</span>(markers<span class="sc">$</span>cellName)</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a><span class="co"># celltype_list &lt;- lapply(celltype_list , function(x) {x[1:min(length(x),50)]} )</span></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> celltype_list[ <span class="fu">unlist</span>(<span class="fu">lapply</span>(celltype_list,length)) <span class="sc">&lt;</span> <span class="dv">100</span> ]</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> celltype_list[ <span class="fu">unlist</span>(<span class="fu">lapply</span>(celltype_list,length)) <span class="sc">&gt;</span> <span class="dv">5</span> ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "Abdominal adipose tissue"       "Adipose tissue"                
  [3] "Adrenal gland"                  "Adventitia"                    
  [5] "Airway epithelium"              "Alveolus"                      
  [7] "Amniotic fluid"                 "Amniotic membrane"             
  [9] "Antecubital vein"               "Anterior cruciate ligament"    
 [11] "Artery"                         "Ascites"                       
 [13] "Bladder"                        "Blood"                         
 [15] "Blood vessel"                   "Bone"                          
 [17] "Bone marrow"                    "Brain"                         
 [19] "Breast"                         "Bronchoalveolar system"        
 [21] "Brown adipose tissue"           "Cartilage"                     
 [23] "Chorionic villus"               "Colon"                         
 [25] "Colorectum"                     "Cornea"                        
 [27] "Corneal endothelium"            "Corneal epithelium"            
 [29] "Corpus luteum"                  "Decidua"                       
 [31] "Deciduous tooth"                "Dental pulp"                   
 [33] "Dermis"                         "Dorsolateral prefrontal cortex"
 [35] "Duodenum"                       "Embryo"                        
 [37] "Embryoid body"                  "Embryonic brain"               
 [39] "Embryonic prefrontal cortex"    "Embryonic stem cell"           
 [41] "Endometrium"                    "Endometrium stroma"            
 [43] "Epithelium"                     "Esophagus"                     
 [45] "Eye"                            "Fat pad"                       
 [47] "Fetal brain"                    "Fetal gonad"                   
 [49] "Fetal kidney"                   "Fetal liver"                   
 [51] "Fetal pancreas"                 "Foreskin"                      
 [53] "Gastric corpus"                 "Gastric epithelium"            
 [55] "Gastric gland"                  "Gastrointestinal tract"        
 [57] "Germ"                           "Gingiva"                       
 [59] "Gonad"                          "Gut"                           
 [61] "Hair follicle"                  "Heart"                         
 [63] "Hippocampus"                    "Inferior colliculus"           
 [65] "Intervertebral disc"            "Intestinal crypt"              
 [67] "Intestine"                      "Jejunum"                       
 [69] "Kidney"                         "Lacrimal gland"                
 [71] "Large intestine"                "Laryngeal squamous epithelium" 
 [73] "Ligament"                       "Limbal epithelium"             
 [75] "Liver"                          "Lung"                          
 [77] "Lymph"                          "Lymph node"                    
 [79] "Lymphoid tissue"                "Mammary epithelium"            
 [81] "Meniscus"                       "Midbrain"                      
 [83] "Molar"                          "Muscle"                        
 [85] "Myocardium"                     "Myometrium"                    
 [87] "Nasal concha"                   "Nasal epithelium"              
 [89] "Nerve"                          "Nucleus pulposus"              
 [91] "Optic nerve"                    "Oral mucosa"                   
 [93] "Osteoarthritic cartilage"       "Ovarian cortex"                
 [95] "Ovarian follicle"               "Ovary"                         
 [97] "Oviduct"                        "Pancreas"                      
 [99] "Pancreatic acinar tissue"       "Pancreatic islet"              
[101] "Periodontal ligament"           "Periosteum"                    
[103] "Peripheral blood"               "Placenta"                      
[105] "Plasma"                         "Pluripotent stem cell"         
[107] "Premolar"                       "Primitive streak"              
[109] "Prostate"                       "Pyloric gland"                 
[111] "Rectum"                         "Renal glomerulus"              
[113] "Retina"                         "Retinal pigment epithelium"    
[115] "Salivary gland"                 "Scalp"                         
[117] "Sclerocorneal tissue"           "Seminal plasma"                
[119] "Serum"                          "Sinonasal mucosa"              
[121] "Skeletal muscle"                "Skin"                          
[123] "Small intestinal crypt"         "Small intestine"               
[125] "Spinal cord"                    "Spleen"                        
[127] "Splenic red pulp"               "Sputum"                        
[129] "Stomach"                        "Subcutaneous adipose tissue"   
[131] "Submandibular gland"            "Sympathetic ganglion"          
[133] "Synovial fluid"                 "Synovium"                      
[135] "Tendon"                         "Testis"                        
[137] "Thymus"                         "Thyroid"                       
[139] "Tonsil"                         "Tooth"                         
[141] "Umbilical cord"                 "Umbilical cord blood"          
[143] "Umbilical vein"                 "Undefined"                     
[145] "Urine"                          "Uterus"                        
[147] "Vagina"                         "Venous blood"                  
[149] "Visceral adipose tissue"        "Vocal fold"                    
[151] "Whartons jelly"                 "White adipose tissue"          
[1] "Peripheral blood"     "Umbilical cord blood" "Venous blood"        </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run fgsea for each of the clusters in the list</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(DGE_list, <span class="cf">function</span>(x){</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  gene_rank <span class="ot">&lt;-</span> <span class="fu">setNames</span>(x<span class="sc">$</span>avg_log2FC, x<span class="sc">$</span>gene)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>( <span class="at">pathways=</span>celltype_list, <span class="at">stats=</span>gene_rank,<span class="at">nperm=</span><span class="dv">10000</span>)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fgseaRes)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">names</span>(DGE_list)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="co"># You can filter and resort the table based on ES, NES or pvalue</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>pval <span class="sc">&lt;</span> <span class="fl">0.01</span> , ]} )</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[ x<span class="sc">$</span>size <span class="sc">&gt;</span> <span class="dv">5</span> , ]} )</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {x[<span class="fu">order</span>(x<span class="sc">$</span>NES,<span class="at">decreasing =</span> T), ]} )</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="co"># show top 3 for each cluster.</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(res,head,<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`0`
                    pathway         pval        padj        ES      NES
1:               Neutrophil 0.0002024496 0.003542869 0.8523815 1.711647
2:   CD1C+_B dendritic cell 0.0000999900 0.003499650 0.7860707 1.711588
3: Mesenchymal stromal cell 0.0033194132 0.029044866 0.8633935 1.609837
   nMoreExtreme size                             leadingEdge
1:            1   15 CD14,CSF3R,JAML,C5AR1,FCGR2A,FCGR1A,...
2:            0   49 S100A8,S100A9,LYZ,S100A12,VCAN,FCN1,...
3:           30    8              CD14,VIM,ANPEP,CD44,PECAM1

$`1`
                    pathway         pval        padj        ES      NES
1:                   T cell 0.0001105217 0.000793454 0.9634109 1.880733
2:             Myeloid cell 0.0001021868 0.000793454 0.8522776 1.852164
3: Mesenchymal stromal cell 0.0038588128 0.012348201 0.8709271 1.649822
   nMoreExtreme size                        leadingEdge
1:            0    7    CD8A,CD3D,CD3G,CD3E,CD2,CD5,...
2:            0   14 CD3D,CD3G,CD3E,CD81,PTPRC,CD84,...
3:           33    6               CD81,B2M,PTPRC,ITGB1

$`2`
   pathway         pval        padj        ES      NES nMoreExtreme size
1:  B cell 0.0001029866 0.002780639 0.9047629 1.801854            0   11
                            leadingEdge
1: CD79A,MS4A1,FCER2,PAX5,CD24,CD19,...

$`3`
                       pathway         pval        padj        ES      NES
1:      CD1C+_A dendritic cell 0.0002217295 0.005127494 0.9084688 1.849273
2:         Natural killer cell 0.0005543237 0.005127494 0.8964502 1.824808
3: AXL+SIGLEC6+ dendritic cell 0.0004023335 0.005127494 0.7337906 1.776492
   nMoreExtreme size                                leadingEdge
1:            1    7                    AREG,LPCAT1,ADAM8,NR4A2
2:            4    7              KLRD1,FCGR3A,KLRC1,NCAM1,NCR1
3:            3   22 PTGDS,CTSW,BHLHE40,TSEN54,CX3CR1,PLAC8,...

$`4`
Empty data.table (0 rows and 8 cols): pathway,pval,padj,ES,NES,nMoreExtreme...

$`5`
   pathway         pval        padj        ES      NES nMoreExtreme size
1:  T cell 0.0001093135 0.003060778 0.9460238 1.702717            0    7
              leadingEdge
1: CD3D,CD3E,CD2,CD5,CD3G

$`6`
               pathway         pval        padj        ES      NES nMoreExtreme
1: Natural killer cell 0.0001092180 0.004041066 0.8961044 1.884108            0
2:            Platelet 0.0007526072 0.009282156 0.8302022 1.783614            6
3:        Myeloid cell 0.0007159660 0.009282156 0.7635915 1.778575            6
   size                          leadingEdge
1:    8 KLRD1,FCGR3A,CD3E,CD2,NCR1,NCAM1,...
2:    9     CCL5,CD63,SPN,BSG,CD47,CD226,...
3:   15 FCGR3A,CD3E,CD81,ZAP70,IL2RB,SPN,...

$`7`
Empty data.table (0 rows and 8 cols): pathway,pval,padj,ES,NES,nMoreExtreme...

$`8`
   pathway        pval       padj        ES      NES nMoreExtreme size
1:  B cell 0.001330332 0.03724928 0.8203607 1.610929           12   13
                                  leadingEdge
1: CD79A,MS4A1,POU2AF1,CD24,FCGR2B,POU2F2,...

$`9`
        pathway         pval       padj        ES      NES nMoreExtreme size
1: Myeloid cell 0.0002004008 0.00741483 0.8081754 1.642042            1   22
2:   Neutrophil 0.0032241290 0.03976426 0.8299506 1.566163           30   10
                              leadingEdge
1: FCGR3A,CSF1R,PECAM1,CD68,ITGAX,SPN,...
2:       FCGR3A,PECAM1,ITGAX,C5AR1,FCGR2A</code></pre>
</div>
</div>
<p>#CT_GSEA8:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(res,<span class="cf">function</span>(x){<span class="fu">as.data.frame</span>(x)[<span class="dv">1</span>,<span class="dv">1</span>]}))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>alldata<span class="sc">$</span>cellmarker_gsea <span class="ot">&lt;-</span> new.cluster.ids[<span class="fu">as.character</span>(alldata<span class="sc">@</span>active.ident)]</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>( <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(alldata,<span class="at">label =</span> T,<span class="at">group.by =</span> <span class="st">"ref_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(alldata,<span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"cellmarker_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-25-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="1056"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you think that the methods overlap well? Where do you see the most inconsistencies?</p>
</div>
</div>
<p>In this case we do not have any ground truth, and we cannot say which method performs best. You should keep in mind, that any celltype classification method is just a prediction, and you still need to use your common sense and knowledge of the biological system to judge if the results make sense.</p>
<p>Finally, lets save the data with predictions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(ctrl,<span class="st">"data/results/ctrl13_qc_dr_int_cl_celltype.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="meta-session" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">7</span> Session info</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.3 (2023-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] fgsea_1.24.0       caret_6.0-94       lattice_0.20-45    scPred_1.9.2      
 [5] rafalib_1.0.0      pheatmap_1.0.12    ggplot2_3.4.2      cowplot_1.1.1     
 [9] dplyr_1.1.1        SeuratObject_4.1.3 Seurat_4.3.0.1    

loaded via a namespace (and not attached):
  [1] fastmatch_1.1-3        plyr_1.8.8             igraph_1.5.0          
  [4] lazyeval_0.2.2         sp_2.0-0               splines_4.2.3         
  [7] BiocParallel_1.32.6    listenv_0.9.0          scattermore_1.2       
 [10] digest_0.6.33          foreach_1.5.2          htmltools_0.5.5       
 [13] fansi_1.0.4            magrittr_2.0.3         tensor_1.5            
 [16] cluster_2.1.4          ROCR_1.0-11            limma_3.54.2          
 [19] recipes_1.0.6          globals_0.16.2         gower_1.0.1           
 [22] matrixStats_1.0.0      hardhat_1.3.0          timechange_0.2.0      
 [25] spatstat.sparse_3.0-2  colorspace_2.1-0       ggrepel_0.9.3         
 [28] xfun_0.39              crayon_1.5.2           jsonlite_1.8.7        
 [31] progressr_0.13.0       spatstat.data_3.0-1    survival_3.5-3        
 [34] zoo_1.8-12             iterators_1.0.14       glue_1.6.2            
 [37] polyclip_1.10-4        gtable_0.3.3           ipred_0.9-14          
 [40] leiden_0.4.3           kernlab_0.9-32         future.apply_1.11.0   
 [43] abind_1.4-5            scales_1.2.1           spatstat.random_3.1-5 
 [46] miniUI_0.1.1.1         Rcpp_1.0.10            viridisLite_0.4.2     
 [49] xtable_1.8-4           reticulate_1.30        stats4_4.2.3          
 [52] lava_1.7.2.1           prodlim_2023.03.31     htmlwidgets_1.6.2     
 [55] httr_1.4.6             RColorBrewer_1.1-3     ellipsis_0.3.2        
 [58] ica_1.0-3              pkgconfig_2.0.3        farver_2.1.1          
 [61] nnet_7.3-18            uwot_0.1.16            deldir_1.0-9          
 [64] utf8_1.2.3             tidyselect_1.2.0       labeling_0.4.2        
 [67] rlang_1.1.1            reshape2_1.4.4         later_1.3.1           
 [70] munsell_0.5.0          tools_4.2.3            cli_3.6.1             
 [73] generics_0.1.3         ggridges_0.5.4         evaluate_0.21         
 [76] stringr_1.5.0          fastmap_1.1.1          yaml_2.3.7            
 [79] goftest_1.2-3          ModelMetrics_1.2.2.2   knitr_1.43            
 [82] fitdistrplus_1.1-11    purrr_1.0.1            RANN_2.6.1            
 [85] pbapply_1.7-2          future_1.33.0          nlme_3.1-162          
 [88] mime_0.12              compiler_4.2.3         rstudioapi_0.14       
 [91] beeswarm_0.4.0         plotly_4.10.2          png_0.1-8             
 [94] spatstat.utils_3.0-3   tibble_3.2.1           stringi_1.7.12        
 [97] Matrix_1.6-0           vctrs_0.6.3            pillar_1.9.0          
[100] lifecycle_1.0.3        spatstat.geom_3.2-1    lmtest_0.9-40         
[103] RcppAnnoy_0.0.21       data.table_1.14.8      irlba_2.3.5.1         
[106] httpuv_1.6.11          patchwork_1.1.2        R6_2.5.1              
[109] promises_1.2.0.1       KernSmooth_2.23-20     gridExtra_2.3         
[112] vipor_0.4.5            parallelly_1.36.0      codetools_0.2-19      
[115] MASS_7.3-58.2          withr_2.5.0            sctransform_0.3.5     
[118] harmony_0.1.1          parallel_4.2.3         grid_4.2.3            
[121] rpart_4.1.19           timeDate_4022.108      tidyr_1.3.0           
[124] class_7.3-21           rmarkdown_2.21         Rtsne_0.16            
[127] spatstat.explore_3.2-1 pROC_1.18.5            shiny_1.7.4           
[130] lubridate_1.9.2        ggbeeswarm_0.7.2      </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2023 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.433</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"openEffect":"zoom","selector":".lightbox","closeEffect":"zoom","descPosition":"bottom","loop":true});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>