---
author: "Åsa Björklund  &  Paulo Czarnewski"
date: "Sept 13, 2019"
output:
  html_document:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    keep_md: yes
    fig_caption: true
  html_notebook:
    self_contained: true
    highlight: tango
    df_print: paged
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=12,tidy=TRUE)
```
<style>
h1, .h1, h2, .h2, h3, .h3, h4, .h4 { margin-top: 50px }
p.caption {font-size: 0.9em;font-style: italic;color: grey;margin-right: 10%;margin-left: 10%;text-align: justify}
</style>

# Differential gene expression

In this tutorial we will cover about Differetial gene expression, which comprises an extensive range of topics and methods. In single cell, differential expresison can have multiple functionalities such as of identifying marker genes for cell populations, as well as differentially regulated genes across conditions (healthy vs control). We will also exercise on how to account the batch information in your test.

We can first load the data from the clustering session. Moreover, we can already decide which clustering resolution to use. First let's define using the `louvain_2` clustering to identifying differentially expressed genes.  

```{r}
suppressPackageStartupMessages({
  library(scater)
  library(scran)
  #library(venn)
  library(cowplot)
  library(ggplot2)
  library(rafalib)
  library(pheatmap)
  library(igraph)
  library(dplyr)
})

sce <- readRDS("data/3pbmc_qc_dr_int_cl.rds")
```

## Cell marker genes
***

Let us first compute a ranking for the highly differential genes in each cluster. There are many different tests and parameters to be chosen that can be used to refine your results. When looking for marker genes, we want genes that are positivelly expressed in a cell type and possibly not expressed in the others.

```{r}
#Compute differentiall expression
markers_genes <- findMarkers( x = sce,
                              clusters = sce$kmeans_5,
                              lfc=.5,
                              pval.type = "all",
                              direction = "up")

#List of dataFrames with the results for each cluster
markers_genes

#Visualizing the expression of one
markers_genes[["1"]]
```

We can now select the top 25 up regulated genes for plotting.

```{r}
#Colect the top 25 genes for each cluster and put the into a single table
top25 <- lapply( names(markers_genes), function(x) { temp <- markers_genes[[x]][1:25, 1:2] ; temp$gene <- rownames(markers_genes[[x]])[1:25] ; temp$cluster <- x ; return(temp) } )
top25 <- as_tibble(do.call(rbind, top25))
top25
```

We can now select the top 25 up regulated genes for plotting.

```{r}
mypar(1,5,mar=c(4,6,3,1))
for(i in unique(top25$cluster)){
  barplot( sort( setNames(-log10(top25$p.value), top25$gene) [top25$cluster == i], F),
           horiz = T,las=1 ,main=paste0(i," vs. rest"),border = "white", yaxs="i",xlab="-log10FC" )
  abline(v=c(0,-log10(0.05)),lty=c(1,2))
}
```

We can visualize them as a heatmap. Here we are selecting the top 5.

```{r}
as_tibble(top25) %>% group_by(cluster)  %>% top_n(-5, p.value) -> top5

scater::plotHeatmap(sce[,order(sce$kmeans_5)], features = unique(top5$gene) ,
                    center=T , zlim = c(-3,3) ,
                    colour_columns_by= "kmeans_5",
                    show_colnames=F , cluster_cols=F,
                    fontsize_row=6,
                    color=colorRampPalette(c("purple","black","yellow"))(90)
                    )
```

We can also plot a violin plot for each gene.

```{r,fig.height=10,fig.width=10}
scater::plotExpression(sce, features = unique(top5$gene), x = "kmeans_5", ncol = 5, colour_by = "kmeans_5", scales="free")
```


## Differential expression across conditions
***

The second way of computing differential expression is to answer which genes are differentially expressed within a cluster. For example, in our case we have libraries comming from 2 different library preparation methods (batches) and we would like to know which genes are influenced the most in a particular cell type. The same concenpt applies if you have instead two or more biological groups (control vs treated, time#0 vs time#1 vs time#2, etc).

For this end, we will first subset our data for the desired cell cluster, then change the cell identities to the variable of comparison (which now in our case is the "Chemistry").

```{r,fig.height=10,fig.width=10}
#Filter cells from that cluster
cell_selection <- sce[ , sce$kmeans_5 == 4 ]
cell_selection$Chemistry <- ifelse(cell_selection$sample_id == "v2.1k","v2","v3")

#Compute differentiall expression
DGE_cell_selection <- findMarkers( x = cell_selection,
                              clusters = cell_selection$Chemistry,
                              lfc=.5,
                              pval.type = "all",
                              direction = "down")
top5_cell_selection <- lapply( names(DGE_cell_selection), function(x) { temp <- DGE_cell_selection[[x]][1:5, 1:2] ; temp$gene <- rownames(DGE_cell_selection[[x]])[1:5] ; temp$cluster <- x ; return(temp) } )
top5_cell_selection <- as_tibble(do.call(rbind, top5_cell_selection))
top5_cell_selection
```

We can now plot the expression across the "Chemistry".

```{r,fig.height=4,fig.width=10}
scater::plotExpression(cell_selection, features = unique(top5_cell_selection$gene), x = "Chemistry", ncol = 5, colour_by = "Chemistry")
```

We can clearly see some patterns across them. Those are the genes that impact the most on your batches (see the dimensionality reduction and integration exercises for more details). We can plot those genes using the integrated and non-integrated UMAP for ilustration.

```{r,fig.asp=.28}
plotlist <- list()
for(i in c("JUND","RPS17","GNAS")){
  plotlist[[i]] <- plotReducedDim(sce,use_dimred = "UMAP_on_PCA",colour_by = i,by_exprs_values = "logcounts",add_ticks = F) +
  scale_fill_gradientn(colours = colorRampPalette(c("grey90","orange3","firebrick","firebrick","red","red" ))(10)) +
  ggtitle(label = i)+ theme(plot.title = element_text(size=20)) }
plot_grid(ncol=3, plotlist = plotlist)
```


Finally, lets save the integrated data for further analysis.

```{r}
saveRDS(sce,"data/3pbmc_qc_dr_int_cl_dge.rds")
```


### Session Info
***

```{r}
sessionInfo()
```



















